<!DOCTYPE HTML>
<html>
  <head>
    <meta charset="UTF-8">
    <title>AME results</title>
    <script type="text/javascript">
      //@JSON_VAR data
      var data = {
        "version": "5.0.3",
        "revision": "b3b3ae8035606cf9a9d0ee3ba8cc5a7f5a38f198",
        "release": "Sun Dec 02 18:41:45 2018 -0800",
        "program": "AME",
        "cmd": [
          "ame", "--oc", "3utr_more_vs_less/all_pearson_totalhits_0.25_fixed",
          "--method", "pearson", "--scoring", "totalhits", "--fix-partition",
          "196", "--hit-lo-fraction", "0.25", "--evalue-report-threshold",
          "1000", "3utr/meme_3utr_fasta_all.txt",
          "/home/diane/software/meme/share/meme-5.0.3/db/motif_databases/RNA/Ray2013_rbp_Mus_musculus.meme",
          "/home/diane/software/meme/share/meme-5.0.3/db/motif_databases/RNA/Ray2013_rbp_Homo_sapiens.meme"
        ],
        "options": {
          "scoring": "total_hits",
          "positive_list": "FASTA",
          "pvalue_method": "Pearson's correlation coefficient",
          "pseudocount": 0.1,
          "fisher_fasta_threshold": 0.001,
          "hit_lo_fraction": 0.25,
          "evalue_report_threshold": 1000,
          "noseq": true,
          "log_fscores": false,
          "log_pwmscores": false,
          "linreg_switchxy": true,
          "fix_partition": true,
          "partition": 196,
          "kmer": 2
        },
        "motif_dbs": [
          {
            "source": "/home/diane/software/meme/share/meme-5.0.3/db/motif_databases/RNA/Ray2013_rbp_Mus_musculus.meme",
            "count": 4
          }, {
            "source": "/home/diane/software/meme/share/meme-5.0.3/db/motif_databases/RNA/Ray2013_rbp_Homo_sapiens.meme",
            "count": 102
          }
        ],
        "alphabet": {
          "name": "RNA",
          "like": "rna",
          "ncore": 4,
          "symbols": [
            {
              "symbol": "A",
              "name": "Adenine",
              "colour": "CC0000"
            }, {
              "symbol": "C",
              "name": "Cytosine",
              "colour": "0000CC"
            }, {
              "symbol": "G",
              "name": "Guanine",
              "colour": "FFB300"
            }, {
              "symbol": "U",
              "aliases": "T",
              "name": "Uracil",
              "colour": "008000"
            }, {
              "symbol": "N",
              "aliases": "X.",
              "name": "Any base",
              "equals": "ACGU"
            }, {
              "symbol": "V",
              "name": "Not U",
              "equals": "ACG"
            }, {
              "symbol": "H",
              "name": "Not G",
              "equals": "ACU"
            }, {
              "symbol": "D",
              "name": "Not C",
              "equals": "AGU"
            }, {
              "symbol": "B",
              "name": "Not A",
              "equals": "CGU"
            }, {
              "symbol": "M",
              "name": "Amino",
              "equals": "AC"
            }, {
              "symbol": "R",
              "name": "Purine",
              "equals": "AG"
            }, {
              "symbol": "W",
              "name": "Weak",
              "equals": "AU"
            }, {
              "symbol": "S",
              "name": "Strong",
              "equals": "CG"
            }, {
              "symbol": "Y",
              "name": "Pyrimidine",
              "equals": "CU"
            }, {
              "symbol": "K",
              "name": "Keto",
              "equals": "GU"
            }
          ]
        },
        "background": {
          "source": "--sequences--",
          "file": "3utr/meme_3utr_fasta_all.txt",
          "frequencies": [0.263386, 0.217669, 0.221038, 0.297907]
        },
        "sequence_db": {
          "source": "3utr/meme_3utr_fasta_all.txt",
          "count": 211,
          "skipped": 0,
          "fasta_scores": true
        },
        "min_sequence_score": 0,
        "max_sequence_score": 299,
        "max_median_sequence_score": 36,
        "motifs": [
          {
            "db": 1,
            "id": "RNCMPT00018",
            "alt": "FUS",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.185248, 0.173961, 0.17341, 0.467383], 
              [0.157055, 0.150585, 0.536551, 0.155808], 
              [0.0181283, 0.945655, 0.0179176, 0.0183], 
              [0.0181283, 0.0179008, 0.945672, 0.0183], 
              [0.0181283, 0.945655, 0.0179176, 0.0183], 
              [0.0181283, 0.0179008, 0.945672, 0.0183], 
              [0.0410696, 0.788621, 0.0408589, 0.129449]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "1.27e-1",
            "evalue": "1.35e1",
            "pearsons_rho": 0.0818629,
            "m": -0.00868497,
            "b": -1.10979,
            "mean_square_error": 0.0324469
          }, {
            "db": 1,
            "id": "RNCMPT00045",
            "alt": "PPRC1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0878437, 0.360131, 0.464009, 0.0880155], 
              [0.0197243, 0.57086, 0.296727, 0.112689], 
              [0.0197243, 0.205958, 0.754422, 0.0198961], 
              [0.0197243, 0.940866, 0.0195136, 0.0198961], 
              [0.0197243, 0.110248, 0.850132, 0.0198961], 
              [0.0197243, 0.853614, 0.106765, 0.0198961], 
              [0.0659442, 0.339489, 0.528451, 0.066116]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "1.61e-1",
            "evalue": "1.70e1",
            "pearsons_rho": 0.0712953,
            "m": -0.00303412,
            "b": -1.10477,
            "mean_square_error": 0.0324733
          }, {
            "db": 1,
            "id": "RNCMPT00052",
            "alt": "RBM4",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0558099, 0.0555824, 0.832626, 0.0559816], 
              [0.00770938, 0.976911, 0.0074987, 0.00788113], 
              [0.00770938, 0.00748193, 0.976928, 0.00788113], 
              [0.00770938, 0.889688, 0.0947216, 0.00788113], 
              [0.00770938, 0.0947048, 0.889705, 0.00788113], 
              [0.00770938, 0.221694, 0.676229, 0.0943677], 
              [0.0574994, 0.234576, 0.568912, 0.139012]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "2.06e-1",
            "evalue": "2.19e1",
            "pearsons_rho": 0.0589062,
            "m": -0.0100146,
            "b": -1.09573,
            "mean_square_error": 0.0324996
          }, {
            "db": 1,
            "id": "RNCMPT00113",
            "alt": "RBM4",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.031184, 0.11498, 0.745122, 0.108714], 
              [0.00679894, 0.815496, 0.170734, 0.00697068], 
              [0.00679894, 0.00657149, 0.979658, 0.00697068], 
              [0.00679894, 0.889613, 0.0966161, 0.00697068], 
              [0.00679894, 0.0965993, 0.88963, 0.00697068], 
              [0.00679894, 0.305907, 0.60301, 0.0842831], 
              [0.0800646, 0.0798372, 0.759862, 0.0802364]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "2.45e-1",
            "evalue": "2.60e1",
            "pearsons_rho": 0.0495631,
            "m": -0.00377447,
            "b": -1.09461,
            "mean_square_error": 0.0325162
          }, {
            "db": 1,
            "id": "RNCMPT00056",
            "alt": "RBM8A",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.195228, 0.113394, 0.577584, 0.113794], 
              [0.022177, 0.667404, 0.0219664, 0.288451], 
              [0.022177, 0.0994919, 0.855982, 0.0223488], 
              [0.022177, 0.855966, 0.0995086, 0.0223488], 
              [0.022177, 0.0994919, 0.855982, 0.0223488], 
              [0.022177, 0.843183, 0.112291, 0.0223488], 
              [0.0676168, 0.157526, 0.612477, 0.16238]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "2.53e-1",
            "evalue": "2.68e1",
            "pearsons_rho": 0.0477702,
            "m": -0.00295131,
            "b": -1.09693,
            "mean_square_error": 0.032519
          }, {
            "db": 1,
            "id": "RNCMPT00073",
            "alt": "SRSF7",
            "len": 8,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.203552, 0.126961, 0.542129, 0.12736], 
              [0.768215, 0.0771227, 0.0771395, 0.0775219], 
              [0.00979595, 0.97065, 0.00958526, 0.0099677], 
              [0.00979595, 0.0095685, 0.970667, 0.0099677], 
              [0.970878, 0.0095685, 0.00958526, 0.0099677], 
              [0.00979595, 0.97065, 0.00958526, 0.0099677], 
              [0.153237, 0.0579635, 0.730438, 0.0583627], 
              [0.339994, 0.13257, 0.394467, 0.13297]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "5.67e-1",
            "evalue": "6.01e1",
            "pearsons_rho": -0.0121784,
            "m": 0.00320858,
            "b": -1.07253,
            "mean_square_error": 0.0325538
          }, {
            "db": 1,
            "id": "RNCMPT00063",
            "alt": "SAMD4A",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.00741884, 0.00719139, 0.9778, 0.00759058], 
              [0.00741884, 0.977783, 0.00720815, 0.00759058], 
              [0.100968, 0.00719139, 0.272491, 0.61935], 
              [0.00741884, 0.00719139, 0.9778, 0.00759058], 
              [0.00741884, 0.00719139, 0.9778, 0.00759058], 
              [0.389066, 0.367589, 0.106221, 0.137124], 
              [0.189657, 0.608357, 0.100802, 0.101185]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "6.27e-1",
            "evalue": "6.64e1",
            "pearsons_rho": -0.023226,
            "m": 0.00166136,
            "b": -1.06569,
            "mean_square_error": 0.0325474
          }, {
            "db": 1,
            "id": "RNCMPT00044",
            "alt": "PCBP2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0473999, 0.776498, 0.0471892, 0.128914], 
              [0.0253343, 0.760897, 0.0251236, 0.188644], 
              [0.0253343, 0.310868, 0.0251236, 0.638673], 
              [0.106409, 0.310868, 0.0251236, 0.5576], 
              [0.108812, 0.73235, 0.0251236, 0.133715], 
              [0.0253343, 0.924035, 0.0251236, 0.025506], 
              [0.158491, 0.623459, 0.0688221, 0.149227]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "6.69e-1",
            "evalue": "7.09e1",
            "pearsons_rho": -0.0314512,
            "m": 0.00034836,
            "b": -1.06331,
            "mean_square_error": 0.0325401
          }, {
            "db": 1,
            "id": "RNCMPT00043",
            "alt": "PABPC4",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.775875, 0.10365, 0.111075, 0.00940053], 
              [0.788338, 0.104447, 0.0978141, 0.00940053], 
              [0.972581, 0.00900134, 0.0090181, 0.00940053], 
              [0.972581, 0.00900134, 0.0090181, 0.00940053], 
              [0.972581, 0.00900134, 0.0090181, 0.00940053], 
              [0.682641, 0.111548, 0.104336, 0.101477], 
              [0.680621, 0.105361, 0.204618, 0.00940053]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.08e-1",
            "evalue": "7.50e1",
            "pearsons_rho": -0.0393815,
            "m": 0.000676985,
            "b": -1.06184,
            "mean_square_error": 0.0325309
          }, {
            "db": 1,
            "id": "RNCMPT00160",
            "alt": "HNRNPH2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.056781, 0.121386, 0.76488, 0.0569528], 
              [0.0306507, 0.0923993, 0.846127, 0.0308224], 
              [0.0999173, 0.0704551, 0.822344, 0.00728312], 
              [0.758321, 0.0704551, 0.0711335, 0.100089], 
              [0.00711137, 0.0704551, 0.809255, 0.11318], 
              [0.00711137, 0.0704551, 0.915151, 0.00728312], 
              [0.137503, 0.136645, 0.649387, 0.0764652]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.40e-1",
            "evalue": "7.84e1",
            "pearsons_rho": -0.0462676,
            "m": 0.00115628,
            "b": -1.05632,
            "mean_square_error": 0.0325213
          }, {
            "db": 1,
            "id": "RNCMPT00002",
            "alt": "ANKHD1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.770667, 0.0763058, 0.0763226, 0.076705], 
              [0.00559396, 0.00536651, 0.983275, 0.00576571], 
              [0.983486, 0.00536651, 0.00538327, 0.00576571], 
              [0.00559396, 0.983258, 0.00538327, 0.00576571], 
              [0.00559396, 0.00536651, 0.983275, 0.00576571], 
              [0.228982, 0.102664, 0.00538327, 0.662972], 
              [0.426936, 0.0799974, 0.100334, 0.392732]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.46e-1",
            "evalue": "7.91e1",
            "pearsons_rho": -0.0476891,
            "m": 0.00726015,
            "b": -1.05956,
            "mean_square_error": 0.0325191
          }, {
            "db": 1,
            "id": "RNCMPT00150",
            "alt": "ESRP2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0859999, 0.0857725, 0.0857892, 0.742437], 
              [0.0384188, 0.0381914, 0.884799, 0.0385906], 
              [0.0141253, 0.0138979, 0.957679, 0.0142971], 
              [0.0141253, 0.0138979, 0.957679, 0.0142971], 
              [0.219703, 0.0376431, 0.704611, 0.0380423], 
              [0.714207, 0.0376431, 0.210108, 0.0380423], 
              [0.206478, 0.107904, 0.197294, 0.488325]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.64e-1",
            "evalue": "8.10e1",
            "pearsons_rho": -0.0518324,
            "m": 0.0011657,
            "b": -1.05216,
            "mean_square_error": 0.0325124
          }, {
            "db": 1,
            "id": "RNCMPT00165",
            "alt": "TIA1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.217876, 0.125897, 0.035239, 0.620988], 
              [0.0114557, 0.0112282, 0.011245, 0.966072], 
              [0.100063, 0.0112282, 0.011245, 0.877464], 
              [0.0114557, 0.0112282, 0.011245, 0.966072], 
              [0.0114557, 0.0112282, 0.011245, 0.966072], 
              [0.0114557, 0.0112282, 0.106815, 0.870502], 
              [0.0830845, 0.47024, 0.184231, 0.262444]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.66e-1",
            "evalue": "8.12e1",
            "pearsons_rho": -0.0521468,
            "m": 0.000596661,
            "b": -1.05566,
            "mean_square_error": 0.0325119
          }, {
            "db": 1,
            "id": "RNCMPT00077",
            "alt": "TIA1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.150808, 0.0627377, 0.0627544, 0.7237], 
              [0.0160377, 0.0158103, 0.0158271, 0.952324], 
              [0.0160377, 0.0158103, 0.0158271, 0.952324], 
              [0.0160377, 0.0158103, 0.0158271, 0.952324], 
              [0.0160377, 0.0158103, 0.0158271, 0.952324], 
              [0.0160377, 0.206883, 0.204854, 0.572225], 
              [0.08519, 0.0849625, 0.564748, 0.265099]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.70e-1",
            "evalue": "8.16e1",
            "pearsons_rho": -0.0530612,
            "m": 0.000632628,
            "b": -1.05575,
            "mean_square_error": 0.0325103
          }, {
            "db": 1,
            "id": "RNCMPT00032",
            "alt": "HuR",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.00775814, 0.00753069, 0.00754745, 0.977165], 
              [0.00775814, 0.00753069, 0.00754745, 0.977165], 
              [0.535079, 0.00753069, 0.177488, 0.279903], 
              [0.088791, 0.00753069, 0.18583, 0.717848], 
              [0.00775814, 0.00753069, 0.00754745, 0.977165], 
              [0.00775814, 0.00753069, 0.00754745, 0.977165], 
              [0.057977, 0.144098, 0.144035, 0.65389]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.73e-1",
            "evalue": "8.20e1",
            "pearsons_rho": -0.0538829,
            "m": 0.00085007,
            "b": -1.05619,
            "mean_square_error": 0.0325089
          }, {
            "db": 1,
            "id": "RNCMPT00167",
            "alt": "HNRNPCL1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.538618, 0.180421, 0.0804529, 0.200507], 
              [0.00659496, 0.00636751, 0.00638427, 0.980653], 
              [0.00659496, 0.00636751, 0.00638427, 0.980653], 
              [0.00659496, 0.00636751, 0.00638427, 0.980653], 
              [0.00659496, 0.00636751, 0.00638427, 0.980653], 
              [0.00659496, 0.00636751, 0.00638427, 0.980653], 
              [0.0542865, 0.131548, 0.254841, 0.559326]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.79e-1",
            "evalue": "8.25e1",
            "pearsons_rho": -0.0551692,
            "m": 0.00131191,
            "b": -1.05479,
            "mean_square_error": 0.0325066
          }, {
            "db": 1,
            "id": "RNCMPT00025",
            "alt": "HNRNPC",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.564305, 0.184745, 0.081044, 0.169905], 
              [0.00743277, 0.00720532, 0.00722208, 0.97814], 
              [0.00743277, 0.00720532, 0.00722208, 0.97814], 
              [0.00743277, 0.00720532, 0.00722208, 0.97814], 
              [0.00743277, 0.00720532, 0.00722208, 0.97814], 
              [0.00743277, 0.00720532, 0.00722208, 0.97814], 
              [0.0556069, 0.147873, 0.271962, 0.524558]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.79e-1",
            "evalue": "8.25e1",
            "pearsons_rho": -0.0551692,
            "m": 0.00131191,
            "b": -1.05479,
            "mean_square_error": 0.0325066
          }, {
            "db": 1,
            "id": "RNCMPT00274",
            "alt": "HuR",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0111512, 0.114188, 0.107857, 0.766804], 
              [0.0111512, 0.0109237, 0.0109405, 0.966986], 
              [0.108015, 0.0109237, 0.0109405, 0.870121], 
              [0.0969422, 0.0109237, 0.0970838, 0.795051], 
              [0.0111512, 0.0109237, 0.0956728, 0.882252], 
              [0.0111512, 0.101175, 0.0956728, 0.792001], 
              [0.0111512, 0.104032, 0.195535, 0.689282]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.91e-1",
            "evalue": "8.39e1",
            "pearsons_rho": -0.0583301,
            "m": 0.000650779,
            "b": -1.05356,
            "mean_square_error": 0.0325007
          }, {
            "db": 1,
            "id": "RNCMPT00159",
            "alt": "RALY",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.121134, 0.0312959, 0.0313126, 0.816257], 
              [0.00732729, 0.00709984, 0.00711661, 0.978457], 
              [0.00732729, 0.00709984, 0.00711661, 0.978457], 
              [0.00732729, 0.00709984, 0.00711661, 0.978457], 
              [0.00732729, 0.00709984, 0.00711661, 0.978457], 
              [0.00732729, 0.0939486, 0.0957843, 0.802941], 
              [0.0793761, 0.17427, 0.550007, 0.196346]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.93e-1",
            "evalue": "8.41e1",
            "pearsons_rho": -0.0587878,
            "m": 0.00121708,
            "b": -1.05384,
            "mean_square_error": 0.0324999
          }, {
            "db": 1,
            "id": "RNCMPT00155",
            "alt": "PABPC1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.755122, 0.0814869, 0.0815037, 0.0818861], 
              [0.325839, 0.123513, 0.516599, 0.0340493], 
              [0.97089, 0.00956452, 0.00958128, 0.00996372], 
              [0.97089, 0.00956452, 0.00958128, 0.00996372], 
              [0.97089, 0.00956452, 0.00958128, 0.00996372], 
              [0.770322, 0.12043, 0.0992838, 0.00996372], 
              [0.68518, 0.203643, 0.00958128, 0.101596]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.95e-1",
            "evalue": "8.43e1",
            "pearsons_rho": -0.0592268,
            "m": 0.000877367,
            "b": -1.05331,
            "mean_square_error": 0.032499
          }, {
            "db": 1,
            "id": "RNCMPT00064",
            "alt": "SART3",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.754564, 0.081673, 0.0816897, 0.0820722], 
              [0.326766, 0.123615, 0.515356, 0.0342612], 
              [0.970262, 0.00977447, 0.00979123, 0.0101737], 
              [0.970262, 0.00977447, 0.00979123, 0.0101737], 
              [0.970262, 0.00977447, 0.00979123, 0.0101737], 
              [0.769589, 0.121097, 0.0991405, 0.0101737], 
              [0.685295, 0.204005, 0.00979123, 0.100908]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.97e-1",
            "evalue": "8.45e1",
            "pearsons_rho": -0.0597332,
            "m": 0.000880888,
            "b": -1.05307,
            "mean_square_error": 0.032498
          }, {
            "db": 1,
            "id": "RNCMPT00112",
            "alt": "HuR",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0593442, 0.0591168, 0.0591335, 0.822406], 
              [0.0593442, 0.0591168, 0.0591335, 0.822406], 
              [0.228538, 0.0120929, 0.0990957, 0.660272], 
              [0.117296, 0.0120929, 0.766338, 0.104272], 
              [0.0123203, 0.0120929, 0.196098, 0.779489], 
              [0.0123203, 0.0120929, 0.0121096, 0.963477], 
              [0.0588974, 0.05867, 0.0586868, 0.823746]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "7.99e-1",
            "evalue": "8.47e1",
            "pearsons_rho": -0.0602675,
            "m": 0.00073022,
            "b": -1.0521,
            "mean_square_error": 0.032497
          }, {
            "db": 1,
            "id": "RNCMPT00158",
            "alt": "CPEB4",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.170259, 0.451301, 0.16479, 0.21365], 
              [0.0977581, 0.0926133, 0.00764894, 0.80198], 
              [0.00785963, 0.00763218, 0.00764894, 0.976858], 
              [0.00785963, 0.00763218, 0.00764894, 0.976858], 
              [0.00785963, 0.00763218, 0.00764894, 0.976858], 
              [0.00785963, 0.00763218, 0.0912151, 0.893293], 
              [0.0316945, 0.031467, 0.0314838, 0.905356]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.19e-1",
            "evalue": "8.68e1",
            "pearsons_rho": -0.0654487,
            "m": 0.000892054,
            "b": -1.05138,
            "mean_square_error": 0.0324864
          }, {
            "db": 1,
            "id": "RNCMPT00186",
            "alt": "PCBP1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.061785, 0.814684, 0.0615743, 0.0619568], 
              [0.102806, 0.865424, 0.0156937, 0.0160762], 
              [0.29089, 0.110541, 0.0156937, 0.582875], 
              [0.373196, 0.015677, 0.0156937, 0.595432], 
              [0.384195, 0.191697, 0.0156937, 0.408415], 
              [0.0159044, 0.952326, 0.0156937, 0.0160762], 
              [0.0159044, 0.952326, 0.0156937, 0.0160762]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.33e-1",
            "evalue": "8.83e1",
            "pearsons_rho": -0.0695357,
            "m": 0.00146855,
            "b": -1.04625,
            "mean_square_error": 0.0324774
          }, {
            "db": 1,
            "id": "RNCMPT00117",
            "alt": "HuR",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0343492, 0.0341217, 0.0341385, 0.897392], 
              [0.0343492, 0.0341217, 0.0341385, 0.897392], 
              [0.140718, 0.0103018, 0.0977096, 0.75127], 
              [0.190276, 0.0997367, 0.61108, 0.098908], 
              [0.0105293, 0.0103018, 0.29253, 0.686639], 
              [0.0105293, 0.0985088, 0.0103186, 0.880644], 
              [0.058387, 0.0581595, 0.146307, 0.737147]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.40e-1",
            "evalue": "8.90e1",
            "pearsons_rho": -0.0714163,
            "m": 0.000751885,
            "b": -1.04753,
            "mean_square_error": 0.032473
          }, {
            "db": 0,
            "id": "RNCMPT00239",
            "alt": "PCBP1",
            "len": 8,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.130592, 0.608264, 0.130381, 0.130764], 
              [0.162674, 0.712773, 0.0620858, 0.0624682], 
              [0.0993581, 0.198416, 0.0161435, 0.686082], 
              [0.0993581, 0.0161267, 0.0161435, 0.868373], 
              [0.126829, 0.0392163, 0.039233, 0.794721], 
              [0.0628069, 0.811617, 0.0625962, 0.0629786], 
              [0.0628069, 0.811617, 0.0625962, 0.0629786], 
              [0.135937, 0.506331, 0.135727, 0.222005]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.43e-1",
            "evalue": "8.94e1",
            "pearsons_rho": -0.0724298,
            "m": 0.00135975,
            "b": -1.04346,
            "mean_square_error": 0.0324707
          }, {
            "db": 1,
            "id": "RNCMPT00086",
            "alt": "ZC3H14",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0187422, 0.0185148, 0.102445, 0.860298], 
              [0.0187422, 0.0185148, 0.0185315, 0.944211], 
              [0.111791, 0.0185148, 0.116004, 0.753689], 
              [0.202629, 0.115507, 0.476179, 0.205685], 
              [0.0187422, 0.0185148, 0.190205, 0.772538], 
              [0.0187422, 0.0185148, 0.112254, 0.850489], 
              [0.0416964, 0.041469, 0.0414858, 0.875348]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.47e-1",
            "evalue": "8.97e1",
            "pearsons_rho": -0.0734497,
            "m": 0.000685119,
            "b": -1.04722,
            "mean_square_error": 0.0324682
          }, {
            "db": 1,
            "id": "RNCMPT00163",
            "alt": "SRSF1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0848188, 0.0845914, 0.745599, 0.0849906], 
              [0.0331522, 0.0329247, 0.900598, 0.0333239], 
              [0.680806, 0.00961029, 0.299575, 0.0100095], 
              [0.102224, 0.00961029, 0.878156, 0.0100095], 
              [0.080593, 0.0799267, 0.829472, 0.0100095], 
              [0.829543, 0.103398, 0.0333375, 0.0337199], 
              [0.201291, 0.221242, 0.519184, 0.0582831]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.50e-1",
            "evalue": "9.01e1",
            "pearsons_rho": -0.0744529,
            "m": 0.00136991,
            "b": -1.04052,
            "mean_square_error": 0.0324658
          }, {
            "db": 1,
            "id": "RNCMPT00004",
            "alt": "BRUNOL4",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0859502, 0.0857227, 0.176176, 0.65215], 
              [0.0142915, 0.014064, 0.773813, 0.19783], 
              [0.0142915, 0.014064, 0.0140808, 0.957563], 
              [0.0142915, 0.014064, 0.761872, 0.209772], 
              [0.0142915, 0.014064, 0.105213, 0.86643], 
              [0.0142915, 0.014064, 0.664581, 0.307062], 
              [0.0839979, 0.0837705, 0.264332, 0.5679]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.55e-1",
            "evalue": "9.06e1",
            "pearsons_rho": -0.0759549,
            "m": 0.000897706,
            "b": -1.04435,
            "mean_square_error": 0.0324621
          }, {
            "db": 1,
            "id": "RNCMPT00162",
            "alt": "LIN28A",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.112863, 0.574068, 0.112652, 0.200417], 
              [0.062792, 0.146541, 0.727703, 0.0629637], 
              [0.0958616, 0.100403, 0.787265, 0.0164712], 
              [0.867038, 0.100403, 0.0160887, 0.0164712], 
              [0.0162994, 0.100403, 0.866827, 0.0164712], 
              [0.0162994, 0.100403, 0.866827, 0.0164712], 
              [0.443536, 0.144908, 0.349282, 0.0622742]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.57e-1",
            "evalue": "9.09e1",
            "pearsons_rho": -0.0767092,
            "m": 0.00169409,
            "b": -1.03852,
            "mean_square_error": 0.0324602
          }, {
            "db": 1,
            "id": "RNCMPT00136",
            "alt": "HuR",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.211018, 0.0335148, 0.0335315, 0.721935], 
              [0.00983874, 0.00961129, 0.00962805, 0.970922], 
              [0.133595, 0.00961129, 0.582335, 0.274459], 
              [0.293512, 0.00961129, 0.594427, 0.10245], 
              [0.00983874, 0.00961129, 0.101589, 0.878961], 
              [0.00983874, 0.00961129, 0.00962805, 0.970922], 
              [0.0337621, 0.0335347, 0.0335514, 0.899151]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.66e-1",
            "evalue": "9.18e1",
            "pearsons_rho": -0.0795945,
            "m": 0.0011504,
            "b": -1.0447,
            "mean_square_error": 0.0324529
          }, {
            "db": 1,
            "id": "RNCMPT00166",
            "alt": "BRUNOL5",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0848566, 0.0846292, 0.084646, 0.745869], 
              [0.011383, 0.0111556, 0.876942, 0.100519], 
              [0.011383, 0.0111556, 0.0111723, 0.966289], 
              [0.011383, 0.0111556, 0.769867, 0.207595], 
              [0.011383, 0.0111556, 0.0967315, 0.88073], 
              [0.011383, 0.0111556, 0.731751, 0.24571], 
              [0.058193, 0.0579655, 0.2288, 0.65504]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.74e-1",
            "evalue": "9.26e1",
            "pearsons_rho": -0.0822811,
            "m": 0.00123244,
            "b": -1.04241,
            "mean_square_error": 0.0324458
          }, {
            "db": 1,
            "id": "RNCMPT00109",
            "alt": "SRSF1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0333552, 0.0331277, 0.899991, 0.0335269], 
              [0.00990441, 0.00967696, 0.970342, 0.0100762], 
              [0.970553, 0.00967696, 0.00969372, 0.0100762], 
              [0.00990441, 0.193169, 0.686743, 0.110183], 
              [0.00990441, 0.00967696, 0.970342, 0.0100762], 
              [0.589534, 0.00967696, 0.390714, 0.0100762], 
              [0.365305, 0.271066, 0.282268, 0.0813607]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.76e-1",
            "evalue": "9.29e1",
            "pearsons_rho": -0.0829513,
            "m": 0.00284449,
            "b": -1.03681,
            "mean_square_error": 0.032444
          }, {
            "db": 1,
            "id": "RNCMPT00034",
            "alt": "KHDRBS3",
            "len": 8,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.212458, 0.129723, 0.527698, 0.130122], 
              [0.966252, 0.0111108, 0.0111276, 0.01151], 
              [0.0113382, 0.0111108, 0.0111276, 0.966423], 
              [0.966252, 0.0111108, 0.0111276, 0.01151], 
              [0.966252, 0.0111108, 0.0111276, 0.01151], 
              [0.966252, 0.0111108, 0.0111276, 0.01151], 
              [0.211479, 0.52856, 0.223987, 0.0359747], 
              [0.233874, 0.314169, 0.320321, 0.131635]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.83e-1",
            "evalue": "9.36e1",
            "pearsons_rho": -0.0855521,
            "m": 0.00395448,
            "b": -1.04136,
            "mean_square_error": 0.0324368
          }, {
            "db": 1,
            "id": "RNCMPT00012",
            "alt": "CPEB2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0340258, 0.793213, 0.0338151, 0.138946], 
              [0.277302, 0.218124, 0.0100002, 0.494574], 
              [0.105055, 0.0967745, 0.0100002, 0.78817], 
              [0.0102109, 0.00998343, 0.0100002, 0.969805], 
              [0.0102109, 0.00998343, 0.0100002, 0.969805], 
              [0.0102109, 0.00998343, 0.0100002, 0.969805], 
              [0.0338576, 0.0336302, 0.214408, 0.718104]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.88e-1",
            "evalue": "9.42e1",
            "pearsons_rho": -0.087461,
            "m": 0.0011825,
            "b": -1.04054,
            "mean_square_error": 0.0324314
          }, {
            "db": 1,
            "id": "RNCMPT00153",
            "alt": "PABPC3",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.264218, 0.0834899, 0.568403, 0.0838891], 
              [0.961469, 0.0127048, 0.0127216, 0.013104], 
              [0.961469, 0.0127048, 0.0127216, 0.013104], 
              [0.961469, 0.0127048, 0.0127216, 0.013104], 
              [0.961469, 0.0127048, 0.0127216, 0.013104], 
              [0.0960367, 0.797706, 0.0127216, 0.0935349], 
              [0.329244, 0.547982, 0.0611962, 0.0615786]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.89e-1",
            "evalue": "9.43e1",
            "pearsons_rho": -0.087823,
            "m": 0.00216127,
            "b": -1.04034,
            "mean_square_error": 0.0324304
          }, {
            "db": 1,
            "id": "RNCMPT00108",
            "alt": "SRSF1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.033984, 0.0337566, 0.898104, 0.0341558], 
              [0.0105233, 0.0102959, 0.968486, 0.0106951], 
              [0.968696, 0.0102959, 0.0103126, 0.0106951], 
              [0.0105233, 0.197345, 0.683991, 0.108141], 
              [0.0105233, 0.0102959, 0.968486, 0.0106951], 
              [0.588886, 0.0102959, 0.390123, 0.0106951], 
              [0.360424, 0.190704, 0.367211, 0.0816612]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.91e-1",
            "evalue": "9.44e1",
            "pearsons_rho": -0.0884064,
            "m": 0.0030222,
            "b": -1.03392,
            "mean_square_error": 0.0324287
          }, {
            "db": 1,
            "id": "RNCMPT00169",
            "alt": "KHDRBS1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.796002, 0.034869, 0.0348858, 0.134243], 
              [0.0110079, 0.0107804, 0.0107972, 0.967414], 
              [0.881129, 0.0107804, 0.0107972, 0.0972931], 
              [0.967243, 0.0107804, 0.0107972, 0.0111796], 
              [0.87539, 0.0107804, 0.0107972, 0.103031], 
              [0.896275, 0.0344362, 0.0344529, 0.0348354], 
              [0.286165, 0.19598, 0.412314, 0.105541]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "8.99e-1",
            "evalue": "9.53e1",
            "pearsons_rho": -0.0914622,
            "m": 0.00211243,
            "b": -1.04102,
            "mean_square_error": 0.0324197
          }, {
            "db": 1,
            "id": "RNCMPT00079",
            "alt": "U2AF2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.135294, 0.13079, 0.0542479, 0.679667], 
              [0.00704073, 0.00681328, 0.00683004, 0.979315], 
              [0.00704073, 0.00681328, 0.00683004, 0.979315], 
              [0.00704073, 0.00681328, 0.00683004, 0.979315], 
              [0.00704073, 0.0986362, 0.00683004, 0.887492], 
              [0.00704073, 0.327244, 0.00683004, 0.658884], 
              [0.0540149, 0.715012, 0.0538042, 0.177168]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.00e-1",
            "evalue": "9.54e1",
            "pearsons_rho": -0.0920392,
            "m": 0.00162693,
            "b": -1.04026,
            "mean_square_error": 0.032418
          }, {
            "db": 1,
            "id": "RNCMPT00107",
            "alt": "SRSF1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0109173, 0.0106899, 0.967303, 0.0110891], 
              [0.186537, 0.0106899, 0.791683, 0.0110891], 
              [0.967513, 0.0106899, 0.0107067, 0.0110891], 
              [0.0109173, 0.107113, 0.760693, 0.121277], 
              [0.0109173, 0.0106899, 0.967303, 0.0110891], 
              [0.760903, 0.0106899, 0.217317, 0.0110891], 
              [0.312737, 0.307748, 0.192921, 0.186594]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.02e-1",
            "evalue": "9.56e1",
            "pearsons_rho": -0.092757,
            "m": 0.00330423,
            "b": -1.03091,
            "mean_square_error": 0.0324158
          }, {
            "db": 1,
            "id": "RNCMPT00185",
            "alt": "KHDRBS2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.356839, 0.083078, 0.476606, 0.0834771], 
              [0.865558, 0.108716, 0.0126718, 0.0130543], 
              [0.0128825, 0.102734, 0.0126718, 0.871712], 
              [0.781462, 0.0126551, 0.0126718, 0.193211], 
              [0.961619, 0.0126551, 0.0126718, 0.0130543], 
              [0.871926, 0.0126551, 0.0126718, 0.102747], 
              [0.515618, 0.318305, 0.0371216, 0.128956]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.02e-1",
            "evalue": "9.56e1",
            "pearsons_rho": -0.0928839,
            "m": 0.0020203,
            "b": -1.039,
            "mean_square_error": 0.0324154
          }, {
            "db": 1,
            "id": "RNCMPT00173",
            "alt": "RBMS3",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.174305, 0.173931, 0.0841126, 0.56765], 
              [0.890867, 0.0362391, 0.0362559, 0.0366383], 
              [0.0124019, 0.0121745, 0.0121912, 0.963233], 
              [0.963062, 0.0121745, 0.0121912, 0.0125737], 
              [0.108477, 0.0121745, 0.0121912, 0.867158], 
              [0.796217, 0.0358481, 0.0358649, 0.132069], 
              [0.149067, 0.246771, 0.365259, 0.238903]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.06e-1",
            "evalue": "9.61e1",
            "pearsons_rho": -0.094627,
            "m": 0.00199421,
            "b": -1.03765,
            "mean_square_error": 0.0324101
          }, {
            "db": 1,
            "id": "RNCMPT00154",
            "alt": "RBM5",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0644427, 0.26979, 0.601153, 0.0646145], 
              [0.875505, 0.0413595, 0.0413763, 0.0417587], 
              [0.666109, 0.107491, 0.119342, 0.107056], 
              [0.0185701, 0.196805, 0.765883, 0.0187418], 
              [0.0185701, 0.107657, 0.85503, 0.0187418], 
              [0.698111, 0.0413396, 0.21881, 0.0417388], 
              [0.152297, 0.0639954, 0.719314, 0.0643946]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.07e-1",
            "evalue": "9.61e1",
            "pearsons_rho": -0.0949252,
            "m": 0.00177095,
            "b": -1.02887,
            "mean_square_error": 0.0324092
          }, {
            "db": 0,
            "id": "RNCMPT00051",
            "alt": "RBM38",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0792129, 0.0789854, 0.57356, 0.268242], 
              [0.00701386, 0.00678641, 0.280934, 0.705265], 
              [0.00701386, 0.00678641, 0.979013, 0.00718561], 
              [0.00701386, 0.00678641, 0.0903365, 0.895863], 
              [0.00701386, 0.00678641, 0.979013, 0.00718561], 
              [0.00701386, 0.106955, 0.170268, 0.715764], 
              [0.0535333, 0.0533058, 0.666293, 0.226868]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.09e-1",
            "evalue": "9.64e1",
            "pearsons_rho": -0.0958352,
            "m": 0.00207257,
            "b": -1.03576,
            "mean_square_error": 0.0324063
          }, {
            "db": 1,
            "id": "RNCMPT00001",
            "alt": "A1CF",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.394673, 0.106072, 0.106089, 0.393167], 
              [0.00897705, 0.0087496, 0.00876636, 0.973508], 
              [0.973336, 0.0087496, 0.00876636, 0.00914879], 
              [0.973336, 0.0087496, 0.00876636, 0.00914879], 
              [0.00897705, 0.0087496, 0.00876636, 0.973508], 
              [0.00897705, 0.0087496, 0.00876636, 0.973508], 
              [0.320844, 0.144171, 0.477091, 0.0578941]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.10e-1",
            "evalue": "9.65e1",
            "pearsons_rho": -0.0963158,
            "m": 0.00386756,
            "b": -1.03976,
            "mean_square_error": 0.0324048
          }, {
            "db": 1,
            "id": "RNCMPT00269",
            "alt": "PTBP1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.481844, 0.211649, 0.108635, 0.197872], 
              [0.0168377, 0.66816, 0.0166271, 0.298376], 
              [0.0168377, 0.0166103, 0.0166271, 0.949925], 
              [0.107018, 0.0166103, 0.0166271, 0.859745], 
              [0.107018, 0.0166103, 0.0166271, 0.859745], 
              [0.0399472, 0.688819, 0.0397365, 0.231497], 
              [0.0635014, 0.162492, 0.0632907, 0.710716]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.13e-1",
            "evalue": "9.68e1",
            "pearsons_rho": -0.0977537,
            "m": 0.00111093,
            "b": -1.03415,
            "mean_square_error": 0.0324002
          }, {
            "db": 1,
            "id": "RNCMPT00157",
            "alt": "PABPN1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.97112, 0.0094879, 0.00950467, 0.0098871], 
              [0.282988, 0.0094879, 0.697637, 0.0098871], 
              [0.97112, 0.0094879, 0.00950467, 0.0098871], 
              [0.97112, 0.0094879, 0.00950467, 0.0098871], 
              [0.194105, 0.0961834, 0.699823, 0.0098871], 
              [0.97112, 0.0094879, 0.00950467, 0.0098871], 
              [0.225068, 0.3117, 0.225867, 0.237366]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.15e-1",
            "evalue": "9.69e1",
            "pearsons_rho": -0.0982582,
            "m": 0.00260155,
            "b": -1.03106,
            "mean_square_error": 0.0323986
          }, {
            "db": 1,
            "id": "RNCMPT00177",
            "alt": "SFPQ",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0893163, 0.0890889, 0.641628, 0.179966], 
              [0.043784, 0.0435566, 0.134952, 0.777706], 
              [0.482123, 0.0201566, 0.39004, 0.107681], 
              [0.194159, 0.0201566, 0.678321, 0.107364], 
              [0.122386, 0.0201566, 0.107299, 0.75016], 
              [0.0430338, 0.0428063, 0.590042, 0.324119], 
              [0.0881014, 0.087874, 0.261819, 0.562206]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.16e-1",
            "evalue": "9.71e1",
            "pearsons_rho": -0.0991263,
            "m": 0.000863874,
            "b": -1.03052,
            "mean_square_error": 0.0323958
          }, {
            "db": 1,
            "id": "RNCMPT00268",
            "alt": "PTBP1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.242274, 0.537161, 0.0627962, 0.157771], 
              [0.0397571, 0.21887, 0.0395465, 0.701826], 
              [0.0165353, 0.0163078, 0.0163246, 0.950833], 
              [0.0165353, 0.0163078, 0.0163246, 0.950833], 
              [0.0165353, 0.109712, 0.0163246, 0.857428], 
              [0.0165353, 0.673392, 0.0163246, 0.293748], 
              [0.0630825, 0.158355, 0.0628718, 0.71569]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.22e-1",
            "evalue": "9.78e1",
            "pearsons_rho": -0.101883,
            "m": 0.00107556,
            "b": -1.03338,
            "mean_square_error": 0.0323868
          }, {
            "db": 1,
            "id": "RNCMPT00024",
            "alt": "HNRNPA2B1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.293636, 0.107243, 0.29449, 0.304631], 
              [0.0592586, 0.0590312, 0.0590479, 0.822661], 
              [0.964107, 0.0118252, 0.011842, 0.0122244], 
              [0.0120527, 0.0118252, 0.963897, 0.0122244], 
              [0.0120527, 0.0118252, 0.963897, 0.0122244], 
              [0.0120527, 0.0118252, 0.963897, 0.0122244], 
              [0.657351, 0.083059, 0.0830758, 0.176514]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.24e-1",
            "evalue": "9.79e1",
            "pearsons_rho": -0.10271,
            "m": 0.00463456,
            "b": -1.02784,
            "mean_square_error": 0.032384
          }, {
            "db": 1,
            "id": "RNCMPT00187",
            "alt": "BRUNOL6",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0613372, 0.0611098, 0.0611266, 0.816427], 
              [0.0142865, 0.014059, 0.957195, 0.0144582], 
              [0.0142865, 0.014059, 0.0140758, 0.957578], 
              [0.0142865, 0.014059, 0.855221, 0.116432], 
              [0.202846, 0.014059, 0.399736, 0.38336], 
              [0.0142865, 0.014059, 0.476974, 0.494679], 
              [0.0612835, 0.0610561, 0.729206, 0.148454]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.33e-1",
            "evalue": "9.89e1",
            "pearsons_rho": -0.107273,
            "m": 0.00198248,
            "b": -1.02885,
            "mean_square_error": 0.0323683
          }, {
            "db": 1,
            "id": "RNCMPT00026",
            "alt": "HNRNPK",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.110569, 0.852856, 0.0180967, 0.0184791], 
              [0.109027, 0.762136, 0.0180967, 0.110741], 
              [0.853084, 0.0180799, 0.0180967, 0.110741], 
              [0.59185, 0.0180799, 0.103723, 0.286349], 
              [0.312515, 0.501003, 0.0418887, 0.144592], 
              [0.0652656, 0.804242, 0.0650549, 0.0654373], 
              [0.0889223, 0.733272, 0.0887116, 0.0890941]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.33e-1",
            "evalue": "9.89e1",
            "pearsons_rho": -0.107523,
            "m": 0.00204989,
            "b": -1.02551,
            "mean_square_error": 0.0323674
          }, {
            "db": 1,
            "id": "RNCMPT00038",
            "alt": "MBNL1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.315288, 0.190378, 0.212052, 0.282282], 
              [0.00988948, 0.118894, 0.861155, 0.0100612], 
              [0.00988948, 0.861139, 0.0096788, 0.119293], 
              [0.00988948, 0.00966203, 0.0096788, 0.97077], 
              [0.00988948, 0.00966203, 0.0096788, 0.97077], 
              [0.116303, 0.033869, 0.815559, 0.0342682], 
              [0.0574159, 0.746792, 0.138204, 0.0575876]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.35e-1",
            "evalue": "9.91e1",
            "pearsons_rho": -0.108435,
            "m": 0.00337537,
            "b": -1.02832,
            "mean_square_error": 0.0323642
          }, {
            "db": 1,
            "id": "RNCMPT00164",
            "alt": "ZNF638",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0916268, 0.180339, 0.178395, 0.549639], 
              [0.0211353, 0.0209078, 0.851417, 0.10654], 
              [0.0211353, 0.0209078, 0.0209246, 0.937032], 
              [0.0211353, 0.0209078, 0.129889, 0.828069], 
              [0.0211353, 0.570728, 0.386831, 0.021307], 
              [0.0211353, 0.0209078, 0.654169, 0.303788], 
              [0.0670467, 0.0668192, 0.152572, 0.713562]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.35e-1",
            "evalue": "9.91e1",
            "pearsons_rho": -0.108712,
            "m": 0.00181953,
            "b": -1.03019,
            "mean_square_error": 0.0323632
          }, {
            "db": 1,
            "id": "RNCMPT00085",
            "alt": "ZC3H10",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.100031, 0.273193, 0.526574, 0.100203], 
              [0.136302, 0.57097, 0.237848, 0.0548791], 
              [0.818506, 0.119214, 0.0309484, 0.0313309], 
              [0.00715516, 0.00692771, 0.97859, 0.0073269], 
              [0.00715516, 0.978573, 0.00694447, 0.0073269], 
              [0.0318288, 0.0316013, 0.904569, 0.0320005], 
              [0.217115, 0.562602, 0.109949, 0.110332]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.37e-1",
            "evalue": "9.93e1",
            "pearsons_rho": -0.109713,
            "m": 0.00516293,
            "b": -1.02628,
            "mean_square_error": 0.0323596
          }, {
            "db": 1,
            "id": "RNCMPT00088",
            "alt": "SRSF10",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.880829, 0.0984073, 0.0101902, 0.0105727], 
              [0.285215, 0.0101735, 0.694039, 0.0105727], 
              [0.887075, 0.0101735, 0.0921783, 0.0105727], 
              [0.101301, 0.0101735, 0.877952, 0.0105727], 
              [0.69425, 0.0101735, 0.285004, 0.0105727], 
              [0.300391, 0.11684, 0.572195, 0.0105727], 
              [0.326619, 0.145532, 0.493474, 0.0343757]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.39e-1",
            "evalue": "9.95e1",
            "pearsons_rho": -0.110673,
            "m": 0.00166945,
            "b": -1.02341,
            "mean_square_error": 0.0323562
          }, {
            "db": 1,
            "id": "RNCMPT00072",
            "alt": "SRSF2",
            "len": 8,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.605686, 0.131299, 0.131316, 0.131698], 
              [0.0176487, 0.0174212, 0.947111, 0.0178204], 
              [0.0176487, 0.0174212, 0.947111, 0.0178204], 
              [0.856748, 0.0174212, 0.017438, 0.108394], 
              [0.0176487, 0.0174212, 0.947111, 0.0178204], 
              [0.581019, 0.0174212, 0.105769, 0.295791], 
              [0.304642, 0.0174212, 0.388871, 0.289067], 
              [0.320481, 0.136046, 0.407029, 0.136445]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.41e-1",
            "evalue": "9.97e1",
            "pearsons_rho": -0.111939,
            "m": 0.00292584,
            "b": -1.0193,
            "mean_square_error": 0.0323516
          }, {
            "db": 0,
            "id": "RNCMPT00215",
            "alt": "PCBP3",
            "len": 8,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.226431, 0.416089, 0.133068, 0.224412], 
              [0.129712, 0.132657, 0.0397644, 0.697866], 
              [0.206208, 0.0160879, 0.0161047, 0.761601], 
              [0.206208, 0.0160879, 0.0161047, 0.761601], 
              [0.0163154, 0.673814, 0.0161047, 0.293766], 
              [0.0163154, 0.951093, 0.0161047, 0.0164871], 
              [0.0400835, 0.879787, 0.0398728, 0.0402553], 
              [0.133103, 0.132876, 0.132893, 0.601129]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.42e-1",
            "evalue": "9.98e1",
            "pearsons_rho": -0.112462,
            "m": 0.00234596,
            "b": -1.02459,
            "mean_square_error": 0.0323496
          }, {
            "db": 1,
            "id": "RNCMPT00055",
            "alt": "RBM5",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.130968, 0.0396372, 0.789357, 0.0400364], 
              [0.880672, 0.0396372, 0.0396539, 0.0400364], 
              [0.69926, 0.0396372, 0.221066, 0.0400364], 
              [0.0167691, 0.0165416, 0.949748, 0.0169408], 
              [0.0167691, 0.0165416, 0.949748, 0.0169408], 
              [0.624571, 0.0640481, 0.0640649, 0.247316], 
              [0.576258, 0.110868, 0.201606, 0.111267]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.46e-1",
            "evalue": "1.00e2",
            "pearsons_rho": -0.115263,
            "m": 0.00249371,
            "b": -1.01979,
            "mean_square_error": 0.0323392
          }, {
            "db": 1,
            "id": "RNCMPT00016",
            "alt": "FMR1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.11163, 0.111402, 0.574464, 0.202504], 
              [0.0195522, 0.0193247, 0.941399, 0.0197239], 
              [0.94161, 0.0193247, 0.0193415, 0.0197239], 
              [0.0195522, 0.941382, 0.0193415, 0.0197239], 
              [0.763708, 0.0193247, 0.197243, 0.0197239], 
              [0.498524, 0.0423297, 0.416417, 0.0427289], 
              [0.134386, 0.0423297, 0.780556, 0.0427289]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.46e-1",
            "evalue": "1.00e2",
            "pearsons_rho": -0.115413,
            "m": 0.00359047,
            "b": -1.01654,
            "mean_square_error": 0.0323386
          }, {
            "db": 1,
            "id": "RNCMPT00019",
            "alt": "SRSF10",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.829759, 0.0566083, 0.0566251, 0.0570075], 
              [0.0974527, 0.0101158, 0.881917, 0.010515], 
              [0.969237, 0.0101158, 0.0101325, 0.010515], 
              [0.108645, 0.0101158, 0.870725, 0.010515], 
              [0.877272, 0.102081, 0.0101325, 0.010515], 
              [0.578367, 0.208429, 0.202688, 0.010515], 
              [0.53466, 0.261603, 0.145995, 0.0577408]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.48e-1",
            "evalue": "1.00e2",
            "pearsons_rho": -0.116181,
            "m": 0.00215898,
            "b": -1.01826,
            "mean_square_error": 0.0323357
          }, {
            "db": 1,
            "id": "RNCMPT00089",
            "alt": "SRSF10",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.899587, 0.0333327, 0.0333494, 0.0337319], 
              [0.181012, 0.009069, 0.800451, 0.00946819], 
              [0.972376, 0.009069, 0.00908576, 0.00946819], 
              [0.0975163, 0.009069, 0.883945, 0.00946819], 
              [0.87957, 0.009069, 0.101892, 0.00946819], 
              [0.506262, 0.103404, 0.380866, 0.00946819], 
              [0.513053, 0.141562, 0.312391, 0.0329946]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.48e-1",
            "evalue": "1.00e2",
            "pearsons_rho": -0.116311,
            "m": 0.00211937,
            "b": -1.0201,
            "mean_square_error": 0.0323352
          }, {
            "db": 0,
            "id": "RNCMPT00062",
            "alt": "KHDRBS1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0140775, 0.0138501, 0.0138669, 0.958206], 
              [0.958034, 0.0138501, 0.0138669, 0.0142493], 
              [0.958034, 0.0138501, 0.0138669, 0.0142493], 
              [0.867509, 0.0138501, 0.0138669, 0.104774], 
              [0.888324, 0.0370869, 0.0371037, 0.0374861], 
              [0.14733, 0.152012, 0.639452, 0.0612055], 
              [0.197689, 0.211483, 0.480625, 0.110203]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.48e-1",
            "evalue": "1.00e2",
            "pearsons_rho": -0.116381,
            "m": 0.00315413,
            "b": -1.0295,
            "mean_square_error": 0.0323349
          }, {
            "db": 1,
            "id": "RNCMPT00067",
            "alt": "SRSF9",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.083989, 0.0837615, 0.6593, 0.17295], 
              [0.0384019, 0.0381745, 0.884849, 0.0385737], 
              [0.545164, 0.0150869, 0.424263, 0.0154861], 
              [0.502335, 0.111105, 0.0151037, 0.371456], 
              [0.0153144, 0.0150869, 0.954113, 0.0154861], 
              [0.0924218, 0.242231, 0.649861, 0.0154861], 
              [0.688485, 0.23231, 0.0394121, 0.0397946]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.48e-1",
            "evalue": "1.01e2",
            "pearsons_rho": -0.116745,
            "m": 0.0021964,
            "b": -1.01655,
            "mean_square_error": 0.0323335
          }, {
            "db": 1,
            "id": "RNCMPT00106",
            "alt": "SRSF1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0341522, 0.0339247, 0.897598, 0.0343239], 
              [0.263593, 0.0104869, 0.715033, 0.0108861], 
              [0.887451, 0.0104869, 0.0105037, 0.0915577], 
              [0.0107144, 0.104534, 0.760224, 0.124528], 
              [0.0107144, 0.0104869, 0.967912, 0.0108861], 
              [0.760435, 0.0104869, 0.218191, 0.0108861], 
              [0.293115, 0.372847, 0.172542, 0.161495]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.49e-1",
            "evalue": "1.01e2",
            "pearsons_rho": -0.117487,
            "m": 0.00335621,
            "b": -1.01646,
            "mean_square_error": 0.0323307
          }, {
            "db": 1,
            "id": "RNCMPT00022",
            "alt": "HNRNPA1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.268197, 0.0836013, 0.271884, 0.376317], 
              [0.0108706, 0.0106431, 0.0106599, 0.967825], 
              [0.967654, 0.0106431, 0.0106599, 0.0110423], 
              [0.0108706, 0.0106431, 0.967443, 0.0110423], 
              [0.0108706, 0.0106431, 0.967443, 0.0110423], 
              [0.0108706, 0.0106431, 0.874417, 0.104068], 
              [0.589462, 0.104523, 0.104539, 0.201476]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.51e-1",
            "evalue": "1.01e2",
            "pearsons_rho": -0.118506,
            "m": 0.00578999,
            "b": -1.01918,
            "mean_square_error": 0.0323268
          }, {
            "db": 1,
            "id": "RNCMPT00023",
            "alt": "HNRNPA1L2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.178395, 0.0867854, 0.267032, 0.467788], 
              [0.0143452, 0.0141178, 0.0141345, 0.957404], 
              [0.957232, 0.0141178, 0.0141345, 0.014517], 
              [0.0143452, 0.0141178, 0.957021, 0.014517], 
              [0.0143452, 0.0141178, 0.957021, 0.014517], 
              [0.0375999, 0.0373725, 0.796179, 0.128847], 
              [0.452041, 0.0834501, 0.0834669, 0.381042]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.51e-1",
            "evalue": "1.01e2",
            "pearsons_rho": -0.118779,
            "m": 0.00481481,
            "b": -1.01856,
            "mean_square_error": 0.0323257
          }, {
            "db": 1,
            "id": "RNCMPT00171",
            "alt": "PABPC5",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.752498, 0.0823615, 0.0823783, 0.0827607], 
              [0.201754, 0.0374949, 0.722857, 0.0378941], 
              [0.956561, 0.0143406, 0.0143574, 0.0147398], 
              [0.956561, 0.0143406, 0.0143574, 0.0147398], 
              [0.956561, 0.0143406, 0.0143574, 0.0147398], 
              [0.46298, 0.112202, 0.19639, 0.228426], 
              [0.11182, 0.180678, 0.0143574, 0.693145]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.52e-1",
            "evalue": "1.01e2",
            "pearsons_rho": -0.119494,
            "m": 0.00183168,
            "b": -1.02615,
            "mean_square_error": 0.0323229
          }, {
            "db": 1,
            "id": "RNCMPT00020",
            "alt": "FXR2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.178618, 0.0879237, 0.55827, 0.17519], 
              [0.0199263, 0.0196988, 0.940276, 0.020098], 
              [0.940486, 0.0196988, 0.0197156, 0.020098], 
              [0.0199263, 0.940259, 0.0197156, 0.020098], 
              [0.3894, 0.0196988, 0.570803, 0.020098], 
              [0.466639, 0.0196988, 0.493564, 0.020098], 
              [0.236302, 0.0426819, 0.677935, 0.0430811]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.54e-1",
            "evalue": "1.01e2",
            "pearsons_rho": -0.120604,
            "m": 0.00374666,
            "b": -1.01349,
            "mean_square_error": 0.0323185
          }, {
            "db": 1,
            "id": "RNCMPT00161",
            "alt": "FXR1",
            "len": 8,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.592969, 0.135539, 0.135555, 0.135938], 
              [0.880384, 0.0397337, 0.0397504, 0.0401329], 
              [0.0168367, 0.377056, 0.0166261, 0.589481], 
              [0.0168367, 0.0166093, 0.949544, 0.0170085], 
              [0.859022, 0.0166093, 0.107359, 0.0170085], 
              [0.10757, 0.858794, 0.0166261, 0.0170085], 
              [0.585298, 0.0166093, 0.381083, 0.0170085], 
              [0.606618, 0.130988, 0.131005, 0.131388]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.55e-1",
            "evalue": "1.01e2",
            "pearsons_rho": -0.121136,
            "m": 0.00563689,
            "b": -1.02129,
            "mean_square_error": 0.0323164
          }, {
            "db": 1,
            "id": "RNCMPT00151",
            "alt": "RBM42",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.760027, 0.0798521, 0.0798688, 0.0802513], 
              [0.970516, 0.0096889, 0.00970566, 0.0100881], 
              [0.00991635, 0.970289, 0.00970566, 0.0100881], 
              [0.00991635, 0.0096889, 0.00970566, 0.970688], 
              [0.970516, 0.0096889, 0.00970566, 0.0100881], 
              [0.520705, 0.317484, 0.0335345, 0.128276], 
              [0.105011, 0.104783, 0.685023, 0.105183]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.63e-1",
            "evalue": "1.02e2",
            "pearsons_rho": -0.12762,
            "m": 0.00996644,
            "b": -1.02385,
            "mean_square_error": 0.03229
          }, {
            "db": 1,
            "id": "RNCMPT00050",
            "alt": "RBM3",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.169774, 0.083078, 0.663671, 0.0834771], 
              [0.717324, 0.037669, 0.206939, 0.0380682], 
              [0.184886, 0.0143486, 0.191861, 0.608904], 
              [0.956539, 0.0143486, 0.0143654, 0.0147478], 
              [0.192072, 0.778816, 0.0143654, 0.0147478], 
              [0.125763, 0.0374481, 0.595138, 0.24165], 
              [0.680753, 0.106277, 0.106294, 0.106676]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.64e-1",
            "evalue": "1.02e2",
            "pearsons_rho": -0.129154,
            "m": 0.00261464,
            "b": -1.0139,
            "mean_square_error": 0.0322835
          }, {
            "db": 1,
            "id": "RNCMPT00036",
            "alt": "LIN28A",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.303804, 0.386554, 0.109914, 0.199729], 
              [0.040381, 0.0401536, 0.878912, 0.0405528], 
              [0.040381, 0.0401536, 0.878912, 0.0405528], 
              [0.879122, 0.0401536, 0.0401703, 0.0405528], 
              [0.0167492, 0.0165217, 0.949808, 0.0169209], 
              [0.693304, 0.0401536, 0.0401703, 0.226372], 
              [0.717631, 0.0633247, 0.15532, 0.0637239]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.66e-1",
            "evalue": "1.02e2",
            "pearsons_rho": -0.130185,
            "m": 0.00333236,
            "b": -1.01084,
            "mean_square_error": 0.0322791
          }, {
            "db": 1,
            "id": "RNCMPT00087",
            "alt": "ZCRB1",
            "len": 8,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.124322, 0.124095, 0.62709, 0.124494], 
              [0.634369, 0.0329197, 0.213945, 0.118767], 
              [0.491839, 0.301671, 0.00909571, 0.197395], 
              [0.0093064, 0.00907895, 0.00909571, 0.97252], 
              [0.0093064, 0.00907895, 0.00909571, 0.97252], 
              [0.972348, 0.00907895, 0.00909571, 0.00947814], 
              [0.972348, 0.00907895, 0.00909571, 0.00947814], 
              [0.13554, 0.135313, 0.413228, 0.315919]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.69e-1",
            "evalue": "1.03e2",
            "pearsons_rho": -0.133327,
            "m": 0.00540735,
            "b": -1.02187,
            "mean_square_error": 0.0322655
          }, {
            "db": 1,
            "id": "RNCMPT00241",
            "alt": "RBM45",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0812288, 0.0810013, 0.756369, 0.0814005], 
              [0.826579, 0.057668, 0.0576848, 0.0580672], 
              [0.123205, 0.718049, 0.124237, 0.034508], 
              [0.010388, 0.0101605, 0.873566, 0.105886], 
              [0.764991, 0.0101605, 0.21429, 0.0105597], 
              [0.239649, 0.69135, 0.0343096, 0.0346921], 
              [0.450375, 0.284757, 0.182809, 0.0820592]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.69e-1",
            "evalue": "1.03e2",
            "pearsons_rho": -0.133343,
            "m": 0.00375311,
            "b": -1.00632,
            "mean_square_error": 0.0322654
          }, {
            "db": 1,
            "id": "RNCMPT00057",
            "alt": "RBMS3",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.581098, 0.202741, 0.107889, 0.108271], 
              [0.0129651, 0.108585, 0.0127544, 0.865694], 
              [0.96137, 0.0127377, 0.0127544, 0.0131369], 
              [0.0129651, 0.0127377, 0.0127544, 0.961542], 
              [0.96137, 0.0127377, 0.0127544, 0.0131369], 
              [0.105143, 0.106106, 0.217839, 0.570913], 
              [0.523484, 0.356191, 0.0599713, 0.0603538]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.71e-1",
            "evalue": "1.03e2",
            "pearsons_rho": -0.135586,
            "m": 0.00353421,
            "b": -1.02524,
            "mean_square_error": 0.0322555
          }, {
            "db": 1,
            "id": "RNCMPT00053",
            "alt": "RBM41",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.445094, 0.057474, 0.0574907, 0.439942], 
              [0.0102825, 0.0100551, 0.0100718, 0.969591], 
              [0.969419, 0.0100551, 0.0100718, 0.0104543], 
              [0.0102825, 0.969191, 0.0100718, 0.0104543], 
              [0.581276, 0.0100551, 0.0100718, 0.398598], 
              [0.0102825, 0.0100551, 0.0100718, 0.969591], 
              [0.105251, 0.105023, 0.194863, 0.594862]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.71e-1",
            "evalue": "1.03e2",
            "pearsons_rho": -0.136098,
            "m": 0.00404659,
            "b": -1.0202,
            "mean_square_error": 0.0322532
          }, {
            "db": 1,
            "id": "RNCMPT00184",
            "alt": "RBM24",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.470767, 0.0819377, 0.0819544, 0.365342], 
              [0.0103283, 0.0101008, 0.969072, 0.0105], 
              [0.292456, 0.0101008, 0.0101176, 0.687326], 
              [0.0103283, 0.0101008, 0.969072, 0.0105], 
              [0.0103283, 0.0101008, 0.0101176, 0.969454], 
              [0.0103283, 0.0101008, 0.969072, 0.0105], 
              [0.534161, 0.0581247, 0.157853, 0.249861]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.77e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.142602,
            "m": 0.00390248,
            "b": -1.01826,
            "mean_square_error": 0.0322234
          }, {
            "db": 1,
            "id": "RNCMPT00084",
            "alt": "YBX2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.87802, 0.0107008, 0.10018, 0.0111], 
              [0.87802, 0.0107008, 0.10018, 0.0111], 
              [0.0109283, 0.877792, 0.10018, 0.0111], 
              [0.967483, 0.0107008, 0.0107176, 0.0111], 
              [0.496403, 0.0107008, 0.10018, 0.392717], 
              [0.0109283, 0.967255, 0.0107176, 0.0111], 
              [0.328642, 0.106056, 0.271289, 0.294013]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.78e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.144262,
            "m": 0.00510877,
            "b": -1.00894,
            "mean_square_error": 0.0322156
          }, {
            "db": 1,
            "id": "RNCMPT00054",
            "alt": "RBM46",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.711459, 0.036462, 0.215218, 0.0368612], 
              [0.962401, 0.0123944, 0.0124111, 0.0127936], 
              [0.0126218, 0.0123944, 0.0124111, 0.962573], 
              [0.0126218, 0.596002, 0.378582, 0.0127936], 
              [0.962401, 0.0123944, 0.0124111, 0.0127936], 
              [0.601393, 0.0368093, 0.131652, 0.230147], 
              [0.301374, 0.109765, 0.197141, 0.39172]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.79e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.145698,
            "m": 0.00344938,
            "b": -1.01317,
            "mean_square_error": 0.0322088
          }, {
            "db": 1,
            "id": "RNCMPT00090",
            "alt": "SRSF10",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.904411, 0.0317247, 0.0317415, 0.0321239], 
              [0.00808152, 0.00785407, 0.97581, 0.00825327], 
              [0.976021, 0.00785407, 0.00787084, 0.00825327], 
              [0.0822059, 0.00785407, 0.901687, 0.00825327], 
              [0.976021, 0.00785407, 0.00787084, 0.00825327], 
              [0.52997, 0.284669, 0.177108, 0.00825327], 
              [0.519476, 0.219648, 0.173454, 0.0874214]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.79e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.1459,
            "m": 0.0042295,
            "b": -1.00499,
            "mean_square_error": 0.0322078
          }, {
            "db": 1,
            "id": "RNCMPT00076",
            "alt": "TARDBP",
            "len": 8,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.3234, 0.130039, 0.222712, 0.323848], 
              [0.0115372, 0.0113098, 0.965443, 0.011709], 
              [0.779952, 0.0113098, 0.108711, 0.100026], 
              [0.965654, 0.0113098, 0.0113266, 0.011709], 
              [0.0115372, 0.0113098, 0.0113266, 0.965825], 
              [0.0115372, 0.0113098, 0.965443, 0.011709], 
              [0.635455, 0.0592143, 0.148864, 0.156468], 
              [0.310411, 0.131114, 0.333584, 0.22489]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.80e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.146344,
            "m": 0.00575084,
            "b": -1.00927,
            "mean_square_error": 0.0322057
          }, {
            "db": 1,
            "id": "RNCMPT00172",
            "alt": "IGF2BP3",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.711152, 0.161393, 0.0635355, 0.0639179], 
              [0.395634, 0.486398, 0.0166101, 0.101359], 
              [0.782221, 0.0998103, 0.0166101, 0.101359], 
              [0.386886, 0.29761, 0.0166101, 0.298893], 
              [0.574437, 0.11006, 0.0166101, 0.298893], 
              [0.199688, 0.766708, 0.0166101, 0.0169926], 
              [0.860401, 0.105995, 0.0166101, 0.0169926]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.81e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.147877,
            "m": 0.0013979,
            "b": -1.01039,
            "mean_square_error": 0.0321982
          }, {
            "db": 1,
            "id": "RNCMPT00168",
            "alt": "RBFOX1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.311855, 0.0525735, 0.0525902, 0.582983], 
              [0.00550739, 0.00527994, 0.983533, 0.00567914], 
              [0.00550739, 0.983516, 0.00529671, 0.00567914], 
              [0.983743, 0.00527994, 0.00529671, 0.00567914], 
              [0.00550739, 0.00527994, 0.00529671, 0.983915], 
              [0.00550739, 0.00527994, 0.983533, 0.00567914], 
              [0.313438, 0.480133, 0.103024, 0.103407]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.83e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.150749,
            "m": 0.0110291,
            "b": -1.01104,
            "mean_square_error": 0.0321841
          }, {
            "db": 1,
            "id": "RNCMPT00170",
            "alt": "RBM6",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.456721, 0.278184, 0.0870111, 0.178084], 
              [0.857493, 0.0169894, 0.0170062, 0.108511], 
              [0.0172168, 0.0169894, 0.0170062, 0.948789], 
              [0.0172168, 0.948389, 0.0170062, 0.0173886], 
              [0.0172168, 0.948389, 0.0170062, 0.0173886], 
              [0.948617, 0.0169894, 0.0170062, 0.0173886], 
              [0.278141, 0.0868003, 0.547858, 0.0871995]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.83e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.150851,
            "m": 0.00529418,
            "b": -1.00561,
            "mean_square_error": 0.0321836
          }, {
            "db": 1,
            "id": "RNCMPT00176",
            "alt": "MSI1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0356984, 0.035471, 0.0354878, 0.893343], 
              [0.963421, 0.0120541, 0.0120708, 0.0124533], 
              [0.0122815, 0.0120541, 0.96321, 0.0124533], 
              [0.111468, 0.101254, 0.202207, 0.585071], 
              [0.361031, 0.0120541, 0.101572, 0.525343], 
              [0.562091, 0.0585217, 0.320466, 0.0589209], 
              [0.103281, 0.103053, 0.690214, 0.103452]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.83e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.151121,
            "m": 0.00344818,
            "b": -1.00571,
            "mean_square_error": 0.0321823
          }, {
            "db": 1,
            "id": "RNCMPT00070",
            "alt": "SNRNP70",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.28932, 0.105842, 0.498597, 0.106241], 
              [0.614427, 0.0361098, 0.0361266, 0.313338], 
              [0.0129353, 0.0127078, 0.0127246, 0.961632], 
              [0.0129353, 0.961233, 0.0127246, 0.013107], 
              [0.961461, 0.0127078, 0.0127246, 0.013107], 
              [0.961461, 0.0127078, 0.0127246, 0.013107], 
              [0.129803, 0.0367804, 0.796238, 0.0371796]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.83e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.151848,
            "m": 0.00715799,
            "b": -1.00737,
            "mean_square_error": 0.0321786
          }, {
            "db": 1,
            "id": "RNCMPT00074",
            "alt": "SRSF9",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.881581, 0.0393347, 0.0393514, 0.0397339], 
              [0.1014, 0.0160819, 0.679802, 0.202716], 
              [0.102501, 0.0160819, 0.864935, 0.0164811], 
              [0.780056, 0.0160819, 0.101189, 0.102673], 
              [0.195539, 0.207589, 0.495064, 0.101807], 
              [0.294008, 0.673411, 0.0160987, 0.0164811], 
              [0.643565, 0.0868083, 0.18242, 0.0872075]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.84e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.153318,
            "m": 0.00240907,
            "b": -0.998815,
            "mean_square_error": 0.0321713
          }, {
            "db": 1,
            "id": "RNCMPT00033",
            "alt": "IGF2BP2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.563047, 0.176815, 0.172223, 0.087915], 
              [0.399104, 0.519191, 0.0406609, 0.0410433], 
              [0.947612, 0.0173237, 0.0173405, 0.0177229], 
              [0.505814, 0.199301, 0.0173405, 0.277544], 
              [0.595274, 0.109841, 0.0173405, 0.277544], 
              [0.110069, 0.854867, 0.0173405, 0.0177229], 
              [0.947612, 0.0173237, 0.0173405, 0.0177229]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.84e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.153933,
            "m": 0.00201564,
            "b": -1.0091,
            "mean_square_error": 0.0321681
          }, {
            "db": 1,
            "id": "RNCMPT00041",
            "alt": "MSI1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0584159, 0.0581884, 0.0582052, 0.825191], 
              [0.96476, 0.0116083, 0.0116251, 0.0120075], 
              [0.0118358, 0.0116083, 0.964549, 0.0120075], 
              [0.124782, 0.104369, 0.116505, 0.654344], 
              [0.435769, 0.0116083, 0.0116251, 0.540999], 
              [0.557757, 0.0350103, 0.371823, 0.0354095], 
              [0.138779, 0.0577416, 0.745338, 0.0581408]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.85e-1",
            "evalue": "1.04e2",
            "pearsons_rho": -0.155316,
            "m": 0.00405686,
            "b": -1.00399,
            "mean_square_error": 0.0321611
          }, {
            "db": 1,
            "id": "RNCMPT00037",
            "alt": "MATR3",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.457411, 0.366453, 0.0878768, 0.0882592], 
              [0.862139, 0.0181894, 0.0182062, 0.101464], 
              [0.101464, 0.0181894, 0.0182062, 0.862141], 
              [0.0184168, 0.944787, 0.0182062, 0.0185886], 
              [0.0184168, 0.0181894, 0.0182062, 0.945187], 
              [0.0184168, 0.0181894, 0.0182062, 0.945187], 
              [0.177602, 0.0885277, 0.644943, 0.0889269]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.87e-1",
            "evalue": "1.05e2",
            "pearsons_rho": -0.159123,
            "m": 0.00474986,
            "b": -1.00607,
            "mean_square_error": 0.0321414
          }, {
            "db": 1,
            "id": "RNCMPT00013",
            "alt": "DAZAP1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0132716, 0.0130441, 0.0130609, 0.960622], 
              [0.960451, 0.0130441, 0.0130609, 0.0134433], 
              [0.102449, 0.0130441, 0.871063, 0.0134433], 
              [0.11826, 0.0130441, 0.668266, 0.20043], 
              [0.202266, 0.0130441, 0.099828, 0.684861], 
              [0.684689, 0.0130441, 0.0985236, 0.203743], 
              [0.3369, 0.0599466, 0.542807, 0.0603458]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.88e-1",
            "evalue": "1.05e2",
            "pearsons_rho": -0.160177,
            "m": 0.00490902,
            "b": -1.00506,
            "mean_square_error": 0.0321358
          }, {
            "db": 1,
            "id": "RNCMPT00049",
            "alt": "RBM28",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0811233, 0.0808959, 0.756685, 0.0812951], 
              [0.511686, 0.036471, 0.0364878, 0.415356], 
              [0.0134497, 0.0132222, 0.959706, 0.0136214], 
              [0.0134497, 0.0132222, 0.013239, 0.960088], 
              [0.959916, 0.0132222, 0.013239, 0.0136214], 
              [0.0134497, 0.0132222, 0.959706, 0.0136214], 
              [0.402411, 0.0853197, 0.163602, 0.348667]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.88e-1",
            "evalue": "1.05e2",
            "pearsons_rho": -0.161348,
            "m": 0.00708875,
            "b": -1.00223,
            "mean_square_error": 0.0321296
          }, {
            "db": 1,
            "id": "RNCMPT00027",
            "alt": "HNRNPL",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.735204, 0.147138, 0.058638, 0.0590204], 
              [0.191791, 0.683744, 0.0121584, 0.112307], 
              [0.96316, 0.0121416, 0.0121584, 0.0125408], 
              [0.111045, 0.669216, 0.0121584, 0.207581], 
              [0.88104, 0.0121416, 0.0121584, 0.0946602], 
              [0.273976, 0.701326, 0.0121584, 0.0125408], 
              [0.714111, 0.213994, 0.0357564, 0.0361388]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.89e-1",
            "evalue": "1.05e2",
            "pearsons_rho": -0.162332,
            "m": 0.00280974,
            "b": -1.00802,
            "mean_square_error": 0.0321243
          }, {
            "db": 1,
            "id": "RNCMPT00110",
            "alt": "SRSF1",
            "len": 8,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.327987, 0.406864, 0.132383, 0.132766], 
              [0.894411, 0.0350581, 0.0350748, 0.0354572], 
              [0.0114178, 0.0111904, 0.965803, 0.0115896], 
              [0.0114178, 0.0111904, 0.965803, 0.0115896], 
              [0.966014, 0.0111904, 0.0112072, 0.0115896], 
              [0.0587163, 0.628157, 0.165909, 0.147218], 
              [0.6479, 0.18837, 0.0816738, 0.0820563], 
              [0.338694, 0.31322, 0.218972, 0.129113]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.89e-1",
            "evalue": "1.05e2",
            "pearsons_rho": -0.16244,
            "m": 0.00531569,
            "b": -0.991301,
            "mean_square_error": 0.0321238
          }, {
            "db": 1,
            "id": "RNCMPT00075",
            "alt": "TUT1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.199743, 0.574023, 0.112926, 0.113308], 
              [0.213311, 0.112909, 0.560472, 0.113308], 
              [0.863618, 0.0453217, 0.0453385, 0.0457209], 
              [0.0229323, 0.0227048, 0.0227216, 0.931642], 
              [0.931471, 0.0227048, 0.0227216, 0.023104], 
              [0.0686139, 0.705322, 0.0684032, 0.15766], 
              [0.0912039, 0.0909765, 0.0909932, 0.726825]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.92e-1",
            "evalue": "1.05e2",
            "pearsons_rho": -0.170836,
            "m": 0.00462227,
            "b": -1.00186,
            "mean_square_error": 0.0320776
          }, {
            "db": 1,
            "id": "RNCMPT00021",
            "alt": "G3BP2",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.739983, 0.0589715, 0.0589882, 0.142057], 
              [0.0356467, 0.0354192, 0.893115, 0.0358184], 
              [0.012187, 0.0119595, 0.963495, 0.0123587], 
              [0.963705, 0.0119595, 0.0119763, 0.0123587], 
              [0.012187, 0.0119595, 0.0119763, 0.963877], 
              [0.310677, 0.0119595, 0.47934, 0.198023], 
              [0.60647, 0.0119595, 0.280796, 0.100774]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.92e-1",
            "evalue": "1.05e2",
            "pearsons_rho": -0.172347,
            "m": 0.00691177,
            "b": -0.994797,
            "mean_square_error": 0.032069
          }, {
            "db": 1,
            "id": "RNCMPT00047",
            "alt": "QKI",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.821559, 0.0593416, 0.0593584, 0.0597408], 
              [0.0159233, 0.95227, 0.0157126, 0.0160951], 
              [0.0159233, 0.0156959, 0.0157126, 0.952669], 
              [0.952497, 0.0156959, 0.0157126, 0.0160951], 
              [0.952497, 0.0156959, 0.0157126, 0.0160951], 
              [0.0159233, 0.817841, 0.0157126, 0.150523], 
              [0.410413, 0.304538, 0.173285, 0.111765]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.95e-1",
            "evalue": "1.05e2",
            "pearsons_rho": -0.182017,
            "m": 0.0122973,
            "b": -0.999002,
            "mean_square_error": 0.0320123
          }, {
            "db": 1,
            "id": "RNCMPT00152",
            "alt": "RBMS1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.108547, 0.10832, 0.193956, 0.589177], 
              [0.959421, 0.0133874, 0.0134042, 0.0137866], 
              [0.0136149, 0.0133874, 0.0134042, 0.959593], 
              [0.959421, 0.0133874, 0.0134042, 0.0137866], 
              [0.104463, 0.110355, 0.127507, 0.657675], 
              [0.754471, 0.218338, 0.0134042, 0.0137866], 
              [0.0136149, 0.587625, 0.384972, 0.0137866]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.96e-1",
            "evalue": "1.06e2",
            "pearsons_rho": -0.186532,
            "m": 0.00757925,
            "b": -0.999842,
            "mean_square_error": 0.0319848
          }, {
            "db": 1,
            "id": "RNCMPT00091",
            "alt": "HNRNPL",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.93788, 0.0205685, 0.0205853, 0.0209677], 
              [0.020796, 0.937652, 0.0205853, 0.0209677], 
              [0.93788, 0.0205685, 0.0205853, 0.0209677], 
              [0.113077, 0.845371, 0.0205853, 0.0209677], 
              [0.476036, 0.112849, 0.390148, 0.0209677], 
              [0.760912, 0.107676, 0.110444, 0.0209677], 
              [0.360936, 0.270621, 0.276926, 0.0915179]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.96e-1",
            "evalue": "1.06e2",
            "pearsons_rho": -0.18717,
            "m": 0.00437054,
            "b": -0.991753,
            "mean_square_error": 0.0319808
          }, {
            "db": 1,
            "id": "RNCMPT00116",
            "alt": "YBX1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.90315, 0.0321446, 0.0321614, 0.0325438], 
              [0.750979, 0.231258, 0.00869074, 0.00907317], 
              [0.00890142, 0.973335, 0.00869074, 0.00907317], 
              [0.973562, 0.00867397, 0.00869074, 0.00907317], 
              [0.00890142, 0.192468, 0.00869074, 0.78994], 
              [0.00890142, 0.973335, 0.00869074, 0.00907317], 
              [0.544627, 0.0589665, 0.230499, 0.165908]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.98e-1",
            "evalue": "1.06e2",
            "pearsons_rho": -0.202144,
            "m": 0.0103453,
            "b": -0.985175,
            "mean_square_error": 0.0318841
          }, {
            "db": 1,
            "id": "RNCMPT00071",
            "alt": "SNRPA",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.264245, 0.0810372, 0.0810539, 0.573664], 
              [0.193179, 0.00985009, 0.00986686, 0.787103], 
              [0.0100775, 0.00985009, 0.969822, 0.0102493], 
              [0.0100775, 0.969805, 0.00986686, 0.0102493], 
              [0.970033, 0.00985009, 0.00986686, 0.0102493], 
              [0.0100775, 0.969805, 0.00986686, 0.0102493], 
              [0.624677, 0.0582391, 0.258446, 0.0586383]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.98e-1",
            "evalue": "1.06e2",
            "pearsons_rho": -0.206063,
            "m": 0.0115106,
            "b": -0.981295,
            "mean_square_error": 0.0318575
          }, {
            "db": 1,
            "id": "RNCMPT00178",
            "alt": "HNRPLL",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.644531, 0.0580491, 0.23897, 0.0584483], 
              [0.0112974, 0.966146, 0.0110868, 0.0114692], 
              [0.966373, 0.01107, 0.0110868, 0.0114692], 
              [0.185877, 0.518854, 0.093045, 0.202224], 
              [0.756959, 0.01107, 0.220502, 0.0114692], 
              [0.0112974, 0.966146, 0.0110868, 0.0114692], 
              [0.966373, 0.01107, 0.0110868, 0.0114692]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.98e-1",
            "evalue": "1.06e2",
            "pearsons_rho": -0.20804,
            "m": 0.00728927,
            "b": -0.989983,
            "mean_square_error": 0.0318438
          }, {
            "db": 1,
            "id": "RNCMPT00083",
            "alt": "YBX1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.976298, 0.00776154, 0.0077783, 0.00816073], 
              [0.827093, 0.156967, 0.0077783, 0.00816073], 
              [0.00798899, 0.976071, 0.0077783, 0.00816073], 
              [0.976298, 0.00776154, 0.0077783, 0.00816073], 
              [0.07819, 0.133683, 0.0077783, 0.780349], 
              [0.00798899, 0.976071, 0.0077783, 0.00816073], 
              [0.491316, 0.154297, 0.171861, 0.182526]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.99e-1",
            "evalue": "1.06e2",
            "pearsons_rho": -0.218742,
            "m": 0.013228,
            "b": -0.975446,
            "mean_square_error": 0.0317677
          }, {
            "db": 1,
            "id": "RNCMPT00149",
            "alt": "ENOX1",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.343107, 0.436842, 0.067441, 0.152609], 
              [0.583289, 0.145613, 0.226512, 0.0445866], 
              [0.0208099, 0.120411, 0.658362, 0.200417], 
              [0.937837, 0.0205824, 0.0205992, 0.0209816], 
              [0.0208099, 0.849988, 0.0205992, 0.108603], 
              [0.846209, 0.0205824, 0.112227, 0.0209816], 
              [0.0663283, 0.0661008, 0.80107, 0.0665]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "9.99e-1",
            "evalue": "1.06e2",
            "pearsons_rho": -0.226232,
            "m": 0.00459658,
            "b": -0.96525,
            "mean_square_error": 0.0317121
          }, {
            "db": 1,
            "id": "RNCMPT00156",
            "alt": "CNOT4",
            "len": 7,
            "motif_evalue": 0,
            "motif_nsites": 20,
            "url": "http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete_eukarya",
            "pwm": [
              [0.0300925, 0.169126, 0.770518, 0.0302642], 
              [0.980973, 0.00620333, 0.00622009, 0.00660252], 
              [0.00643078, 0.911043, 0.0759226, 0.00660252], 
              [0.980973, 0.00620333, 0.00622009, 0.00660252], 
              [0.00643078, 0.184335, 0.802632, 0.00660252], 
              [0.980973, 0.00620333, 0.00622009, 0.00660252], 
              [0.300821, 0.184294, 0.20976, 0.305125]
            ],
            "pos": 196,
            "neg": 0,
            "tp": -1,
            "fp": -1,
            "positive_thresh": 1.83017,
            "true_positive_thresh": 0,
            "number_of_tests": 1,
            "corrected_pvalue": "1.00e0",
            "evalue": "1.06e2",
            "pearsons_rho": -0.238134,
            "m": 0.00864216,
            "b": -0.966466,
            "mean_square_error": 0.0316196
          }
        ]
      };
    </script>
    <script type="text/javascript">
var site_url = "http://meme-suite.org/";

    </script>
    <script type="text/javascript">
/*
 * $
 *
 * Shorthand function for getElementById
 */
function $(el) {
  return document.getElementById(el);
}


/*
 * See http://stackoverflow.com/a/5450113/66387
 * Does string multiplication like the perl x operator.
 */
function string_mult(pattern, count) {
    if (count < 1) return '';
    var result = '';
    while (count > 1) {
        if (count & 1) result += pattern;
        count >>= 1, pattern += pattern;
    }
    return result + pattern;
}

/*
 * See http://stackoverflow.com/questions/814613/how-to-read-get-data-from-a-url-using-javascript
 * Slightly modified with information from
 * https://developer.mozilla.org/en/DOM/window.location
 */
function parse_params() {
  "use strict";
  var search, queryStart, queryEnd, query, params, nvPairs, i, nv, n, v;
  search = window.location.search;
  queryStart = search.indexOf("?") + 1;
  queryEnd   = search.indexOf("#") + 1 || search.length + 1;
  query      = search.slice(queryStart, queryEnd - 1);

  if (query === search || query === "") return {};

  params  = {};
  nvPairs = query.replace(/\+/g, " ").split("&");

  for (i = 0; i < nvPairs.length; i++) {
    nv = nvPairs[i].split("=");
    n  = decodeURIComponent(nv[0]);
    v  = decodeURIComponent(nv[1]);
    // allow a name to be used multiple times
    // storing each value in the array
    if (!(n in params)) {
      params[n] = [];
    }
    params[n].push(nv.length === 2 ? v : null);
  }
  return params;
}

/*
 * coords
 *
 * Calculates the x and y offset of an element.
 * From http://www.quirksmode.org/js/findpos.html
 * with alterations to take into account scrolling regions
 */
function coords(elem) {
  var myX = myY = 0;
  if (elem.getBoundingClientRect) {
    var rect;
    rect = elem.getBoundingClientRect();
    myX = rect.left + ((typeof window.pageXOffset !== "undefined") ?
        window.pageXOffset : document.body.scrollLeft);
    myY = rect.top + ((typeof window.pageYOffset !== "undefined") ?
        window.pageYOffset : document.body.scrollTop);
  } else {
    // this fall back doesn't properly handle absolutely positioned elements
    // inside a scrollable box
    var node;
    if (elem.offsetParent) {
      // subtract all scrolling
      node = elem;
      do {
        myX -= node.scrollLeft ? node.scrollLeft : 0;
        myY -= node.scrollTop ? node.scrollTop : 0;
      } while (node = node.parentNode);
      // this will include the page scrolling (which is unwanted) so add it back on
      myX += (typeof window.pageXOffset !== "undefined") ? window.pageXOffset : document.body.scrollLeft;
      myY += (typeof window.pageYOffset !== "undefined") ? window.pageYOffset : document.body.scrollTop;
      // sum up offsets
      node = elem;
      do {
        myX += node.offsetLeft;
        myY += node.offsetTop;
      } while (node = node.offsetParent);
    }
  }
  return [myX, myY];
}

/*
 * position_popup
 *
 * Positions a popup relative to an anchor element.
 *
 * The available positions are:
 * 0 - Centered below the anchor.
 */
function position_popup(anchor, popup, position) {
  "use strict";
  var a_x, a_y, a_w, a_h, p_x, p_y, p_w, p_h;
  var a_xy, spacer, margin, scrollbar, page_w;
  // define constants
  spacer = 5;
  margin = 15;
  scrollbar = 15;
  // define the positions and widths
  a_xy = coords(anchor);
  a_x = a_xy[0];
  a_y = a_xy[1];
  a_w = anchor.offsetWidth;
  a_h = anchor.offsetHeight;
  p_w = popup.offsetWidth;
  p_h = popup.offsetHeight;
  page_w = null;
  if (window.innerWidth) {
    page_w = window.innerWidth;
  } else if (document.body) {
    page_w = document.body.clientWidth;
  }
  // check the position type is defined
  if (typeof position !== "number") {
    position = 0;
  }
  // calculate the popup position
  switch (position) {
    case 1:
      p_x = a_x + a_w + spacer;
      p_y = a_y + (a_h / 2) - (p_h / 2);
      break;
    case 0:
    default:
      p_x = a_x + (a_w / 2) - (p_w / 2);
      p_y = a_y + a_h + spacer;
      break;
  }
  // constrain the popup position
  if (p_x < margin) {
    p_x = margin;
  } else if (page_w != null && (p_x + p_w) > (page_w - margin - scrollbar)) {
    p_x = page_w - margin - scrollbar - p_w;
  }
  if (p_y < margin) {
    p_y = margin;
  }
  // position the popup
  popup.style.left = p_x + "px";
  popup.style.top = p_y + "px";
}

function lookup_help_popup(popup_id) {
  var _body, pop, info;
  pop = document.getElementById(popup_id);
  if (pop == null) {
    _body = document.getElementsByTagName("body")[0];
    pop = document.createElement("div");
    pop.className = "pop_content";
    pop.id = popup_id;
    pop.style.backgroundColor = "#FFC";
    pop.style.borderColor = "black";
    info = document.createElement("p");
    info.style.fontWeight = "bold";
    info.appendChild(document.createTextNode("Error: No popup for topic \"" + popup_id + "\"."));
    pop.appendChild(info);
    // this might cause problems with the menu, but as this only happens
    // when something is already wrong I don't think that's too much of a problem
    _body.insertBefore(pop, _body.firstChild);
  }
  if (document.getElementsByTagName('body')[0].hasAttribute("data-autobtns")) {
    if (!/\bauto_buttons\b/.test(pop.className)) {
      pop.className += " auto_buttons";
      var back_btn_sec = document.createElement("div");
      back_btn_sec.className = "nested_only pop_back_sec";
      var back_btn = document.createElement("span");
      back_btn.className = "pop_back";
      back_btn.appendChild(document.createTextNode("<< back"));
      back_btn.addEventListener("click", function(e) {
        help_return();
      }, false);
      back_btn_sec.appendChild(back_btn);
      pop.insertBefore(back_btn_sec, pop.firstChild);
      var close_btn_sec = document.createElement("div");
      close_btn_sec.className = "pop_close_sec";
      var close_btn = document.createElement("span");
      close_btn.className = "pop_close";
      close_btn.appendChild(document.createTextNode("close"));
      close_btn.addEventListener("click", function(e) {
        help_popup();
      }, false);
      close_btn_sec.appendChild(close_btn);
      pop.appendChild(close_btn_sec);
    }
  }
  return pop;
}

/*
 * help_popup
 *
 * Moves around help pop-ups so they appear
 * below an activator.
 */
function help_popup(activator, popup_id) {
  "use strict";
  var pop;
  // set default values
  if (typeof help_popup.popup === "undefined") {
    help_popup.popup = [];
  }
  if (typeof help_popup.activator === "undefined") {
    help_popup.activator = null;
  }
  var last_pop = (help_popup.popup.length > 0 ? help_popup.popup[help_popup.popup.length - 1] : null);
  if (typeof(activator) == "undefined") { // no activator so hide
    if (last_pop != null) {
      last_pop.style.display = 'none';
      help_popup.popup = [];
    }
    return;
  }
  pop = lookup_help_popup(popup_id);
  if (pop == last_pop) {
    if (activator == help_popup.activator) {
      //hide popup (as we've already shown it for the current help button)
      last_pop.style.display = 'none';
      help_popup.popup = [];
      return; // toggling complete!
    }
  } else if (last_pop != null) {
    //activating different popup so hide current one
    last_pop.style.display = 'none';
  }
  help_popup.popup = [pop];
  help_popup.activator = activator;
  toggle_class(pop, "nested", false);
  //must make the popup visible to measure it or it has zero width
  pop.style.display = 'block';
  position_popup(activator, pop);
}

/*
 * help_refine
 * 
 * Intended for links within a help popup. Stores a stack of state so
 * you can go back.
 */
function help_refine(popup_id) {
  if (help_popup.popup == null || help_popup.popup.length == 0 || help_popup.activator == null) {
    //throw new Error("Cannot refine a help popup when one is not shown!");
    var pop = lookup_help_popup(popup_id);
    var act_id = popup_id + '_act';
    var activator = document.getElementById(act_id);
    help_popup(activator, popup_id);
  }
  var pop = lookup_help_popup(popup_id);
  var last_pop = help_popup.popup[help_popup.popup.length - 1];
  if (pop == last_pop) return; // slightly odd, but no real cause for alarm
  help_popup.popup.push(pop);
  toggle_class(pop, "nested", true);
  last_pop.style.display = "none";
  //must make the popup visible to measure it or it has zero width
  pop.style.display = "block";
  position_popup(help_popup.activator, pop);
}

/*
 * help_return
 * 
 * Intended for links within a help popup. Stores a stack of state so
 * you can go back.
 */
function help_return() {
  if (help_popup.popup == null || help_popup.popup.length == 0 || help_popup.activator == null) {
    throw new Error("Can not return to a earlier help popup when one is not shown!");
  }
  var last_pop = help_popup.popup.pop();
  last_pop.style.display = "none";
  var pop = (help_popup.popup.length > 0 ? help_popup.popup[help_popup.popup.length - 1] : null);
  if (pop != null) {
    toggle_class(pop, "nested", help_popup.popup.length > 1);
    pop.style.display = "block";
    position_popup(help_popup.activator, pop);
  } else {
    help_popup.activator = null;
  }
}

/*
 * update_scroll_pad
 *
 * Creates padding at the bottom of the page to allow
 * scrolling of anything into view.
 */
function update_scroll_pad() {
  var page, pad;
  page = (document.compatMode === "CSS1Compat") ? document.documentElement : document.body;
  pad = $("scrollpad");
  if (pad === null) {
    pad = document.createElement("div");
    pad.id = 'scrollpad';
    document.getElementsByTagName('body')[0].appendChild(pad);
  }
  pad.style.height = Math.abs(page.clientHeight - 100) + "px";
}

function substitute_classes(node, remove, add) {
  "use strict";
  var list, all, i, cls, classes;
  list = node.className.split(/\s+/);
  all = {};
  for (i = 0; i < list.length; i++) {
    if (list[i].length > 0) all[list[i]] = true;
  }
  for (i = 0; i < remove.length; i++) {
    if (all.hasOwnProperty(remove[i])) {
      delete all[remove[i]];
    }
  }
  for (i = 0; i < add.length; i++) {
    all[add[i]] = true;
  }
  classes = "";
  for (cls in all) {
    classes += cls + " ";
  }
  node.className = classes;
}

/*
 * toggle_class
 *
 * Adds or removes a class from the node. If the parameter 'enabled' is not 
 * passed then the existence of the class will be toggled, otherwise it will be
 * included if enabled is true.
 */
function toggle_class(node, cls, enabled) {
  var classes = node.className;
  var list = classes.replace(/^\s+/, '').replace(/\s+$/, '').split(/\s+/);
  var found = false;
  for (var i = 0; i < list.length; i++) {
    if (list[i] == cls) {
      list.splice(i, 1);
      i--;
      found = true;
    }
  }
  if (typeof enabled == "undefined") {
    if (!found) list.push(cls);
  } else {
    if (enabled) list.push(cls);
  }
  node.className = list.join(" ");
}

/*
 * find_child
 *
 * Searches child nodes in depth first order and returns the first it finds
 * with the className specified.
 * TODO replace with querySelector
 */
function find_child(node, className) {
  var pattern;
  if (node == null || typeof node !== "object") {
    return null;
  }
  if (typeof className === "string") {
    pattern = new RegExp("\\b" + className + "\\b");
  } else {
    pattern = className;
  }
  if (node.nodeType == Node.ELEMENT_NODE && 
      pattern.test(node.className)) {
    return node;
  } else {
    var result = null;
    for (var i = 0; i < node.childNodes.length; i++) {
      result = find_child(node.childNodes[i], pattern);
      if (result != null) break;
    }
    return result;
  }
}

/*
 * find_parent
 *
 * Searches parent nodes outwards from the node and returns the first it finds
 * with the className specified.
 */
function find_parent(node, className) {
  var pattern;
  pattern = new RegExp("\\b" + className + "\\b");
  do {
    if (node.nodeType == Node.ELEMENT_NODE && 
        pattern.test(node.className)) {
      return node;
    }
  } while (node = node.parentNode);
  return null;
}

/*
 * find_parent_tag
 *
 * Searches parent nodes outwards from the node and returns the first it finds
 * with the tag name specified. HTML tags should be specified in upper case.
 */
function find_parent_tag(node, tag_name) {
  do {
    if (node.nodeType == Node.ELEMENT_NODE && node.tagName == tag_name) {
      return node;
    }
  } while (node = node.parentNode);
  return null;
}

/*
 * __toggle_help
 *
 * Uses the 'topic' property of the this object to
 * toggle display of a help topic.
 *
 * This function is not intended to be called directly.
 */
function __toggle_help(e) {
  if (!e) e = window.event;
  if (e.type === "keydown") {
    if (e.keyCode !== 13 && e.keyCode !== 32) {
      return;
    }
    // stop a submit or something like that
    e.preventDefault();
  }

  help_popup(this, this.getAttribute("data-topic"));
}

function setup_help_button(button) {
  "use strict";
  var topic;
  if (button.hasAttribute("data-topic")) {
    topic = button.getAttribute("data-topic");
    if (document.getElementById(topic) != null) {
      button.tabIndex = "0"; // make keyboard selectable
      button.addEventListener("click", function() {
        help_popup(button, topic);
      }, false);
      button.addEventListener("keydown", function(e) {
        // toggle only on Enter or Spacebar, let other keys do their thing
        if (e.keyCode !== 13 && e.keyCode !== 32) return;
        // stop a submit or something like that
        e.preventDefault();
        help_popup(button, topic);
      }, false);
    } else {
      button.style.visibility = "hidden";
    }
  }
  button.className += " active";
}

/*
 * help_button
 *
 * Makes a help button for the passed topic.
 */
function help_button(topic) {
  var btn = document.createElement("div");
  btn.className = "help";
  btn.setAttribute("data-topic", topic);
  setup_help_button(btn);
  return btn;
}

/*
 * prepare_download
 *
 * Sets the attributes of a link to setup a file download using the given content.
 * If no link is provided then create one and click it.
 */
function prepare_download(content, mimetype, filename, link) {
  "use strict";
  // if no link is provided then create one and click it
  var click_link = false;
  if (!link) {
    link = document.createElement("a");
    click_link = true;
  }
  try {
    // Use a BLOB to convert the text into a data URL.
    // We could do this manually with a base 64 conversion.
    // This will only be supported on modern browsers,
    // hence the try block.
    var blob = new Blob([content], {type: mimetype});
    var reader = new FileReader();
    reader.onloadend = function() {
      // If we're lucky the browser will also support the download
      // attribute which will let us suggest a file name to save the link.
      // Otherwise it is likely that the filename will be unintelligible. 
      link.setAttribute("download", filename);
      link.href = reader.result;
      if (click_link) {
        // must add the link to click it
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
      }
    }
    reader.readAsDataURL(blob);
  } catch (error) {
    if (console && console.log) console.log(error);
    // probably an old browser
    link.href = "";
    link.visible = false;
  }
}

/*
 * add_cell
 *
 * Add a cell to the table row.
 */
function add_cell(row, node, cls, click_action) {
  var cell = row.insertCell(row.cells.length);
  if (node) cell.appendChild(node);
  if (cls && cls !== "") cell.className = cls;
  if (click_action) cell.addEventListener("click", click_action, false);
}

/*
 * add_header_cell
 *
 * Add a header cell to the table row.
 */
function add_header_cell(row, node, help_topic, cls, colspan) {
  var th = document.createElement("th");
  if (node) th.appendChild(node);
  if (help_topic && help_topic !== "") th.appendChild(help_button(help_topic));
  if (cls && cls !== "") th.className = cls;
  if (typeof colspan == "number" && colspan > 1) th.colSpan = colspan;
  row.appendChild(th);
}

/*
 * add_text_cell
 *
 * Add a text cell to the table row.
 */
function add_text_cell(row, text, cls, action) {
  var node = null;
  if (typeof(text) != 'undefined') node = document.createTextNode(text);
  add_cell(row, node, cls, action);
}

/*
 * add_text_header_cell
 *
 * Add a text header cell to the table row.
 */
function add_text_header_cell(row, text, help_topic, cls, action, colspan) {
  var node = null;
  if (typeof(text) != 'undefined') {
    var nbsp = (help_topic ? "\u00A0" : "");
    var str = "" + text;
    var parts = str.split(/\n/);
    if (parts.length === 1) {
      if (action) {
        node = document.createElement("span");
        node.appendChild(document.createTextNode(str + nbsp));
      } else {
        node = document.createTextNode(str + nbsp);
      }
    } else {
      node = document.createElement("span");
      for (var i = 0; i < parts.length; i++) {
        if (i !== 0) {
          node.appendChild(document.createElement("br"));
        }
        node.appendChild(document.createTextNode(parts[i]));
      }
    }
    if (action) {
      node.addEventListener("click", action, false);
      node.style.cursor = "pointer";
    }
  }
  add_header_cell(row, node, help_topic, cls, colspan);
}

function setup_help() {
  "use strict";
  var help_buttons, i;
  help_buttons = document.querySelectorAll(".help:not(.active)");
  for (i = 0; i < help_buttons.length; i++) {
    setup_help_button(help_buttons[i]);
  }
}

function setup_scrollpad() {
  "use strict";
  if (document.getElementsByTagName('body')[0].hasAttribute("data-scrollpad") && document.getElementById("scrollpad") == null) {
    window.addEventListener("resize", update_scroll_pad, false);
    update_scroll_pad();
  }
}

// anon function to avoid polluting global scope
(function() {
  "use strict";
  window.addEventListener("load", function load(evt) {
    window.removeEventListener("load", load, false);
    setup_help();
    setup_scrollpad();
  }, false);
})();

/*
 *  make_link
 *
 *  Creates a text node and if a URL is specified it surrounds it with a link.
 *  If the URL doesn't begin with "http://" it automatically adds it, as
 *  relative links don't make much sense in this context.
 */
function make_link(text, url) {
  var textNode = null;
  var link = null;
  if (typeof text !== "undefined" && text !== null) textNode = document.createTextNode(text);
  if (typeof url === "string") {
    if (url.indexOf("//") == -1) {
      url = "http://" + url;
    }
    link = document.createElement('a');
    link.href = url;
    if (textNode) link.appendChild(textNode);
    return link;
  }
  return textNode;
}

//
// Function to create an HTML paragraph describing the 
// MEME Suite background model source.
//
function make_background_source(title, source, text) {
  var paraNode = document.createElement("P");
  var titleNode = document.createElement("B");
  var textNode1 = document.createTextNode("\u00A0\u00A0\u00A0\u00A0" + title + ": ");
  titleNode.appendChild(textNode1);
  var source_text = ((source == "--motif--") ? "the (first) motif file" : (source == "--nrdb--") ? "an old version of the NCBI non-redundant database" : (source == "--uniform--") ? "the uniform model" : (source == "--query--") ? "the query file" : (source == "--sequences--") ? "built from the (primary) sequences" : ("the file '" + source + "'"));
  if (text) { return source_text; }
  var textNode2 = document.createTextNode(source_text);
  paraNode.appendChild(titleNode);
  paraNode.appendChild(textNode2);
  return paraNode;
}

// Function to create a help button
function make_help_button(container, help_topic) {
  container.appendChild(help_button(help_topic));
}

    </script>
    <script type="text/javascript">
//======================================================================
// start Alphabet object
//======================================================================
var Alphabet = function(alphabet, background) {
  "use strict";
  var i, j, sym, aliases, complement, comp_e_sym, ambigs, generate_background;
  generate_background = (background == null);
  if (generate_background) {
    background = [];
    for (i = 0; i < alphabet.ncore; i++) background[i] = 1.0 / alphabet.ncore;
  } else if (alphabet.ncore != background.length) {
    throw new Error("The background length does not match the alphabet length.");
  }
  this.name = alphabet.name;
  this.like = (alphabet.like != null ? alphabet.like.toUpperCase() : null);
  this.ncore = alphabet.ncore;
  this.symbols = alphabet.symbols;
  this.background = background;
  this.genbg = generate_background;
  this.encode = {};
  this.encode2core = {};
  this.complement = {};
  // check if all symbols are same case
  var seen_uc = false;
  var seen_lc = false;
  var check_case = function (syms) {
    var s, sym;
    if (typeof syms === "string") {
      for (s = 0; s < syms.length; s++) {
        sym = syms.charAt(s);
        if (sym >= 'a' && sym <= 'z') seen_lc = true;
        else if (sym >= 'A' && sym <= 'Z') seen_uc = true;
      }
    }
  };
  for (i = 0; i < this.symbols.length; i++) {
    check_case(this.symbols[i].symbol);
    check_case(this.symbols[i].aliases);
  }
  // now map symbols to indexes
  var update_array = function(array, syms, index) {
    var s, sym;
    if (typeof syms === "string") {
      for (s = 0; s < syms.length; s++) {
        sym = syms.charAt(s);
        array[sym] = index;
        // when only a single case is used, then encode as case insensitive
        if (seen_uc != seen_lc) {
          if (sym >= 'a' && sym <= 'z') {
            array[sym.toUpperCase()] = index;
          } else if (sym >= 'A' && sym <= 'Z') {
            array[sym.toLowerCase()] = index;
          }
        }
      }
    }
  }
  // map core symbols to index
  for (i = 0; i < this.ncore; i++) {
    update_array(this.encode2core, this.symbols[i].symbol, i);
    update_array(this.encode, this.symbols[i].symbol, i);
    update_array(this.encode2core, this.symbols[i].aliases, i);
    update_array(this.encode, this.symbols[i].aliases, i);
  }
  // map ambiguous symbols to index
  ambigs = {};
  for (i = this.ncore; i < this.symbols.length; i++) {
    update_array(this.encode, this.symbols[i].symbol, i);
    update_array(this.encode, this.symbols[i].aliases, i);
    ambigs[this.symbols[i].equals] = i;
  }
  // determine complements
  for (i = 0; i < this.ncore; i++) {
    complement = this.symbols[i].complement;
    if (typeof complement === "string") {
      this.complement[i] = this.encode2core[complement];
    }
  }
  next_symbol:
  for (i = this.ncore; i < this.symbols.length; i++) {
    complement = "";
    for (j = 0; j < this.symbols[i].equals.length; j++) {
      comp_e_sym = this.complement[this.encode2core[this.symbols[i].equals.charAt(j)]];
      if (typeof comp_e_sym !== "number") continue next_symbol;
      complement += this.symbols[comp_e_sym].symbol;
    }
    complement = complement.split("").sort().join("");
    if (typeof ambigs[complement] === "number") {
      this.complement[i] = ambigs[complement];
    }
  }
  // determine case insensitivity
  this.case_insensitive = true;
  if (seen_uc == seen_lc) {
    // when there is a mixture of cases it probably won't
    // be case insensitive but we still need to check
    loop:
    for (i = 0; i < this.symbols.length; i++) {
      sym = this.symbols[i].symbol;
      if (sym >= 'A' && sym <= 'Z') {
        if (this.encode[sym.toLowerCase()] != i) {
          this.case_insensitive = false;
          break loop;
        }
      } else if (sym >= 'a' && sym <= 'z') {
        if (this.encode[sym.toUpperCase()] != i) {
          this.case_insensitive = false;
          break loop;
        }
      }
      aliases = this.symbols[i].aliases;
      if (aliases != null) {
        for (j = 0; j < aliases.length; j++) {
          sym = aliases.charAt(j);
          if (sym >= 'A' && sym <= 'Z') {
            if (this.encode[sym.toLowerCase()] != i) {
              this.case_insensitive = false;
              break loop;
            }
          } else if (sym >= 'a' && sym <= 'z') {
            if (this.encode[sym.toUpperCase()] != i) {
              this.case_insensitive = false;
              break loop;
            }
          }
        }
      }
    }
  }
  // normalise aliases to remove the prime symbol and eliminate
  // the alternate cases when the alphabet is case insensitive
  var seen, out;
  for (i = 0; i < this.symbols.length; i++) {
    sym = this.symbols[i].symbol;
    aliases = this.symbols[i].aliases;
    if (typeof aliases != "string") aliases = "";
    seen = {};
    out = [];
    if (this.case_insensitive) {
      sym = sym.toUpperCase();
      aliases = aliases.toUpperCase();
    }
    seen[sym] = true;
    for (j = 0; j < aliases.length; j++) {
      if (!seen[aliases.charAt(j)]) {
        seen[aliases.charAt(j)] = true;
        out.push(aliases.charAt(j));
      }
    }
    this.symbols[i].aliases = out.sort().join("");
  }
};
// return the name of the alphabet
Alphabet.prototype.get_alphabet_name = function() {
  return this.name;
};
// return if the alphabet can be complemented
Alphabet.prototype.has_complement = function() {
  return (typeof this.symbols[0].complement === "string");
};
// return true if an uppercase letter has the same meaning as the lowercase form
Alphabet.prototype.is_case_insensitive = function() {
  return this.case_insensitive;
};
// return the information content of an alphabet letter
Alphabet.prototype.get_ic = function() {
  return Math.log(this.ncore) / Math.LN2;
};
// return the count of the core alphabet symbols
Alphabet.prototype.get_size_core = function() {
  return this.ncore;
};
// return the count of all alphabet symbols
Alphabet.prototype.get_size_full = function() {
  return this.symbols.length;
};
// return the symbol for the given alphabet index
Alphabet.prototype.get_symbol = function(alph_index) {
  "use strict";
  if (alph_index < 0 || alph_index >= this.symbols.length) {
    throw new Error("Alphabet index out of bounds");
  }
  return this.symbols[alph_index].symbol;
};
// return the aliases for the given alphabet index
Alphabet.prototype.get_aliases = function(alph_index) {
  "use strict";
  if (alph_index < 0 || alph_index >= this.symbols.length) {
    throw new Error("Alphabet index out of bounds");
  }
  var sym_obj = this.symbols[alph_index];
  return (sym_obj.aliases != null ? sym_obj.aliases : "");
};
// return the name for the given alphabet index
Alphabet.prototype.get_name = function(alph_index) {
  "use strict";
  var sym;
  if (alph_index < 0 || alph_index >= this.symbols.length) {
    throw new Error("Alphabet index out of bounds");
  }
  sym = this.symbols[alph_index];
  return (typeof sym.name === "string" ? sym.name : sym.symbol);
};
// return the alphabet it is like or null
Alphabet.prototype.get_like = function() {
  "use strict";
  return this.like;
};
// return the index of the complement for the given alphabet index
Alphabet.prototype.get_complement = function(alph_index) {
  var comp_e_sym = this.complement[alph_index];
  if (typeof comp_e_sym === "number") {
    return comp_e_sym;
  } else {
    return -1;
  }
};
// return a string containing the core symbols
Alphabet.prototype.get_symbols = function() {
  "use strict";
  var i, core_symbols;
  core_symbols = "";
  for (i = 0; i < this.ncore; i++) {
    core_symbols += this.symbols[i].symbol;
  }
  return core_symbols;
};
// return if the background was not a uniform generated background
Alphabet.prototype.has_bg = function() {
  "use strict";
  return !this.genbg;
};
// get the background frequency for the index
Alphabet.prototype.get_bg_freq = function(alph_index) {
  "use strict";
  var freq, i, symbols;
  if (alph_index >= 0) {
    if (alph_index < this.ncore) {
      return this.background[alph_index];
    } else if (alph_index < this.symbols.length) {
      freq = 0;
      symbols = this.symbols[alph_index].equals;
      for (i = 0; i < symbols.length; i++) {
        freq += this.background[this.encode2core[symbols.charAt(i)]];
      }
      return freq;
    } 
  }
  throw new Error("The alphabet index is out of range.");
};
// get the colour of the index
Alphabet.prototype.get_colour = function(alph_index) {
  "use strict";
  if (alph_index < 0 || alph_index >= this.symbols.length) {
    throw new Error("BAD_ALPHABET_INDEX");
  }
  if (typeof this.symbols[alph_index].colour != "string") {
    return "black";
  }
  return "#" + this.symbols[alph_index].colour;
};
// get the rgb components of the colour at the index
Alphabet.prototype.get_rgb = function(alph_index) {
  "use strict";
  if (alph_index < 0 || alph_index >= this.symbols.length) {
    throw new Error("BAD_ALPHABET_INDEX");
  }
  if (typeof this.symbols[alph_index].colour != "string") {
    return {"red": 0, "green": 0, "blue": 0};
  }
  var colour = this.symbols[alph_index].colour;
  var red = parseInt(colour.substr(0, 2), 16) / 255;
  var green = parseInt(colour.substr(2, 2), 16) / 255;
  var blue = parseInt(colour.substr(4, 2), 16) / 255;
  return {"red": red, "green": green, "blue": blue};
};
// convert a symbol into the index
Alphabet.prototype.get_index = function(letter) {
  "use strict";
  var alph_index;
  alph_index = this.encode[letter];
  if (typeof alph_index === "undefined") {
    return -1;
  }
  return alph_index;
};
// convert a symbol into the list of core indexes that it equals
Alphabet.prototype.get_indexes = function(letter) {
  "use strict";
  var alph_index, comprise_str, i, comprise_list;
  alph_index = this.encode[letter];
  if (typeof alph_index === "undefined") {
    throw new Error("Unknown letter");
  }
  comprise_str = this.symbols[alph_index].equals;
  comprise_list = [];
  if (typeof comprise_str == "string") {
    for (i = 0; i < comprise_str.length; i++) {
      comprise_list.push(this.encode2core[comprise_str.charAt(i)]);
    }
  } else {
    comprise_list.push(alph_index);
  }
  return comprise_list;
};
// check if a symbol is the primary way of representing the symbol in the alphabet
Alphabet.prototype.is_prime_symbol = function(letter) {
  var alph_index;
  alph_index = this.encode[letter];
  if (alph_index == null) return false;
  if (this.is_case_insensitive()) {
    return (this.symbols[alph_index].symbol.toUpperCase() == letter.toUpperCase());
  } else {
    return (this.symbols[alph_index].symbol == letter);
  }
};
// compare 2 alphabets
Alphabet.prototype.equals = function(other) {
  "use strict";
  var i, sym1, sym2;
  // first check that it's actually an alphabet object
  if (!(typeof other === "object" && other != null && other instanceof Alphabet)) {
    return false;
  }
  // second shortcircuit if it's the same object
  if (this === other) return true;
  // compare
  if (this.name !== other.name) return false;
  if (this.ncore !== other.ncore) return false;
  if (this.symbols.length !== other.symbols.length) return false;
  for (i = 0; i < this.symbols.length; i++) {
    sym1 = this.symbols[i];
    sym2 = other.symbols[i];
    if (sym1.symbol !== sym2.symbol) return false;
    if (sym1.aliases !== sym2.aliases) return false;
    if (sym1.name !== sym2.name) return false;
    if (typeof sym1.colour !== typeof sym2.colour || 
        (typeof sym1.colour === "string" && typeof sym2.colour === "string" &&
         parseInt(sym1.colour, 16) != parseInt(sym2.colour, 16))) {
      return false;
    }
    if (sym1.complement !== sym2.complement) return false;
    if (sym1.equals !== sym2.equals) return false;
  }
  return true;
};
Alphabet.prototype.check_core_subset = function(super_alph) {
  var complement_same = true;
  var seen_set = {};
  var sub_i, sub_symbol, super_i, super_symbol;
  for (sub_i = 0; sub_i < this.ncore; sub_i++) {
    sub_symbol = this.symbols[sub_i];
    super_i = super_alph.encode[sub_symbol.symbol]; 
    if (super_i == null) return 0;
    super_symbol = super_alph.symbols[super_i];
    if (seen_set[super_i]) return 0;
    seen_set[super_i] = true;
    // check complement
    if (sub_symbol.complement != null && super_symbol.complement != null) {
      if (super_alph.encode[sub_symbol.complement] != super_alph.encode[super_symbol.complement]) {
        complement_same = false;
      }
    } else if (sub_symbol.complement != null || super_symbol.complement != null) {
      complement_same = false;
    }
  }
  return (complement_same ? 1 : -1);
};
// convert a sequence to its reverse complement
Alphabet.prototype.invcomp_seq = function(seq) {
  "use strict";
  var syms, i, e_sym, comp_e_sym;
  if (!this.has_complement()) throw new Error("Alphabet must be complementable");
  syms = seq.split("");
  for (i = 0; i < syms.length; i++) {
    e_sym = this.encode[syms[i]];
    if (typeof e_sym === "undefined") {
      e_sym = this.ncore; // wildcard
    }
    comp_e_sym = this.complement[e_sym];
    if (typeof comp_e_sym === "undefined") {
      comp_e_sym = e_sym; // not complementable
    }
    syms[i] = this.symbols[comp_e_sym].symbol;
  }
  return syms.reverse().join("");
};
// convert the alphabet to the text version
Alphabet.prototype.as_text = function() {
  "use strict";
  function name_as_text(name) {
    var i, c, out;
    out = "\"";
    for (i = 0; i < name.length; i++) {
      c = name.charAt(i);
      if (c == "\"") {
        out += "\\\"";
      } else if (c == "/") {
        out += "\\/";
      } else if (c == "\\") {
        out += "\\\\";
      } else {
        out += c;
      }
    }
    out += "\"";
    return out;
  }
  function symbol_as_text(sym) {
    var out;
    out = sym.symbol;
    if (typeof sym.name === "string" && sym.name != sym.symbol) {
      out += " " + name_as_text(sym.name);
    }
    if (typeof sym.colour === "string") {
      out += " " + sym.colour;
    }
    return out;
  }
  var out, i, j, c, sym;
  out = "";
  // output core symbols with 2 way complements
  for (i = 0; i < this.ncore; i++) {
    c = this.complement[i];
    if (typeof c === "number" && i < c && this.complement[c] === i) {
      out += symbol_as_text(this.symbols[i]) + " ~ " + symbol_as_text(this.symbols[c]) + "\n";  
    }
  }
  // output core symbols with no complement
  for (i = 0; i < this.ncore; i++) {
    if (typeof this.complement[i] === "undefined") {
      out += symbol_as_text(this.symbols[i]) + "\n";
    }
  }
  // output ambiguous symbols that have comprising characters
  for (i = this.ncore; i < this.symbols.length; i++) {
    if (this.symbols[i].equals.length == 0) break;
    out += symbol_as_text(this.symbols[i]) + " = " + this.symbols[i].equals + "\n";
    if (typeof this.symbols[i].aliases === "string") {
      for (j = 0; j < this.symbols[i].aliases.length; j++) {
        if (this.symbols[i].aliases.charAt(j) == this.symbols[i].symbol) continue;
        out += this.symbols[i].aliases.charAt(j) + " = " + this.symbols[i].equals + "\n";
      }
    }
  }
  // output aliases of core symbols
  for (i = 0; i < this.ncore; i++) {
    if (typeof this.symbols[i].aliases === "string") {
      for (j = 0; j < this.symbols[i].aliases.length; j++) {
        if (this.symbols[i].aliases.charAt(j) == this.symbols[i].symbol) continue;
        out += this.symbols[i].aliases.charAt(j) + " = " + this.symbols[i].symbol + "\n";
      }
    }
  }
  // output gap symbols
  i = this.symbols.length - 1;
  if (this.symbols[i].equals.length == 0) {
    out += symbol_as_text(this.symbols[i]) + " =\n";
    if (typeof this.symbols[i].aliases === "string") {
      for (j = 0; j < this.symbols[i].aliases.length; j++) {
        if (this.symbols[i].aliases.charAt(j) == this.symbols[i].symbol) continue;
        out += this.symbols[i].aliases.charAt(j) + " =\n";
      }
    }
  }
  return out;
};
// output the alphabet as it appears in minimal MEME format
Alphabet.prototype.as_meme = function() {
  "use strict";
  function name_as_text(name) {
    var i, c, out;
    out = "\"";
    for (i = 0; i < name.length; i++) {
      c = name.charAt(i);
      if (c == "\"") {
        out += "\\\"";
      } else if (c == "/") {
        out += "\\/";
      } else if (c == "\\") {
        out += "\\\\";
      } else {
        out += c;
      }
    }
    out += "\"";
    return out;
  }
  if (this.equals(AlphStd.DNA)) {
    return "ALPHABET= ACGT\n";
  } else if (this.equals(AlphStd.PROTEIN)) {
    return "ALPHABET= ACDEFGHIKLMNPQRSTVWY\n";
  } else {
    return "ALPHABET" + 
      (this.name != null ? " " + name_as_text(this.name) : "") + 
      (this.like != null ? " " + this.like + "-LIKE" : "") + "\n" +
      this.as_text() + "END ALPHABET\n";
  }
};

// Returns a table showing all the letters in the alphabet
Alphabet.prototype.as_table = function() {
  "use strict";
  var i, j, row, th, td, aliases, equals, sym;
  var table = document.createElement("table");
  // create the core symbol header
  row = table.insertRow(table.rows.length);
  th = document.createElement("th");
  th.appendChild(document.createTextNode("Symbol(s)"));
  row.appendChild(th);
  th = document.createElement("th");
  th.appendChild(document.createTextNode("Name"));
  row.appendChild(th);
  th = document.createElement("th");
  if (this.has_complement()) {
    th.appendChild(document.createTextNode("Complement"));
  }
  row.appendChild(th);
  // list the core symbols
  for (i = 0; i < this.ncore; i++) {
    row = table.insertRow(table.rows.length);
    td = document.createElement("td");
    if (this.symbols[i].colour != null) {
      td.style.color = '#' + this.symbols[i].colour;
    }
    td.appendChild(document.createTextNode(this.symbols[i].symbol));
    aliases = this.get_aliases(i);
    if (aliases.length > 0) {
      td.appendChild(document.createTextNode(' ' + aliases.split('').join(' ')));
    }
    row.appendChild(td);
    td = document.createElement("td");
    if (this.symbols[i].name != null) {
      td.appendChild(document.createTextNode(this.symbols[i].name));
    }
    row.appendChild(td);
    td = document.createElement("td");
    if (this.symbols[i].complement != null) {
      td.style.color = this.get_colour(this.get_index(this.symbols[i].complement));
      td.appendChild(document.createTextNode(this.symbols[i].complement));
    }
    row.appendChild(td);
  }
  // create the ambiguous symbol header
  row = table.insertRow(table.rows.length);
  th = document.createElement("th");
  th.appendChild(document.createTextNode("Symbol(s)"));
  row.appendChild(th);
  th = document.createElement("th");
  th.appendChild(document.createTextNode("Name"));
  row.appendChild(th);
  th = document.createElement("th");
  th.appendChild(document.createTextNode("Matches"));
  row.appendChild(th);
  // list the ambiguous symbols
  for (i = this.ncore; i < this.symbols.length; i++) {
    row = table.insertRow(table.rows.length);
    td = document.createElement("td");
    if (this.symbols[i].colour != null) {
      td.style.color = '#' + this.symbols[i].colour;
    }
    td.appendChild(document.createTextNode(this.symbols[i].symbol));
    aliases = this.get_aliases(i);
    if (aliases.length > 0) {
      td.appendChild(document.createTextNode(' ' + aliases.split('').join(' ')));
    }
    row.appendChild(td);
    td = document.createElement("td");
    if (this.symbols[i].name != null) {
      td.appendChild(document.createTextNode(this.symbols[i].name));
    }
    row.appendChild(td);
    td = document.createElement("td");
    equals = this.symbols[i].equals.split('');
    for (j = 0; j < equals.length; j++) {
      if (j != 0) td.appendChild(document.createTextNode(' '));
      sym = document.createElement("span");
      sym.style.color = this.get_colour(this.get_index(equals[j]));
      sym.appendChild(document.createTextNode(equals[j]));
      td.appendChild(sym);
    }
    row.appendChild(td);
  }
  return table;
};

// returns a dictionary of the colours for EPS
Alphabet.prototype._as_eps_dict = function() {
  "use strict";
  var i, sym, rgb;
  var out = "/fullColourDict <<\n";
  for (i = 0; i < this.ncore; i++) {
    sym = this.get_symbol(i);
    sym = sym.replace(/\\/g, "\\\\");
    sym = sym.replace(/\(/g, "\\(");
    sym = sym.replace(/\)/g, "\\)");
    rgb = this.get_rgb(i);
    out += " (" + sym + ") [" + rgb.red.toFixed(4) + " " + rgb.green.toFixed(4) + " " + rgb.blue.toFixed(4) + "]\n";
  }
  out += ">> def\n";
  out += "/mutedColourDict <<\n";
  for (i = 0; i < this.ncore; i++) {
    sym = this.get_symbol(i);
    sym = sym.replace(/\\/g, "\\\\");
    sym = sym.replace(/\(/g, "\\(");
    sym = sym.replace(/\)/g, "\\)");
    rgb = Alphabet.lighten_colour(this.get_rgb(i));
    out += " (" + sym + ") [" + rgb.red.toFixed(4) + " " + rgb.green.toFixed(4) + " " + rgb.blue.toFixed(4) + "]\n";
  }
  out += ">> def\n";
  return out;
};

// return the alphabet name or a list of primary symbols
Alphabet.prototype.toString = function() {
  "use strict";
  if (this.name != null) {
    return this.name;
  } else {
    return this.get_symbols();
  }
};

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Helper functions
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

// Convert a colour specified in RGB colourspace values into LAB colourspace
Alphabet.rgb2lab = function(rgb) {
  "use strict";
  var xyzHelper, labHelper;
  // XYZ helper
  xyzHelper = function(value) {
    if (value > 0.0445) {
      value = (value + 0.055) / 1.055;
      value = Math.pow(value, 2.4);
    } else {
      value /= 12.92;
    }
    value *= 100;
    return value;
  };
  // lab helper
  labHelper = function(value) {
    if (value > 0.008856) {
      value = Math.pow(value, 1.0 / 3.0);
    } else {
      value = (7.787 * value) + (16.0 / 116.0);
    }
    return value;
  };
  // convert into XYZ colourspace
  var c1, c2, c3;
  if (typeof rgb == "number") {
    c1 = xyzHelper(((rgb >> 16) & 0xFF) / 255.0);
    c2 = xyzHelper(((rgb >> 8) & 0xFF) / 255.0);
    c3 = xyzHelper((rgb & 0xFF) / 255.0);
  } else {
    c1 = xyzHelper(rgb.red);
    c2 = xyzHelper(rgb.green);
    c3 = xyzHelper(rgb.blue);
  }
  var x = (c1 * 0.4124) + (c2 * 0.3576) + (c3 * 0.1805);
  var y = (c1 * 0.2126) + (c2 * 0.7152) + (c3 * 0.0722);
  var z = (c1 * 0.0193) + (c2 * 0.1192) + (c3 * 0.9505);
  // convert into Lab colourspace
  c1 = labHelper(x / 95.047);
  c2 = labHelper(y / 100.0);
  c3 = labHelper(z / 108.883);
  var l = (116.0 * c2) - 16;
  var a = 500.0 * (c1 - c2);
  var b = 200.0 * (c2 - c3);
  return {"l": l, "a": a, "b": b};
};

// Convert a colour specified in HSV colourspace into RGB colourspace
Alphabet.hsv2rgb = function(hue, sat, value, output_object) {
  // achromatic (grey)
  var r = value;
  var g = value;
  var b = value;
  if (sat != 0) {
    var h = hue / 60.0;
    var i = Math.floor(h);
    var f = h - i;
    var p = value * (1.0 - sat);
    var q = value * (1.0 - (sat * f));
    var t = value * (1.0 - (sat * (1.0 - f)));
    if (i == 0) {
      r = value;
      g = t;
      b = p;
    } else if (i == 1) {
      r = q;
      g = value;
      b = p;
    } else if (i == 2) {
      r = p;
      g = value;
      b = t;
    } else if (i == 3) {
      r = p;
      g = q;
      b = value;
    } else if (i == 4) {
      r = t;
      g = p;
      b = value;
    } else {
      r = value;
      g = p;
      b = q;
    }
  }
  if (output_object) {
    return {"red": r, "green": g, "blue": b};
  } else {
    return (Math.floor(r * 255) << 15) | (Math.floor(g * 255) << 8) | (Math.floor(b * 255));
  }
};

// Calculate a distance score between two colours in LAB colourspace
Alphabet.lab_dist = function(lab1, lab2) {
  var c1 = Math.sqrt((lab1.l * lab1.l) + (lab1.a * lab1.a));
  var c2 = Math.sqrt((lab2.l * lab2.l) + (lab2.a * lab2.a));
  var dc = c1 - c2;
  var dl = lab1.l - lab2.l;
  var da = lab1.a - lab2.a;
  var db = lab1.b - lab2.b;
  // we don't want NaN due to rounding errors so fudge things a bit...
  var dh = 0;
  var dh_squared = (da * da) + (db * db) - (dc * dc);
  if (dh_squared > 0) {
    dh = Math.sqrt(dh_squared);
  }
  var first = dl;
  var second = dc / (1.0 + (0.045 * c1));
  var third = dh / (1.0 + (0.015 * c1));
  return Math.sqrt((first * first) + (second * second) + (third * third));
};

// convert an RGB value into a HSL value
Alphabet.rgb2hsl = function(rgb) {
  "use strict";
  var min, max, delta, h, s, l, r, g, b;
  if (typeof rgb == "number") {
    r = ((rgb >> 16) & 0xFF) / 255.0;
    g = ((rgb >> 8) & 0xFF) / 255.0;
    b = (rgb & 0xFF) / 255.0;
  } else {
    r = rgb.red;
    g = rgb.green;
    b = rgb.blue;
  }
  min = Math.min(r, g, b);
  max = Math.max(r, g, b);
  delta = max - min;
  l = min + (delta / 2);
  if (max == min) {
    h = 0; // achromatic (grayscale)
    s = 0;
  } else {
    if (l > 0.5) {
      s = delta / (2 - max - min);
    } else {
      s = delta / (max + min);
    }
    if (max == r) {
      h = (g - b) / delta;
      if (g < b) h += 6;
    } else if (max == g) {
      h = ((b - r) / delta) + 2;
    } else {
      h = ((r - g) / delta) + 4;
    }
    h /= 6;
  }
  return {"h": h, "s": s, "l": l};
};

// convert a HSL value into an RGB value
Alphabet.hsl2rgb = function(hsl, output_object) {
  "use strict";
  function _hue(p, q, t) {
    "use strict";
    if (t < 0) t += 1;
    else if (t > 1) t -= 1;
    if (t < (1.0 / 6.0)) {
      return p + ((q - p) * 6.0 * t);
    } else if (t < 0.5) {
      return q;
    } else if (t < (2.0 / 3.0)) {
      return p + ((q - p) * ((2.0 / 3.0) - t) * 6.0);
    } else {
      return p;
    }
  }
  var r, g, b, p, q;
  if (hsl.s == 0) {
    // achromatic (grayscale)
    r = hsl.l;
    g = hsl.l;
    b = hsl.l;
  } else {
    if (hsl.l < 0.5) {
      q = hsl.l * (1 + hsl.s);
    } else {
      q = hsl.l + hsl.s - (hsl.l * hsl.s);
    }
    p = (2 * hsl.l) - q;
    r = _hue(p, q, hsl.h + (1.0 / 3.0));
    g = _hue(p, q, hsl.h);
    b = _hue(p, q, hsl.h - (1.0 / 3.0));
  }
  if (output_object) {
    return {"red": r, "green": g, "blue": b};
  } else {
    return (Math.floor(r * 255) << 15) | (Math.floor(g * 255) << 8) | (Math.floor(b * 255));
  }
};

Alphabet.lighten_colour = function(rgb) {
  "use strict";
  var hsl = Alphabet.rgb2hsl(rgb);
  hsl.l += (1.0 - hsl.l) * 2 / 3;
  return Alphabet.hsl2rgb(hsl, typeof rgb != "number");
};

//======================================================================
// end Alphabet object
//======================================================================

//======================================================================
// start StandardAlphabet object
//======================================================================

// an extension of the alphabet object to support some additional fields 
// only present in standard alphabets.
var StandardAlphabet = function(enum_code, enum_name, alphabet_data) {
  Alphabet.apply(this, [alphabet_data]);
  this.enum_code = enum_code;
  this.enum_name = enum_name;
};
StandardAlphabet.prototype = Alphabet.prototype;
StandardAlphabet.prototype.constructor = StandardAlphabet;

// A unique code for this standard alphabet.
// This code will be a power of 2 to enable creation of bitsets for
// a selection of standard alphabets.
StandardAlphabet.prototype.get_code = function() {
  return this.enum_code;
};

// A unique name for this standard alphabet.
// this name will be all upper case and the same as the property that
// refers to this alphabet in the AlphStd collection.
StandardAlphabet.prototype.get_enum = function() {
  return this.enum_name;
};

//======================================================================
// end StandardAlphabet object
//======================================================================

// A collection of standard alphabets.
var AlphStd = {
  RNA: new StandardAlphabet(1, "RNA", {
    "name": "RNA",
    "like": "RNA",
    "ncore": 4,
    "symbols": [
      {"symbol": "A", "name": "Adenine", "colour": "CC0000"},
      {"symbol": "C", "name": "Cytosine", "colour": "0000CC"},
      {"symbol": "G", "name": "Guanine", "colour": "FFB300"},
      {"symbol": "U", "name": "Uracil", "colour": "008000",
        "aliases": "T"},
      {"symbol": "N", "name": "Any base", "equals": "ACGU", "aliases": "X."},
      {"symbol": "V", "name": "Not U", "equals": "ACG"},
      {"symbol": "H", "name": "Not G", "equals": "ACU"},
      {"symbol": "D", "name": "Not C", "equals": "AGU"},
      {"symbol": "B", "name": "Not A", "equals": "CGU"},
      {"symbol": "M", "name": "Amino", "equals": "AC"},
      {"symbol": "R", "name": "Purine", "equals": "AG"},
      {"symbol": "W", "name": "Weak", "equals": "AU"}, 
      {"symbol": "S", "name": "Strong", "equals": "CG"},
      {"symbol": "Y", "name": "Pyrimidine", "equals": "CU"},
      {"symbol": "K", "name": "Keto", "equals": "GU"}
    ]
  }), 
  DNA: new StandardAlphabet(2, "DNA", {
    "name": "DNA",
    "like": "DNA",
    "ncore": 4,
    "symbols": [
      {"symbol": "A", "name": "Adenine", "colour": "CC0000", "complement": "T"},
      {"symbol": "C", "name": "Cytosine", "colour": "0000CC", "complement": "G"},
      {"symbol": "G", "name": "Guanine", "colour": "FFB300", "complement": "C"},
      {"symbol": "T", "name": "Thymine", "colour": "008000", "complement": "A",
        "aliases": "U"},
      {"symbol": "N", "name": "Any base", "equals": "ACGT", "aliases": "X."},
      {"symbol": "V", "name": "Not T", "equals": "ACG"},
      {"symbol": "H", "name": "Not G", "equals": "ACT"},
      {"symbol": "D", "name": "Not C", "equals": "AGT"},
      {"symbol": "B", "name": "Not A", "equals": "CGT"},
      {"symbol": "M", "name": "Amino", "equals": "AC"},
      {"symbol": "R", "name": "Purine", "equals": "AG"},
      {"symbol": "W", "name": "Weak", "equals": "AT"}, 
      {"symbol": "S", "name": "Strong", "equals": "CG"},
      {"symbol": "Y", "name": "Pyrimidine", "equals": "CT"},
      {"symbol": "K", "name": "Keto", "equals": "GT"}
    ]
  }), 
  PROTEIN: new StandardAlphabet(4, "PROTEIN", {
    "name": "Protein",
    "like": "PROTEIN",
    "ncore": 20,
    "symbols": [
      {"symbol": "A", "name": "Alanine", "colour": "0000CC"},
      {"symbol": "C", "name": "Cysteine", "colour": "0000CC"},
      {"symbol": "D", "name": "Aspartic acid", "colour": "FF00FF"},
      {"symbol": "E", "name": "Glutamic acid", "colour": "FF00FF"},
      {"symbol": "F", "name": "Phenylalanine", "colour": "0000CC"},
      {"symbol": "G", "name": "Glycine", "colour": "FFB300"},
      {"symbol": "H", "name": "Histidine", "colour": "FFCCCC"},
      {"symbol": "I", "name": "Isoleucine", "colour": "0000CC"},
      {"symbol": "K", "name": "Lysine", "colour": "CC0000"},
      {"symbol": "L", "name": "Leucine", "colour": "0000CC"},
      {"symbol": "M", "name": "Methionine", "colour": "0000CC"},
      {"symbol": "N", "name": "Asparagine", "colour": "008000"},
      {"symbol": "P", "name": "Proline", "colour": "FFFF00"},
      {"symbol": "Q", "name": "Glutamine", "colour": "008000"},
      {"symbol": "R", "name": "Arginine", "colour": "CC0000"},
      {"symbol": "S", "name": "Serine", "colour": "008000"},
      {"symbol": "T", "name": "Threonine", "colour": "008000"},
      {"symbol": "V", "name": "Valine", "colour": "0000CC"},
      {"symbol": "W", "name": "Tryptophan", "colour": "0000CC"},
      {"symbol": "Y", "name": "Tyrosine", "colour": "33E6CC"},
      {"symbol": "X", "name": "Any amino acid", "equals": "ACDEFGHIKLMNPQRSTVWY", "aliases": "*."},
      {"symbol": "B", "name": "Asparagine or Aspartic acid", "equals": "DN"}, 
      {"symbol": "Z", "name": "Glutamine or Glutamic acid", "equals": "EQ"}, 
      {"symbol": "J", "name": "Leucine or Isoleucine", "equals": "IL"}
    ]
  })
};

//======================================================================
// start Symbol object
//======================================================================
var Symbol = function(alph_index, scale, alphabet) {
  "use strict";
  //variable prototype
  this.symbol = alphabet.get_symbol(alph_index);
  this.scale = scale;
  this.colour = alphabet.get_colour(alph_index);
};

Symbol.prototype.get_symbol = function() {
  "use strict";
  return this.symbol;
};

Symbol.prototype.get_scale = function() {
  "use strict";
  return this.scale;
};

Symbol.prototype.get_colour = function() {
  "use strict";
  return this.colour;
};

Symbol.prototype.toString = function() {
  "use strict";
  return this.symbol + " " + (Math.round(this.scale*1000)/10) + "%";
};

function compare_symbol(sym1, sym2) {
  "use strict";
  if (sym1.get_scale() < sym2.get_scale()) {
    return -1;
  } else if (sym1.get_scale() > sym2.get_scale()) {
    return 1;
  } else {
    return 0;
  }
}
//======================================================================
// end Symbol object
//======================================================================

//======================================================================
// start Pspm object
//======================================================================
var Pspm = function(matrix, name, ltrim, rtrim, nsites, evalue, pssm, alt) {
  "use strict";
  var row, col, data, row_sum, delta, evalue_re;
  if (typeof name !== "string") {
    name = "";
  }
  this.name = name;
  //construct
  if (matrix instanceof Pspm) {
    // copy constructor
    this.alph_length = matrix.alph_length;
    this.motif_length = matrix.motif_length;
    this.name = matrix.name;
    this.alt = matrix.alt;
    this.nsites = matrix.nsites;
    this.evalue = matrix.evalue;
    this.ltrim = matrix.ltrim;
    this.rtrim = matrix.rtrim;
    this.pspm = [];
    for (row = 0; row < matrix.motif_length; row++) {
      this.pspm[row] = [];
      for (col = 0; col < matrix.alph_length; col++) {
        this.pspm[row][col] = matrix.pspm[row][col];
      }
    }
    if (matrix.pssm != null) {
      this.pssm = [];
      for (row = 0; row < matrix.motif_length; row++) {
        this.pspm[row] = [];
        for (col = 0; col < matrix.alph_length; col++) {
          this.pssm[row][col] = matrix.pssm[row][col];
        }
      }
    }
  } else {
    // check parameters
    if (ltrim == null) {
      ltrim = 0;
    } else if (typeof ltrim !== "number" || ltrim % 1 !== 0 || ltrim < 0) {
      throw new Error("ltrim must be a non-negative integer, got: " + ltrim);
    }
    if (rtrim == null) {
      rtrim = 0;
    } else if (typeof rtrim !== "number" || rtrim % 1 !== 0 || rtrim < 0) {
      throw new Error("rtrim must be a non-negative integer, got: " + rtrim);
    }
    if (nsites != null) {
      if (typeof nsites !== "number" || nsites < 0) {
        throw new Error("nsites must be a positive number, got: " + nsites);
      } else if (nsites == 0) {
        nsites = null;
      }
    }
    if (evalue != null) {
      if (typeof evalue === "number") {
        if (evalue < 0) {
          throw new Error("evalue must be a non-negative number, got: " + evalue);
        }
      } else if (typeof evalue === "string") {
        evalue_re = /^((?:[+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)|inf)$/;
        if (!evalue_re.test(evalue)) {
          throw new Error("evalue must be a non-negative number, got: " + evalue);
        }
      } else {
        throw new Error("evalue must be a non-negative number, got: " + evalue);
      }
    }
    // set properties
    this.name = name;
    this.alt = alt;
    this.nsites = nsites;
    this.evalue = evalue;
    this.ltrim = ltrim;
    this.rtrim = rtrim;
    if (typeof matrix === "string") {
      // string constructor
      data = parse_pspm_string(matrix);
      this.alph_length = data["alph_length"];
      this.motif_length = data["motif_length"];
      this.pspm = data["pspm"];
      if (this.evalue == null) {
        if (data["evalue"] != null) {
          this.evalue = data["evalue"];
        } else {
          this.evalue = 0;
        }
      }
      if (this.nsites == null) {
        if (typeof data["nsites"] === "number") {
          this.nsites = data["nsites"];
        } else {
          this.nsites = 20;
        }
      }
    } else {
      // assume pspm is a nested array
      this.motif_length = matrix.length;
      this.alph_length = (matrix.length > 0 ? matrix[0].length : 0);
      if (this.nsites == null) {
        this.nsites = 20;
      }
      if (this.evalue == null) {
        this.evalue = 0;
      }
      this.pspm = [];
      // copy pspm and check
      for (row = 0; row < this.motif_length; row++) {
        if (this.alph_length != matrix[row].length) {
          throw new Error("COLUMN_MISMATCH");
        }
        this.pspm[row] = [];
        row_sum = 0;
        for (col = 0; col < this.alph_length; col++) {
          this.pspm[row][col] = matrix[row][col];
          row_sum += this.pspm[row][col];
        }
        delta = 0.1;
        if (isNaN(row_sum) || (row_sum > 1 && (row_sum - 1) > delta) || 
            (row_sum < 1 && (1 - row_sum) > delta)) {
          throw new Error("INVALID_SUM");
        }
      }
      // copy pssm
      if (pssm != null) {
        this.pssm = [];
        for (row = 0; row < this.motif_length; row++) {
          this.pssm[row] = [];
          for (col = 0; col < this.alph_length; col++) {
            this.pssm[row][col] = pssm[row][col];
          }
        }
      }
    }
  }
};

Pspm.prototype.copy = function() {
  "use strict";
  return new Pspm(this);
};

Pspm.prototype.reverse = function() {
  "use strict";
  var x, y, temp, temp_trim;
  //reverse
  x = 0;
  y = this.motif_length-1;
  while (x < y) {
    temp = this.pspm[x];
    this.pspm[x] = this.pspm[y];
    this.pspm[y] = temp;
    x++;
    y--;
  }
  // reverse pssm (if defined)
  if (typeof this.pssm !== "undefined") {
    //reverse
    x = 0;
    y = this.motif_length-1;
    while (x < y) {
      temp = this.pssm[x];
      this.pspm[x] = this.pssm[y];
      this.pssm[y] = temp;
      x++;
      y--;
    }
  }
  //swap triming
  temp_trim = this.ltrim;
  this.ltrim = this.rtrim;
  this.rtrim = temp_trim;
  return this; //allow function chaining...
};

Pspm.prototype.reverse_complement = function(alphabet) {
  "use strict";
  var x, y, temp, i, row, c, temp_trim;
  if (this.alph_length != alphabet.get_size_core()) {
    throw new Error("The alphabet size does not match the size of the pspm.");
  }
  if (!alphabet.has_complement()) {
    throw new Error("The specified alphabet can not be complemented.");
  }
  // reverse motif
  this.reverse();
  //complement
  for (x = 0; x < this.motif_length; x++) {
    row = this.pspm[x];
    for (i = 0; i < row.length; i++) {
      c = alphabet.get_complement(i);
      if (c < i) continue;
      temp = row[i];
      row[i] = row[c];
      row[c] = temp;
    }
  }
  // complement pssm (if defined)
  if (typeof this.pssm !== "undefined") {
    //complement
    for (x = 0; x < this.motif_length; x++) {
      row = this.pssm[x];
      for (i = 0; i < row.length; i++) {
        c = alphabet.get_complement(i);
        if (c < i) continue;
        temp = row[i];
        row[i] = row[c];
        row[c] = temp;
      }
    }
  }
  return this; //allow function chaining...
};

Pspm.prototype.get_stack = function(position, alphabet, ssc) {
  "use strict";
  var row, stack_ic, alphabet_ic, stack, i, sym;
  if (this.alph_length != alphabet.get_size_core()) {
    throw new Error("The alphabet size does not match the size of the pspm.");
  }
  row = this.pspm[position];
  stack_ic = this.get_stack_ic(position, alphabet);
  if (ssc) stack_ic -= this.get_error(alphabet);
  alphabet_ic = alphabet.get_ic();
  stack = [];
  for (i = 0; i < this.alph_length; i++) {
    sym = new Symbol(i, row[i]*stack_ic/alphabet_ic, alphabet);
    if (sym.get_scale() <= 0) {
      continue;
    }
    stack.push(sym);
  }
  stack.sort(compare_symbol);
  return stack;
};

Pspm.prototype.get_stack_ic = function(position, alphabet) {
  "use strict";
  var row, H, i;
  if (this.alph_length != alphabet.get_size_core()) {
    throw new Error("The alphabet size does not match the size fo the pspm.");
  }
  row = this.pspm[position];
  H = 0;
  for (i = 0; i < this.alph_length; i++) {
    if (row[i] === 0) {
      continue;
    }
    H -= (row[i] * (Math.log(row[i]) / Math.LN2));
  }
  return alphabet.get_ic() - H;
};

Pspm.prototype.get_error = function(alphabet) {
  "use strict";
  if (this.nsites === 0) {
    return 0;
  }
  return (alphabet.get_size_core()-1) / (2 * Math.LN2 * this.nsites);
};

Pspm.prototype.get_motif_length = function() {
  "use strict";
  return this.motif_length;
};

Pspm.prototype.get_alph_length = function() {
  "use strict";
  return this.alph_length;
};

Pspm.prototype.get_left_trim = function() {
  "use strict";
  return this.ltrim;
};

Pspm.prototype.get_right_trim = function() {
  "use strict";
  return this.rtrim;
};

Pspm.prototype.as_best_match = function(alphabet) {
  "use strict";
  var match, odds, best_odds, best_index;
  var i, j;
  match = "";
  for (i = 0; i < this.motif_length; i++) {
    best_index = 0;
    best_odds = this.pspm[i][0] / alphabet.get_bg_freq(0);
    for (j = 1; j < this.alph_length; j++) {
      odds = this.pspm[i][j] / alphabet.get_bg_freq(j);
      if (odds > best_odds) {
        best_odds = odds;
        best_index = j;
      }
    }
    match += alphabet.get_symbol(best_index);
  }
  return match;
};

Pspm.prototype.as_count_matrix = function() {
  "use strict";
  var count, count_text, text;
  var i, j;
  text = "";
  for (i = 0; i < this.motif_length; i++) {
    if (i !== 0) {
      text += "\n";
    }
    for (j = 0; j < this.alph_length; j++) {
      if (j !== 0) {
        text += " ";
      }
      count = Math.round(this.nsites * this.pspm[i][j]);
      count_text = "" + count;
      // pad up to length of 4
      if (count_text.length < 4) {
        text += (new Array(5 - count_text.length)).join(" ") + count_text;
      } else {
        text += count_text;
      }
    }
  }
  return text; 
};

Pspm.prototype.as_probability_matrix = function() {
  "use strict";
  var text;
  var i, j;
  text = "";
  for (i = 0; i < this.motif_length; i++) {
    if (i !== 0) {
      text += "\n";
    }
    for (j = 0; j < this.alph_length; j++) {
      if (j !== 0) {
        text += " ";
      }
      text += this.pspm[i][j].toFixed(6);
    }
  }
  return text; 
};

Pspm.prototype.as_score_matrix = function(alphabet, pseudo) {
  "use strict";
  var me, score, out, row, col, score_text;
  me = this;
  if (typeof this.pssm === "undefined") {
    if (!(typeof alphabet === "object" && alphabet != null && alphabet instanceof Alphabet)) {
      throw new Error("The alphabet is required to generate the pssm.");
    }
    if (typeof pseudo === "undefined") {
      pseudo = 0.01;
    } else if (typeof pseudo !== "number" || pseudo < 0) {
      throw new Error("Expected positive number for pseudocount");
    }
    score = function(row, col) {
      "use strict";
      var p, bg, p2;
      p = me.pspm[row][col];
      bg = alphabet.get_bg_freq(col);
      p2 = (p * me.nsites + bg * pseudo) / (me.nsites + pseudo);
      return (p2 > 0 ? Math.round((Math.log(p2 / bg) / Math.LN2) * 100) : -10000);
    };
  } else {
    score = function(row, col) {
      "use strict";
      return me.pssm[row][col];
    };
  }
  out = "";
  for (row = 0; row < this.motif_length; row++) {
    for (col = 0; col < this.alph_length; col++) {
      if (col !== 0) {
        out += " ";
      }
      score_text = "" + score(row, col);
      // pad out to 6 characters
      if (score_text.length < 6) {
        out += (new Array(7 - score_text.length)).join(" ") + score_text;
      } else {
        out += score_text;
      }
    }
    out += "\n";
  }
  return out;
}

Pspm.prototype.as_pspm = function() {
  "use strict";
  return "letter-probability matrix: alength= " + this.alph_length + 
      " w= " + this.motif_length + " nsites= " + this.nsites + 
      " E= " + (typeof this.evalue === "number" ? 
          this.evalue.toExponential() : this.evalue) + "\n" +
      this.as_probability_matrix();
};

Pspm.prototype.as_pssm = function(alphabet, pseudo) {
  "use strict";
  return "log-odds matrix: alength= " + this.alph_length + 
      " w= " + this.motif_length + 
      " E= " + (typeof this.evalue == "number" ?
          this.evalue.toExponential() : this.evalue) + "\n" +
      this.as_score_matrix(alphabet, pseudo);
};

Pspm.prototype.as_meme = function(options) {
  var with_header, with_pspm, with_pssm, version, alphabet, bg_source, pseudocount, strands;
  var out, alen, i;
  // get the options
  if (typeof options !== "object" || options === null) {
    options = {};
  }
  with_header = (typeof options["with_header"] === "boolean" ? options["with_header"] : false);
  with_pspm = (typeof options["with_pspm"] === "boolean" ? options["with_pspm"] : false);
  with_pssm = (typeof options["with_pssm"] === "boolean" ? options["with_pssm"] : false);
  if (!with_pspm && !with_pssm) with_pspm = true;
  if (with_header) {
    if (typeof options["version"] === "string" && /^\d+(?:\.\d+){0,2}$/.test(options["version"])) {
      version = options["version"];
    } else if (typeof options["version"] === "number") {
      version = options["version"].toFixed(0);
    } else {
      version = "4";
    }
    if (typeof options["strands"] === "number" && options["strands"] === 1) {
      strands = 1;
    } else {
      strands = 2;
    }
    if (typeof options["bg_source"] === "string") {
      bg_source = options["bg_source"];
    } else {
      bg_source = "unknown source";
    }
    if (typeof options["alphabet"] === "object" && options["alphabet"] != null
        && options["alphabet"] instanceof Alphabet) {
      alphabet = options["alphabet"];
    } else {
      throw new Error("The alphabet is required to generate the header.");
    }
  }
  // now create the output
  out = "";
  if (with_header) {
    out = "MEME version " + version + "\n\n";
    out += alphabet.as_meme() + "\n";
    if (alphabet.has_complement()) { // assume DNA has both strands unless otherwise specified
      out += "strands: " + (strands === 1 ? "+" : "+ -") + "\n\n";
    }
    out += "Background letter frequencies (from " + bg_source + "):\n";
    alen = alphabet.get_size_core();
    for (i = 0; i < alen; i++) {
      if (i !== 0) {
        if (i % 9 === 0) { // maximum of nine entries per line
          out += "\n";
        } else {
          out += " ";
        }
      }
      out += alphabet.get_symbol(i) + " " + alphabet.get_bg_freq(i).toFixed(3);
    }
  }
  out += "\n\n";
  out += "MOTIF " + this.name + (this.alt == null ? "" : " " + this.alt);
  if (with_pssm) {
    out += "\n\n";
    out += this.as_pssm(options["alphabet"], options["pseudocount"]);
  }
  if (with_pspm) {
    out += "\n\n";
    out += this.as_pspm();
  }
  return out;
}

Pspm.prototype.toString = function() {
  "use strict";
  var str, i, row;
  str = "";
  for (i = 0; i < this.pspm.length; i++) {
    row = this.pspm[i];
    str += row.join("\t") + "\n";
  }
  return str;
};

function parse_pspm_properties(str) {
  "use strict";
  var parts, i, eqpos, before, after, properties, prop, num, num_re;
  num_re = /^((?:[+]?[0-9]*\.?[0-9]+(?:[eE][-+]?[0-9]+)?)|inf)$/;
  parts = trim(str).split(/\s+/);
  // split up words containing =
  for (i = 0; i < parts.length;) {
    eqpos = parts[i].indexOf("=");
    if (eqpos != -1) {
      before = parts[i].substr(0, eqpos);
      after = parts[i].substr(eqpos+1);
      if (before.length > 0 && after.length > 0) {
        parts.splice(i, 1, before, "=", after);
        i += 3;
      } else if (before.length > 0) {
        parts.splice(i, 1, before, "=");
        i += 2;
      } else if (after.length > 0) {
        parts.splice(i, 1, "=", after);
        i += 2;
      } else {
        parts.splice(i, 1, "=");
        i++;
      }
    } else {
      i++;
    }
  }
  properties = {};
  for (i = 0; i < parts.length; i += 3) {
    if (parts.length - i < 3) {
      throw new Error("Expected PSPM property was incomplete. "+
          "Remaing parts are: " + parts.slice(i).join(" "));
    }
    if (parts[i+1] !== "=") {
      throw new Error("Expected '=' in PSPM property between key and " +
          "value but got " + parts[i+1]); 
    }
    prop = parts[i].toLowerCase();
    num = parts[i+2];
    if (!num_re.test(num)) {
      throw new Error("Expected numeric value for PSPM property '" + 
          prop + "' but got '" + num + "'");
    }
    properties[prop] = num;
  }
  return properties;
}

function parse_pspm_string(pspm_string) {
  "use strict";
  var header_re, lines, first_line, line_num, col_num, alph_length, 
      motif_length, nsites, evalue, pspm, i, line, match, props, parts,
      j, prob;
  header_re = /^letter-probability\s+matrix:(.*)$/i;
  lines = pspm_string.split(/\n/);
  first_line = true;
  line_num = 0;
  col_num = 0;
  alph_length;
  motif_length;
  nsites;
  evalue;
  pspm = [];
  for (i = 0; i < lines.length; i++) {
    line = trim(lines[i]);
    if (line.length === 0) { 
      continue;
    }
    // check the first line for a header though allow matrices without it
    if (first_line) {
      first_line = false;
      match = header_re.exec(line);
      if (match !== null) {
        props = parse_pspm_properties(match[1]);
        if (props.hasOwnProperty("alength")) {
          alph_length = parseFloat(props["alength"]);
          if (alph_length != 4 && alph_length != 20) {
            throw new Error("PSPM property alength should be 4 or 20" +
                " but got " + alph_length);
          }
        }
        if (props.hasOwnProperty("w")) {
          motif_length = parseFloat(props["w"]);
          if (motif_length % 1 !== 0 || motif_length < 1) {
            throw new Error("PSPM property w should be an integer larger " +
                "than zero but got " + motif_length);
          }
        }
        if (props.hasOwnProperty("nsites")) {
          nsites = parseFloat(props["nsites"]);
          if (nsites <= 0) {
            throw new Error("PSPM property nsites should be larger than " +
                "zero but got " + nsites);
          }
        }
        if (props.hasOwnProperty("e")) {
          evalue = props["e"];
          if (evalue < 0) {
            throw new Error("PSPM property evalue should be " +
                "non-negative but got " + evalue);
          }
        }
        continue;
      }
    }
    pspm[line_num] = [];
    col_num = 0;
    parts = line.split(/\s+/);
    for (j = 0; j < parts.length; j++) {
      prob = parseFloat(parts[j]);
      if (prob != parts[j] || prob < 0 || prob > 1) {
        throw new Error("Expected probability but got '" + parts[j] + "'"); 
      }
      pspm[line_num][col_num] = prob;
      col_num++;
    }
    line_num++;
  }
  if (typeof motif_length === "number") {
    if (pspm.length != motif_length) {
      throw new Error("Expected PSPM to have a motif length of " + 
          motif_length + " but it was actually " + pspm.length);
    }
  } else {
    motif_length = pspm.length;
  }
  if (typeof alph_length !== "number") {
    alph_length = pspm[0].length;
    if (alph_length != 4 && alph_length != 20) {
      throw new Error("Expected length of first row in the PSPM to be " +
          "either 4 or 20 but got " + alph_length);
    }
  }
  for (i = 0; i < pspm.length; i++) {
    if (pspm[i].length != alph_length) {
      throw new Error("Expected PSPM row " + i + " to have a length of " + 
          alph_length + " but the length was " + pspm[i].length);
    }
  }
  return {"pspm": pspm, "motif_length": motif_length, 
    "alph_length": alph_length, "nsites": nsites, "evalue": evalue};
}
//======================================================================
// end Pspm object
//======================================================================

//======================================================================
// start Logo object
//======================================================================

var Logo = function(alphabet, options) {
  "use strict";
  this.alphabet = alphabet;
  this.fine_text = "";
  this.x_axis = 1;
  this.y_axis = true;
  this.xlate_nsyms = 1;
  this.xlate_start = null;
  this.xlate_end = null;
  this.pspm_list = [];
  this.pspm_column = [];
  this.rows = 0;
  this.columns = 0;
  if (typeof options === "string") {
    // the old method signature had fine_text here so we support that
    this.fine_text = options;
  } else if (typeof options === "object" && options != null) {
    this.fine_text = (typeof options.fine_text === "string" ? options.fine_text : "");
    this.x_axis = (typeof options.x_axis === "boolean" ? (options.x_axis ? 1 : 0) : 1);
    if (options.x_axis_hidden != null && options.x_axis_hidden) this.x_axis = -1;
    this.y_axis = (typeof options.y_axis === "boolean" ? options.y_axis : true);
    this.xlate_nsyms = (typeof options.xlate_nsyms === "number" ? options.xlate_nsyms : this.xlate_nsyms);
    this.xlate_start = (typeof options.xlate_start === "number" ? options.xlate_start : this.xlate_start);
    this.xlate_end = (typeof options.xlate_end === "number" ? options.xlate_end : this.xlate_end);
  }
};

Logo.prototype.add_pspm = function(pspm, column) {
  "use strict";
  var col;
  if (typeof column === "undefined") {
    column = 0;
  } else if (column < 0) {
    throw new Error("Column index out of bounds.");
  }
  this.pspm_list[this.rows] = pspm;
  this.pspm_column[this.rows] = column;
  this.rows++;
  col = column + pspm.get_motif_length();
  if (col > this.columns) {
    this.columns = col;
  }
};

Logo.prototype.get_columns = function() {
  "use strict";
  return this.columns;
};

Logo.prototype.get_xlate_nsyms = function() {
  "use strict";
  return this.xlate_nsyms;
};

Logo.prototype.get_xlate_start = function() {
  "use strict";
  return (this.xlate_start != null ? this.xlate_start : 0);
};

Logo.prototype.get_xlate_end = function() {
  "use strict";
  return (this.xlate_end != null ? this.xlate_end : this.columns * this.xlate_nsyms);
};

Logo.prototype.get_xlate_columns = function() {
  "use strict";
  return this.get_xlate_end() - this.get_xlate_start();
};

Logo.prototype.get_rows = function() {
  "use strict";
  return this.rows;
};

Logo.prototype.get_pspm = function(row_index) {
  "use strict";
  if (row_index < 0 || row_index >= this.rows) {
    throw new Error("INDEX_OUT_OF_BOUNDS");
  }
  return this.pspm_list[row_index];
};

Logo.prototype.get_offset = function(row_index) {
  "use strict";
  if (row_index < 0 || row_index >= this.rows) {
    throw new Error("INDEX_OUT_OF_BOUNDS");
  }
  return this.pspm_column[row_index];
};

Logo.prototype._as_eps_data = function(ssc, errbars) {
  var i, j, pos, stack_pos, pspm, stack, sym, out;
  out = "";
  for (i = 0; i < this.rows; i++) {
    out += "\nStartLine\n";
    // Indent
    for (j = 0; j < this.pspm_column[i]; j++) {
      out += "() startstack\nendstack\n\n";
    }
    pspm = this.pspm_list[i];
    if (pspm.get_left_trim() > 0) {
      out += "MuteColour\nDrawTrimEdge\n" + pspm.get_left_trim() + " DrawTrimBg\n";
    }
    for (pos = 0; pos < pspm.get_motif_length(); pos++) {
      if (pos != 0 && pos == pspm.get_left_trim()) { // enable full colour
        out += "DrawTrimEdge\nRestoreColour\n";
      } else if (pos == (pspm.get_motif_length() - pspm.get_right_trim())) {
        out += "MuteColour\n" + pspm.get_right_trim() + " DrawTrimBg\n";
      }
      out += "(" + (pos + 1) + ") startstack\n";
      stack = pspm.get_stack(pos, this.alphabet, ssc);
      for (stack_pos = 0; stack_pos < stack.length; stack_pos++) {
        sym = stack[stack_pos];
        out += " " + (sym.get_scale() * this.alphabet.get_ic()) + " (" + sym.get_symbol() + ") numchar\n";
      }
      if (errbars) {
        out += " " + pspm.get_error(this.alphabet) + " Ibeam\n";
      }
      out += "endstack\n\n";
    }
    if (pspm.get_right_trim() > 0 || pspm.get_left_trim() == pspm.get_motif_length()) {
      out += "RestoreColour\n";
    }
    out += "EndLine\n";
  }
  return out;
};

Logo.prototype.as_eps = function(options) {
  "use strict";
  if (this.xlate_nsyms != 1) throw new Error("Unsupported setting xlate_nsyms for EPS");
  if (this.xlate_start != null) throw new Error("Unsupported setting xlate_start for EPS");
  if (this.xlate_end != null) throw new Error("Unsupported setting xlate_end for EPS");

  var LOGOHEIGHT = 7.5; // default height of line in cm
  var cm2pts, height, width, now, ssc, errbars;
  if (typeof options === "undefined") {
    options = {};
  }
  cm2pts = 72 / 2.54;
  if (typeof options.logo_height == "number") {
    height = options.logo_height;
  } else {
    height = LOGOHEIGHT * this.rows;
  }
  if (typeof options.logo_width == "number") {
    width = options.logo_width;
  } else {
    width = this.columns + 2;
  }
  now = new Date();
  ssc = (typeof options.ssc == "boolean" ? options.ssc : false);
  errbars = (typeof options.show_error_bar == "boolean" ? options.show_error_bar : ssc);
  var values = {
    "LOGOHEIGHT": height,
    "LOGOWIDTH": width,
    "BOUNDINGHEIGHT": Math.round(height * cm2pts),
    "BOUNDINGWIDTH": Math.round(width * cm2pts),
    "LOGOLINEHEIGHT": (height / this.rows),
    "CHARSPERLINE": this.columns,
    "BARBITS": this.alphabet.get_ic(),
    "LOGOTYPE": (this.alphabet.has_complement() ? "NA" : "AA"),
    "CREATIONDATE": now.getDate() + "." + (now.getMonth() + 1) + "." + now.getFullYear() + " " + now.getHours() + ":" + now.getMinutes() + ":" + now.getSeconds(),
    "ERRORBARFRACTION": (typeof options.error_bar_fraction == "number" ? options.error_bar_fraction : 1.0),
    "TICBITS": (typeof options.ticbits == "number" ? options.ticbits : 1.0),
    "TITLE": (typeof options.title == "string" ? options.title : ""),
    "FINEPRINT": (typeof options.fineprint == "string" ? options.fineprint : this.fine_text),
    "XAXISLABEL": (typeof options.xaxislabel == "string" ? options.xaxislabel : ""),
    "YAXISLABEL": (typeof options.yaxislabel == "string" ? options.yaxislabel : "bits"),
    "SSC": ssc,
    "YAXIS": (typeof options.show_y_axis == "boolean" ? options.show_y_axis : this.y_axis),
    "SHOWENDS": (typeof options.show_ends == "boolean" ? options.show_ends : false),
    "ERRBAR": errbars,
    "OUTLINE": (typeof options.show_outline == "boolean" ? options.show_outline : false),
    "NUMBERING": (typeof options.show_numbering == "boolean" ? options.show_numbering : this.x_axis != 0),
    "SHOWINGBOX": (typeof options.show_box == "boolean" ? options.show_box : false),
    "CREATOR": (typeof options.creator == "string" ? options.creator : "motif_logo.js"),
    "FONTSIZE": (typeof options.label_font_size == "number" ? options.label_font_size : 12),
    "TITLEFONTSIZE": (typeof options.title_font_size == "number" ? options.title_font_size : 12),
    "SMALLFONTSIZE": (typeof options.small_font_size == "number" ? options.small_font_size : 6),
    "TOPMARGIN" : (typeof options.top_margin == "number" ? options.top_margin : 0.9),
    "BOTTOMMARGIN": (typeof options.bottom_margin == "number" ? options.bottom_margin : 0.9),
    "COLORDICT": this.alphabet._as_eps_dict(),
    "DATA": this._as_eps_data(ssc, errbars)
  };
  // now this requires that the script containing the template has been imported!
  return motif_logo_template(values);
};

//======================================================================
// end Logo object
//======================================================================

// calculate the exact size (in pixels) of an object drawn on the
// canvas assuming that the background of the canvas is transparent.
function canvas_bounds(ctx, cwidth, cheight) {
  "use strict";
  var data, r, c, top_line, bottom_line, left_line, right_line, 
      txt_width, txt_height;

  // extract the image data
  data = ctx.getImageData(0, 0, cwidth, cheight).data;

  // set initial values
  top_line = -1; bottom_line = -1; left_line = -1; right_line = -1;
  txt_width = 0; txt_height = 0;

  // Find the top-most line with a non-transparent pixel
  for (r = 0; r < cheight; r++) {
    for (c = 0; c < cwidth; c++) {
      if (data[r * cwidth * 4 + c * 4 + 3]) {
        top_line = r;
        break;
      }
    }
    if (top_line != -1) {
      break;
    }
  }
  
  // Only bother looking if we found at least one set pixel... 
  if (top_line != -1) {

    //find the last line with a non-transparent pixel
    for (r = cheight-1; r >= top_line; r--) {
      for(c = 0; c < cwidth; c++) {
        if(data[r * cwidth * 4 + c * 4 + 3]) {
          bottom_line = r;
          break;
        }
      }
      if (bottom_line != -1) {
        break;
      }
    }
    // calculate height
    txt_height = bottom_line - top_line + 1;

    // Find the left-most line with a non-transparent pixel
    for (c = 0; c < cwidth; c++) {
      for (r = top_line; r <= bottom_line; r++) {
        if (data[r * cwidth * 4 + c * 4 + 3]) {
          left_line = c;
          break;
        }
      }
      if (left_line != -1) {
        break;
      }
    }

    //find the right most line with a non-transparent pixel
    for (c = cwidth-1; c >= left_line; c--) {
      for(r = top_line; r <= bottom_line; r++) {
        if(data[r * cwidth * 4 + c * 4 + 3]) {
          right_line = c;
          break;
        }
      }
      if (right_line != -1) {
        break;
      }
    }
    txt_width = right_line - left_line + 1;
  }

  //return the bounds
  return {bound_top: top_line, bound_bottom: bottom_line, 
    bound_left: left_line, bound_right: right_line, width: txt_width, 
    height: txt_height};
}

//======================================================================
// start RasterizedAlphabet
//======================================================================

// Rasterize Alphabet
// 1) Measure width of text at default font for all symbols in alphabet
// 2) sort in width ascending
// 3) Drop the top and bottom 10% (designed to ignore outliers like 'W' and 'I')
// 4) Calculate the average as the maximum scaling factor (designed to stop I becoming a rectangular blob).
// 5) Assume scale of zero would result in width of zero, interpolate scale required to make perfect width font
// 6) Draw text onto temp canvas at calculated scale
// 7) Find bounds of drawn text
// 8) Paint on to another canvas at the desired height (but only scaling width to fit if larger).
var RasterizedAlphabet = function(alphabet, logo_scale, font, width) {
  "use strict";
  var default_size, safety_pad, canvas, ctx, middle, baseline, widths, sizes,
      i, sym, size, tenpercent, avg_width, scale, 
      target_width, target_height;
  //variable prototypes
  this.alphabet = alphabet;
  this.scale = logo_scale;
  this.sym_cache = {};
  this.stack_num_cache = [];
  this.scale_num_cache = [];
  // size of canvas
  default_size = 60; // size of measuring canvas
  safety_pad = 20; // pixels to pad around so we don't miss the edges
  // create a canvas to do our measuring
  canvas = document.createElement("canvas");
  if (!canvas.getContext) throw new Error("No canvas support");
  canvas.width = default_size + 2 * safety_pad;
  canvas.height = default_size + 2 * safety_pad;
  middle = Math.round(canvas.width / 2);
  baseline = Math.round(canvas.height - safety_pad);
  ctx = canvas.getContext('2d');
  if (!supports_text(ctx)) throw new Error("Canvas does not support text");
  ctx.font = font;
  ctx.textAlign = "center";
  ctx.translate(middle, baseline);
  // list of widths
  widths = [];
  sizes = [];
  //now measure each letter in the alphabet
  for (i = 0; i < alphabet.get_size_core(); ++i) {
    // reset the canvas
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    ctx.fillStyle = alphabet.get_colour(i);
    // draw the test text
    ctx.fillText(alphabet.get_symbol(i), 0, 0);
    //measure
    size = canvas_bounds(ctx, canvas.width, canvas.height);
    if (size.width === 0) throw new Error("Invisible symbol!");
    widths.push(size.width);
    sizes[i] = size;
  }
  //sort the widths
  widths.sort(function(a,b) {return a - b;});
  //drop 10% of the items off each end
  tenpercent = Math.floor(widths.length / 10);
  for (i = 0; i < tenpercent; ++i) {
    widths.pop();
    widths.shift();
  }
  //calculate average width
  avg_width = 0;
  for (i = 0; i < widths.length; ++i) {
    avg_width += widths[i];
  }
  avg_width /= widths.length;
  // calculate the target width
  target_width = width * this.scale * 2;
  // calculate scales
  for (i = 0; i < alphabet.get_size_core(); ++i) {
    sym = alphabet.get_symbol(i);
    size = sizes[i];
    // calculate scale
    scale = target_width / Math.max(avg_width, size.width);
    // estimate scaled height
    target_height = size.height * scale;
    // create an appropriately sized canvas
    canvas = document.createElement("canvas");
    canvas.width = target_width;
    canvas.height = target_height + safety_pad * 2;
    // calculate the middle
    middle = Math.round(canvas.width / 2);
    // calculate the baseline
    baseline = Math.round(canvas.height - safety_pad);
    // get the context and prepare to draw the rasterized text
    ctx = canvas.getContext('2d');
    ctx.font = font;
    ctx.fillStyle = alphabet.get_colour(i);
    ctx.textAlign = "center";
    ctx.translate(middle, baseline);
    ctx.save();
    ctx.scale(scale, scale);
    // draw the text
    ctx.fillText(sym, 0, 0);
    ctx.restore();
    this.sym_cache[sym] = {"image": canvas, "size": canvas_bounds(ctx, canvas.width, canvas.height)};
  }
};

RasterizedAlphabet.prototype.get_alphabet = function() {
  return this.alphabet;
};

RasterizedAlphabet.prototype.get_scale = function() {
  return this.scale;
};

RasterizedAlphabet.prototype.draw_stack_sym = function(ctx, letter, dx, dy, dWidth, dHeight) {
  "use strict";
  var entry, image, size;
  entry = this.sym_cache[letter];
  image = entry.image;
  size = entry.size;
  ctx.drawImage(image, 0, size.bound_top -1, image.width, size.height+1, dx, dy, dWidth, dHeight);
};

RasterizedAlphabet.prototype.draw_stack_num = function(ctx, font, stack_width, index) {
  var image, image_ctx, text_length;
  if (index >= this.stack_num_cache.length) {
    image = document.createElement("canvas");
    // measure the text
    image_ctx = image.getContext('2d');
    image_ctx.save();
    image_ctx.font = font;
    text_length = image_ctx.measureText("" + (index + 1)).width;
    image_ctx.restore();
    // resize the canvas to fit
    image.width = Math.ceil(stack_width);
    image.height = Math.ceil(text_length);
    // draw the text
    image_ctx = image.getContext('2d');
    image_ctx.translate(Math.round(stack_width / 2), 0);
    image_ctx.font = font;
    image_ctx.textBaseline = "middle";
    image_ctx.textAlign = "right";
    image_ctx.rotate(-(Math.PI / 2));
    image_ctx.fillText("" + (index + 1), 0, 0);
    this.stack_num_cache[index] = image;
  } else {
    image = this.stack_num_cache[index];
  }
  ctx.drawImage(image, 0, 0);
}

RasterizedAlphabet.prototype.draw_scale_num = function(ctx, font, num) {
  var image, image_ctx, text_size, m_length;
  if (num >= this.scale_num_cache.length) {
    image = document.createElement("canvas");
    // measure the text
    image_ctx = image.getContext('2d');
    image_ctx.font = font;
    text_size = image_ctx.measureText("" + num);
    if (text_size.actualBoundingBoxAscent && text_size.actualBoundingBoxDesent) {
      // resize the canvas to fit
      image.width = Math.ceil(text_size.width);
      image.height = Math.ceil(text_size.actualBoundingBoxAscent + text_size.actualBoundingBoxDesent);
      // draw the text
      image_ctx = image.getContext('2d');
      image_ctx.font = font;
      image_ctx.textAlign = "right";
      image_ctx.fillText("" + num, image.width, text_size.actualBoundingBoxAscent);
    } else {
      // measure width of 'm' to approximate height, we double it later anyway
      m_length = image_ctx.measureText("m").width;
      // resize the canvas to fit
      image.width = Math.ceil(text_size.width);
      image.height = Math.ceil(2 * m_length);
      // draw the text
      image_ctx = image.getContext('2d');
      image_ctx.font = font;
      image_ctx.textAlign = "right";
      image_ctx.textBaseline = "middle";
      image_ctx.fillText("" + num, image.width, m_length);
    }
    this.scale_num_cache[num] = image;
  } else {
    image = this.scale_num_cache[num];
  }
  ctx.drawImage(image, -image.width, -Math.round(image.height / 2))
}

//======================================================================
// end RasterizedAlphabet
//======================================================================

//======================================================================
// start LogoMetrics object
//======================================================================

var LogoMetrics = function(ctx, logo_columns, logo_rows, has_names, has_finetext, x_axis, y_axis) {
  "use strict";
  var i, row_height;
  //variable prototypes
  this.pad_top = (has_names ? 5 : 0);
  this.pad_left = (y_axis ? 10 : 0);
  this.pad_right = (has_finetext ? 15 : 0);
  this.pad_bottom = 0;
  this.pad_middle = 20;
  this.name_height = 14;
  this.name_font = "bold " + this.name_height + "px Times, sans-serif";
  this.name_spacer = 0;
  this.y_axis = y_axis;
  this.y_label = "bits";
  this.y_label_height = 12;
  this.y_label_font = "bold " + this.y_label_height + "px Helvetica, sans-serif";
  this.y_label_spacer = 3;
  this.y_num_height = 12;
  this.y_num_width = 0;
  this.y_num_font = "bold " + this.y_num_height + "px Helvetica, sans-serif";
  this.y_tic_width = 5;
  this.stack_pad_left = 0;
  this.stack_font = "bold 25px Helvetica, sans-serif";
  this.stack_height = 90;
  this.stack_width = 26;
  this.stacks_pad_right = 5;
  this.x_axis = x_axis;
  this.x_num_above = 2;
  this.x_num_height = 12;
  this.x_num_width = 0;
  this.x_num_font = "bold " + this.x_num_height + "px Helvetica, sans-serif";
  this.fine_txt_height = 6;
  this.fine_txt_above = 2;
  this.fine_txt_font = "normal " + this.fine_txt_height + "px Helvetica, sans-serif";
  this.letter_metrics = new Array();
  this.summed_width = 0;
  this.summed_height = 0;
  //calculate the width of the y axis numbers
  ctx.font = this.y_num_font;
  for (i = 0; i <= 2; i++) {
    this.y_num_width = Math.max(this.y_num_width, ctx.measureText("" + i).width);
  }
  //calculate the width of the x axis numbers (but they are rotated so it becomes height)
  if (x_axis == 1) {
    ctx.font = this.x_num_font;
    for (i = 1; i <= logo_columns; i++) {
      this.x_num_width = Math.max(this.x_num_width, ctx.measureText("" + i).width);
    }
  } else if (x_axis == 0) {
    this.x_num_height = 4;
    this.x_num_width = 4;
  } else {
    this.x_num_height = 0;
    this.x_num_width = 0;
  }
  
  //calculate how much vertical space we want to draw this
  //first we add the padding at the top and bottom since that's always there
  this.summed_height += this.pad_top + this.pad_bottom;
  //all except the last row have the same amount of space allocated to them
  if (logo_rows > 1) {
    row_height = this.stack_height + this.pad_middle;
    if (has_names) {
      row_height += this.name_height;
      //the label is allowed to overlap into the spacer
      row_height += Math.max(this.y_num_height/2, this.name_spacer); 
      //the label is allowed to overlap the space used by the other label
      row_height += Math.max(this.y_num_height/2, this.x_num_height + this.x_num_above); 
    } else {
      row_height += this.y_num_height/2; 
      //the label is allowed to overlap the space used by the other label
      row_height += Math.max(this.y_num_height/2, this.x_num_height + this.x_num_above); 
    }
    this.summed_height += row_height * (logo_rows - 1);
  }
  //the last row has the name and fine text below it but no padding
  this.summed_height += this.stack_height + (this.y_axis ? this.y_num_height/2 : 0);

  var fine_txt_total = (has_finetext ? this.fine_txt_height + this.fine_txt_above : 0);
  if (has_names) {
    this.summed_height += fine_txt_total + this.name_height;
    this.summed_height += Math.max((this.y_axis ? this.y_num_height/2 : 0), 
        this.x_num_height + this.x_num_above + this.name_spacer);
  } else {
    this.summed_height += Math.max((this.y_axis ? this.y_num_height/2 : 0), 
        this.x_num_height + this.x_num_above + fine_txt_total);
  }

  //calculate how much horizontal space we want to draw this
  //first add the padding at the left and right since that's always there
  this.summed_width += this.pad_left + this.pad_right;
  if (this.y_axis) {
    //add on the space for the y-axis label
    this.summed_width += this.y_label_height + this.y_label_spacer;
    //add on the space for the y-axis
    this.summed_width += this.y_num_width + this.y_tic_width;
  }
  //add on the space for the stacks
  this.summed_width += (this.stack_pad_left + this.stack_width) * logo_columns;
  //add on the padding after the stacks (an offset from the fine text)
  this.summed_width += this.stacks_pad_right;

};

//======================================================================
// end LogoMetrics object
//======================================================================

//found this trick at http://talideon.com/weblog/2005/02/detecting-broken-images-js.cfm
function image_ok(img) {
  "use strict";
  // During the onload event, IE correctly identifies any images that
  // weren't downloaded as not complete. Others should too. Gecko-based
  // browsers act like NS4 in that they report this incorrectly.
  if (!img.complete) {
    return false;
  }
  // However, they do have two very useful properties: naturalWidth and
  // naturalHeight. These give the true size of the image. If it failed
  // to load, either of these should be zero.
  if (typeof img.naturalWidth !== "undefined" && img.naturalWidth === 0) {
    return false;
  }
  // No other way of checking: assume it's ok.
  return true;
}
  
function supports_text(ctx) {
  "use strict";
  if (!ctx.fillText) {
    return false;
  }
  if (!ctx.measureText) {
    return false;
  }
  return true;
}

//draws the scale, returns the width
function draw_scale(ctx, metrics, alphabet_ic, raster) {
  "use strict";
  var tic_height, i;
  tic_height = metrics.stack_height / alphabet_ic;
  ctx.save();
  ctx.translate(metrics.y_label_height, metrics.y_num_height/2);
  //draw the axis label
  ctx.save();
  ctx.font = metrics.y_label_font;
  ctx.translate(0, metrics.stack_height/2);
  ctx.rotate(-(Math.PI / 2));
  ctx.textAlign = "center";
  ctx.fillText("bits", 0, 0);
  ctx.restore();

  ctx.translate(metrics.y_label_spacer + metrics.y_num_width, 0);

  //draw the axis tics
  ctx.save();
  ctx.translate(0, metrics.stack_height);
  for (i = 0; i <= alphabet_ic; i++) {
    //draw the number
    ctx.save();
    ctx.translate(-1, 0);
    raster.draw_scale_num(ctx, metrics.y_num_font, i);
    ctx.restore();
    //draw the tic
    ctx.fillRect(0, -1, metrics.y_tic_width, 2);
    //prepare for next tic
    ctx.translate(0, -tic_height);
  }
  ctx.restore();

  ctx.fillRect(metrics.y_tic_width - 2, 0, 2, metrics.stack_height)

  ctx.restore();
}

function draw_stack_num(ctx, metrics, row_index, raster) {
  "use strict";
  ctx.save();
  ctx.translate(0, Math.round(metrics.stack_height + metrics.x_num_above));
  if (metrics.x_axis == 1) {
    raster.draw_stack_num(ctx, metrics.x_num_font, metrics.stack_width, row_index);
  } else if (metrics.x_axis == 0) {
    // draw dots instead of the numbers (good for small logos)
    ctx.beginPath();
    var radius = Math.round(metrics.x_num_height / 2);
    ctx.arc(Math.round(metrics.stack_width / 2), radius, radius, 0, 2 * Math.PI, false);
    ctx.fill();
  }
  ctx.restore();
}

function draw_stack(ctx, metrics, symbols, raster) {
  "use strict";
  var preferred_pad, sym_min, i, sym, sym_height, pad;
  preferred_pad = 0;
  sym_min = 5;

  ctx.save();//1
  ctx.translate(0, metrics.stack_height);
  for (i = 0; i < symbols.length; i++) {
    sym = symbols[i];
    sym_height = metrics.stack_height * sym.get_scale();
    
    pad = preferred_pad;
    if (sym_height - pad < sym_min) {
      pad = Math.min(pad, Math.max(0, sym_height - sym_min));
    }
    sym_height -= pad;

    //translate to the correct position
    ctx.translate(0, -(pad/2 + sym_height));

    //draw
    raster.draw_stack_sym(ctx, sym.get_symbol(), 0, 0, metrics.stack_width, sym_height);
    //translate past the padding
    ctx.translate(0, -(pad/2));
  }
  ctx.restore();//1
}

function draw_dashed_line(ctx, pattern, start, x1, y1, x2, y2) {
  "use strict";
  var x, y, len, i, dx, dy, tlen, theta, mulx, muly, lx, ly;
  dx = x2 - x1;
  dy = y2 - y1;
  tlen = Math.pow(dx*dx + dy*dy, 0.5);
  theta = Math.atan2(dy,dx);
  mulx = Math.cos(theta);
  muly = Math.sin(theta);
  lx = [];
  ly = [];
  for (i = 0; i < pattern; ++i) {
    lx.push(pattern[i] * mulx);
    ly.push(pattern[i] * muly);
  }
  i = start;
  x = x1;
  y = y1;
  len = 0;
  ctx.beginPath();
  while (len + pattern[i] < tlen) {
    ctx.moveTo(x, y);
    x += lx[i];
    y += ly[i];
    ctx.lineTo(x, y);
    len += pattern[i];
    i = (i + 1) % pattern.length;
    x += lx[i];
    y += ly[i];
    len += pattern[i];
    i = (i + 1) % pattern.length;
  }
  if (len < tlen) {
    ctx.moveTo(x, y);
    x += mulx * (tlen - len);
    y += muly * (tlen - len);
    ctx.lineTo(x, y);
  }
  ctx.stroke();
}

function draw_trim_background(ctx, metrics, left_start, left_end, left_divider, right_start, right_end, right_divider) {
  "use strict";
  var left_size = left_end - left_start;
  var right_size = right_end - right_start;
  var line_x;

  ctx.save();//s8
  ctx.fillStyle = "rgb(240, 240, 240)";
  if (left_size > 0) {
    ctx.fillRect(left_start * metrics.stack_width, 0, left_size * metrics.stack_width, metrics.stack_height);
  }
  if (right_size > 0) {
    ctx.fillRect(right_start * metrics.stack_width, 0, right_size * metrics.stack_width, metrics.stack_height);
  }
  ctx.fillStyle = "rgb(51, 51, 51)";
  if (left_size > 0 && left_divider) {
    line_x = (left_end * metrics.stack_width) - 0.5;
    draw_dashed_line(ctx, [3], 0, line_x, 0, line_x, metrics.stack_height);
  }
  if (right_size > 0 && right_divider) {
    line_x = (right_start * metrics.stack_width) + 0.5;
    draw_dashed_line(ctx, [3], 0, line_x, 0, line_x, metrics.stack_height);
  }
  ctx.restore();//s8
}

function size_logo_on_canvas(logo, canvas, show_names, scale) {
  "use strict";
  var draw_name, draw_finetext, metrics;
  draw_name = (typeof show_names === "boolean" ? show_names : (logo.get_rows() > 1));
  draw_finetext = (logo.fine_text.length > 0);
  if (canvas.width !== 0 && canvas.height !== 0) {
    return;
  }
  metrics = new LogoMetrics(canvas.getContext('2d'), 
      logo.get_xlate_columns(), logo.get_rows(), draw_name, draw_finetext, logo.x_axis, logo.y_axis);
  if (typeof scale == "number") {
    //resize the canvas to fit the scaled logo
    canvas.width = metrics.summed_width * scale;
    canvas.height = metrics.summed_height * scale;
  } else {
    if (canvas.width === 0 && canvas.height === 0) {
      canvas.width = metrics.summed_width;
      canvas.height = metrics.summed_height;
    } else if (canvas.width === 0) {
      canvas.width = metrics.summed_width * (canvas.height / metrics.summed_height);
    } else if (canvas.height === 0) {
      canvas.height = metrics.summed_height * (canvas.width / metrics.summed_width);
    }
  }
}

function draw_logo_on_canvas(logo, canvas, show_names, scale) {
  "use strict";
  var i, draw_name, draw_finetext, ctx, metrics, raster, pspm_i, pspm, 
      offset, col_index, motif_position, ssc;
  ssc = false;
  draw_name = (typeof show_names === "boolean" ? show_names : (logo.get_rows() > 1));
  draw_finetext = (logo.fine_text.length > 0);
  ctx = canvas.getContext('2d');
  //assume that the user wants the canvas scaled equally so calculate what the best width for this image should be
  metrics = new LogoMetrics(ctx, logo.get_xlate_columns(), logo.get_rows(), draw_name, draw_finetext, logo.x_axis, logo.y_axis);
  if (typeof scale == "number") {
    //resize the canvas to fit the scaled logo
    canvas.width = metrics.summed_width * scale;
    canvas.height = metrics.summed_height * scale;
  } else {
    if (canvas.width === 0 && canvas.height === 0) {
      scale = 1;
      canvas.width = metrics.summed_width;
      canvas.height = metrics.summed_height;
    } else if (canvas.width === 0) {
      scale = canvas.height / metrics.summed_height;
      canvas.width = metrics.summed_width * scale;
    } else if (canvas.height === 0) {
      scale = canvas.width / metrics.summed_width;
      canvas.height = metrics.summed_height * scale;
    } else {
      scale = Math.min(canvas.width / metrics.summed_width, canvas.height / metrics.summed_height);
    }
  }
  // cache the raster based on the assumption that we will be drawing a lot
  // of logos the same size and alphabet
  if (typeof draw_logo_on_canvas.raster_cache === "undefined") {
    draw_logo_on_canvas.raster_cache = [];
  }
  for (i = 0; i < draw_logo_on_canvas.raster_cache.length; i++) {
    raster = draw_logo_on_canvas.raster_cache[i];
    if (raster.get_alphabet().equals(logo.alphabet) &&
        Math.abs(raster.get_scale() - scale) < 0.1) break;
    raster = null;
  }
  if (raster == null) {
    raster = new RasterizedAlphabet(logo.alphabet, scale, metrics.stack_font, metrics.stack_width);
    draw_logo_on_canvas.raster_cache.push(raster);
  }
  ctx = canvas.getContext('2d');
  ctx.save();//s1
  ctx.scale(scale, scale);
  ctx.save();//s2
  ctx.save();//s7
  //create margin
  ctx.translate(Math.round(metrics.pad_left), Math.round(metrics.pad_top));
  for (pspm_i = 0; pspm_i < logo.get_rows(); ++pspm_i) {
    pspm = logo.get_pspm(pspm_i);
    offset = logo.get_offset(pspm_i);
    //optionally draw name if this isn't the last row or is the only row 
    if (draw_name && (logo.get_rows() == 1 || pspm_i != (logo.get_rows()-1))) {
      ctx.save();//s4
      ctx.translate(Math.round(metrics.summed_width/2), Math.round(metrics.name_height));
      ctx.font = metrics.name_font;
      ctx.textAlign = "center";
      ctx.fillText(pspm.name, 0, 0);
      ctx.restore();//s4
      ctx.translate(0, Math.round(metrics.name_height + 
          Math.min(0, metrics.name_spacer - metrics.y_num_height/2)));
    }
    //draw scale
    if (logo.y_axis) draw_scale(ctx, metrics, logo.alphabet.get_ic(), raster);
    ctx.save();//s5
    //translate across past the scale
    if (logo.y_axis) {
      ctx.translate(Math.round(metrics.y_label_height + metrics.y_label_spacer + 
        metrics.y_num_width + metrics.y_tic_width), Math.round(metrics.y_num_height / 2));
    }
    //draw the trimming background
    if (pspm.get_left_trim() > 0 || pspm.get_right_trim() > 0) {
      var left_start = offset * logo.get_xlate_nsyms();
      var left_end = (offset + pspm.get_left_trim()) * logo.get_xlate_nsyms();
      var left_divider = true;
      if (left_end < logo.get_xlate_start() || left_start > logo.get_xlate_end()) {
        // no overlap
        left_start = 0;
        left_end = 0;
        left_divider = false;
      } else {
        if (left_start < logo.get_xlate_start()) {
          left_start = logo.get_xlate_start();
        }
        if (left_end > logo.get_xlate_end()) {
          left_end = logo.get_xlate_end();
          left_divider = false;
        }
        left_start -= logo.get_xlate_start();
        left_end -= logo.get_xlate_start();
        if (left_end < left_start) {
          left_start = 0;
          left_end = 0;
          left_divider = false;
        }
      }
      var right_end = (offset + pspm.get_motif_length()) * logo.get_xlate_nsyms();
      //var right_start = right_end - (pspm.get_left_trim() * logo.get_xlate_nsyms());
      var right_start = right_end - (pspm.get_right_trim() * logo.get_xlate_nsyms());
      var right_divider = true;
      if (right_end < logo.get_xlate_start() || right_start > logo.get_xlate_end()) {
        // no overlap
        right_start = 0;
        right_end = 0;
        right_divider = false;
      } else {
        if (right_start < logo.get_xlate_start()) {
          right_start = logo.get_xlate_start();
          right_divider = false;
        }
        if (right_end > logo.get_xlate_end()) {
          right_end = logo.get_xlate_end();
        }
        right_start -= logo.get_xlate_start();
        right_end -= logo.get_xlate_start();
        if (right_end < right_start) {
          right_start = 0;
          right_end = 0;
          right_divider = false;
        }
      }
      draw_trim_background(ctx, metrics, left_start, left_end, left_divider, right_start, right_end, right_divider);
    }
    //draw letters
    var xlate_col;
    for (xlate_col = logo.get_xlate_start(); xlate_col < logo.get_xlate_end(); xlate_col++) {
      ctx.translate(metrics.stack_pad_left,0);
      col_index = Math.floor(xlate_col / logo.get_xlate_nsyms());
      if (xlate_col % logo.get_xlate_nsyms() == 0) {
        if (col_index >= offset && col_index < (offset + pspm.get_motif_length())) {
          motif_position = col_index - offset;
          draw_stack_num(ctx, metrics, motif_position, raster);
          draw_stack(ctx, metrics, pspm.get_stack(motif_position, logo.alphabet, ssc), raster);
        }
      } else {
        if (col_index >= offset && col_index < (offset + pspm.get_motif_length())) {
          ctx.save();// s5.1
          ctx.translate(0, Math.round(metrics.stack_height));
          // TODO draw a dot or dash or something to indicate continuity of the motif
          ctx.restore(); //s5.1
        }
      }
      ctx.translate(Math.round(metrics.stack_width), 0);
    }
    ctx.restore();//s5
    ////optionally draw name if this is the last row but isn't the only row 
    if (draw_name && (logo.get_rows() != 1 && pspm_i == (logo.get_rows()-1))) {
      //translate vertically past the stack and axis's        
      ctx.translate(0, metrics.y_num_height/2 + metrics.stack_height + 
          Math.max(metrics.y_num_height/2, metrics.x_num_above + metrics.x_num_width + metrics.name_spacer));

      ctx.save();//s6
      ctx.translate(metrics.summed_width/2, metrics.name_height);
      ctx.font = metrics.name_font;
      ctx.textAlign = "center";
      ctx.fillText(pspm.name, 0, 0);
      ctx.restore();//s6
      ctx.translate(0, metrics.name_height);
    } else {
      //translate vertically past the stack and axis's        
      ctx.translate(0, metrics.y_num_height/2 + metrics.stack_height + 
          Math.max(metrics.y_num_height/2, metrics.x_num_above + metrics.x_num_width));
    }
    //if not the last row then add middle padding
    if (pspm_i != (logo.get_rows() -1)) {
      ctx.translate(0, metrics.pad_middle);
    }
  }
  ctx.restore();//s7
  if (logo.fine_text.length > 0) {
    ctx.translate(metrics.summed_width - metrics.pad_right, metrics.summed_height - metrics.pad_bottom);
    ctx.font = metrics.fine_txt_font;
    ctx.textAlign = "right";
    ctx.fillText(logo.fine_text, 0,0);
  }
  ctx.restore();//s2
  ctx.restore();//s1
}

function create_canvas(c_width, c_height, c_id, c_title, c_display) {
  "use strict";
  var canvas = document.createElement("canvas");
  //check for canvas support before attempting anything
  if (!canvas.getContext) {
    return null;
  }
  var ctx = canvas.getContext('2d');
  //check for html5 text drawing support
  if (!supports_text(ctx)) {
    return null;
  }
  //size the canvas
  canvas.width = c_width;
  canvas.height = c_height;
  canvas.id = c_id;
  canvas.title = c_title;
  canvas.style.display = c_display;
  return canvas;
}

function logo_1(alphabet, fine_text, pspm) {
  "use strict";
  var logo = new Logo(alphabet, fine_text);
  logo.add_pspm(pspm);
  return logo;
}

function logo_2(alphabet, fine_text, target, query, query_offset) {
  "use strict";
  var logo = new Logo(alphabet, fine_text);
  if (query_offset < 0) {
    logo.add_pspm(target, -query_offset);
    logo.add_pspm(query);
  } else {
    logo.add_pspm(target);
    logo.add_pspm(query, query_offset);
  }      
  return logo;
}

/*
 * Specifies an alternate source for an image.
 * If the image with the image_id specified has
 * not loaded then a generated logo will be used 
 * to replace it.
 *
 * Note that the image must either have dimensions
 * or a scale must be set.
 */
function alternate_logo(logo, image_id, scale) {
  "use strict";
  var image = document.getElementById(image_id);
  if (!image) {
    alert("Can't find specified image id (" +  image_id + ")");
    return;
  }
  //if the image has loaded then there is no reason to use the canvas
  if (image_ok(image)) {
    return;
  }
  //the image has failed to load so replace it with a canvas if we can.
  var canvas = create_canvas(image.width, image.height, image_id, image.title, image.style.display);
  if (canvas === null) {
    return;
  }
  //draw the logo on the canvas
  draw_logo_on_canvas(logo, canvas, null, scale);
  //replace the image with the canvas
  image.parentNode.replaceChild(canvas, image);
}

/*
 * Specifies that the element with the specified id
 * should be replaced with a generated logo.
 */
function replace_logo(logo, replace_id, scale, title_txt, display_style) {
  "use strict";
  var element = document.getElementById(replace_id);
  if (!replace_id) {
    alert("Can't find specified id (" + replace_id + ")");
    return;
  }
  //found the element!
  var canvas = create_canvas(50, 120, replace_id, title_txt, display_style);
  if (canvas === null) {
    return;
  }
  //draw the logo on the canvas
  draw_logo_on_canvas(logo, canvas, null, scale);
  //replace the element with the canvas
  element.parentNode.replaceChild(canvas, element);
}

/*
 * Fast string trimming implementation found at
 * http://blog.stevenlevithan.com/archives/faster-trim-javascript
 *
 * Note that regex is good at removing leading space but
 * bad at removing trailing space as it has to first go through
 * the whole string.
 */
function trim (str) {
  "use strict";
  var ws, i;
  str = str.replace(/^\s\s*/, '');
  ws = /\s/; i = str.length;
  while (ws.test(str.charAt(--i)));
  return str.slice(0, i + 1);
}

    </script>
    <script type="text/javascript">
//
// simple-shared-doc.js
//
// Function to replace the innerHTML of element "id" with the HTML indicated by "doc_type".
// Easier to read and update than the more flexible approach in shared-doc.js. 
//
function print_doc(id, doc_type) {
  var html;
  switch (doc_type) {
    case 'motif-consensus':
      html = `
	<p id="consensus_doc"> 
	   A <b>consensus sequence</b> is constructed from each column in a
	   motif's frequency matrix using the <b>"50% rule"</b>
	   as follows:
	</p>
	<ol>
	  <li>The letter frequencies in the column are sorted in decreasing order.</li>
	  <li>Letters with frequency less 50% of the maximum are discarded.</li>
	  <li>The letter used in this position in the consensus sequence is determined
	  by the first rule below that applies:</li>
	  <ul>
	    <li>If there is only one letter left, or if the remaining letters exactly match
	    an ambiguous symbol in the alphabet, the <b>letter</b> or <b>ambiguous symbol</b>,
	    respectively, is used.</li>
	    <li>Otherwise, if the remaining set contains at least 50% of the core
	    symbols in the alphabet, the alphabet's <b>wildcard</b>
	    (e.g., "N" for DNA or RNA, and "X" for protein) is used.</li>
	    <li>Otherwise, the letter with the <b>maximum frequency</b> is used.</li>
	  </ul>
	</ol>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'ame-results-header':
      html = `
	<p>
          AME outputs a tab-separated values (TSV) file ('ame.tsv') that contains one line for each motif
	  found to be significantly enriched.
	  The lines are sorted in order of decreasing statistical significance.
          The first line contains the (tab-separated) names of the fields.
	  Your command line is given at the end of the file in a comment line starting with the
	  character '#'.
	</p>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'ame-results-table':
      html = `
	<p>
	  The <a href="` + site_url + `/doc/ame-output-format.html#tsv_results">AME Results TSV Format</a> 
	  differs depending on the choice of method for measuring motif enrichment, as shown in the following table. 
	</p>
	<table style="width:100%" border=1>
	  <tr>
	    <th>Method</th>
	    <th colspan=19>Output Columns</th>
	  </tr>
	  <tr>
	    <th></th> 
	    <th>1</th>
	    <th>2</th>
	    <th>3</th>
	    <th>4</th>
	    <th>5</th>
	    <th>6</th>
	    <th>7</th>
	    <th>8</th>
	    <th>9</th>
	    <th>10</th>
	    <th>11</th>
	    <th>12</th>
	    <th>13 </th>
	    <th>14 </th>
	    <th>15 </th>
	    <th>16 </th>
	    <th>17 </th>
	    <th>18 </th>
	    <th>19 </th>
	  </tr>
	  <tr>
	    <td>fisher</td>
	    <td>rank </td>
	    <td>motif_DB </td>
	    <td>motif_ID </td>
	    <td>motif_alt_ID </td>
	    <td>consensus </td>
	    <td>p-value </td>
	    <td>adj_p-value </td>
	    <td>E-value</td>
	    <td>tests </td>
	    <td>FASTA_max </td>
	    <td>pos </td>
	    <td>neg </td>
	    <td>PWM_min</td>
	    <td>TP </td>
	    <td>%TP </td>
	    <td>FP </td>
	    <td>%FP </td>
	    </td> <td>
	    </td> <td>
	  </tr>
	  <tr>
	    <td>ranksum</td>
	    <td>rank </td>
	    <td>motif_DB </td>
	    <td>motif_ID </td>
	    <td>motif_alt_ID </td>
	    <td>consensus </td>
	    <td>p-value </td>
	    <td>adj_p-value </td>
	    <td>E-value</td>
	    <td>tests </td>
	    <td>FASTA_max </td>
	    <td>pos </td>
	    <td>neg</td>
	    <td>U </td>
	    <td>pleft </td>
	    <td>pright </td>
	    <td>pboth </td>
	    <td>adj_pleft </td>
	    <td>adj_pright </td>
	    <td>adj_both</td>
	  </tr>
	  <tr>
	    <td>3dmhg</td>
	    <td>rank </td>
	    <td>motif_DB </td>
	    <td>motif_ID </td>
	    <td>motif_alt_ID </td>
	    <td>consensus </td>
	    <td>p-value </td>
	    <td>adj_p-value </td>
	    <td>E-value</td>
	    <td>tests </td>
	    <td>FASTA_max </td>
	    <td>pos </td>
	    <td>neg</td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	  </tr>
	  <tr>
	    <td>4dmhg</td>
	    <td>rank </td>
	    <td>motif_DB </td>
	    <td>motif_ID </td>
	    <td>motif_alt_ID </td>
	    <td>consensus </td>
	    <td>p-value </td>
	    <td>adj_p-value </td>
	    <td>E-value</td>
	    <td>tests </td>
	    <td>FASTA_max </td>
	    <td>pos </td>
	    <td>neg</td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	  </tr>
	  <tr>
	    <td>pearson</td>
	    <td>rank </td>
	    <td>motif_DB </td>
	    <td>motif_ID </td>
	    <td>motif_alt_ID </td>
	    <td>consensus </td>
	    <td>p-value </td>
	    <td>adj_p-value </td>
	    <td>E-value</td>
	    <td>tests </td>
	    <td>Pearson_CC </td>
	    <td>mean_squared_error </td>
	    <td>slope </td>
	    <td>intercept</td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	  </tr>
	  <tr>
	    <td>spearman</td>
	    <td>rank </td>
	    <td>motif_DB </td>
	    <td>motif_ID </td>
	    <td>motif_alt_ID </td>
	    <td>consensus </td>
	    <td>p-value </td>
	    <td>adj_p-value </td>
	    <td>E-value</td>
	    <td>tests </td>
	    <td>Spearman_CC</td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	    </td> <td>
	  </tr>
        </table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'ame-results-description':
      html = `
        <p>
          The names and meanings of each of the fields in <a href="` + site_url + `/doc/ame-output-format.html#tsv_results">AME Results TSV Format</a>,
          which depend on the method of enrichment analysis you chose, are described below.
        </p>
        <table class="dark" style="width:100%" border=1>
	  <tr> <th>method</th> <th>field</th> <th>name</th> <th>contents</th> </tr>	
    	  <tr> <td>all</td> <td>1</td> <td>rank</td> <td>The rank of the significance of the motif in the sorted results.</td> </tr>
    	  <tr> <td>all</td> <td>2</td> <td>motif_DB</td> <td> ` + get_doc_text('motif-db', 'the motif.') + `</td> </tr>
    	  <tr> <td>all</td> <td>3</td> <td>motif_ID</td> <td> ` + get_doc_text('motif-id') + `</td> </tr>
    	  <tr> <td>all</td> <td>4</td> <td>motif_alt_ID</td> <td> ` + get_doc_text('motif-alt-id') + `</td> </tr>
    	  <tr> <td>all</td> <td>5</td> <td>consensus</td> <td> ` + get_doc_text('motif-cons') + `</td> </tr>
    	  <tr> <td>all</td> <td>6</td> <td>p-value</td> <td> ` + get_doc_text('ame-pvalue') + `</td> </tr>
    	  <tr> <td>all</td> <td>7</td> <td>adj_p-value</td> <td> ` + get_doc_text('ame-adj-pvalue') + `</td> </tr>
    	  <tr> <td>all</td> <td>8</td> <td>E-value</td> <td> ` + get_doc_text('ame-evalue') + `</td> </tr>
    	  <tr> <td>all</td> <td>9</td> <td>tests</td> <td>The number of tests performed; used in correcting the <i>p</i>-value</td> </tr>
	  <tr> <td colspan=4></td>
          <tr> <td>fisher, ranksum, 3dmhg, 4dmhg</td> <td>10</td> <td>FASTA_max</td> <td>The optimal threshold for <b>labeling</b> sequences as positive;
            sequences with FASTA score less than or equal to the threshold are labeled as positive;
            this field will be contain the number of primary sequences if you provided
            a control file (using option <code>--control</code>); this field will contain the size of the
            partition if you specified one (using option <code>--fix-partition</code>).</td> </tr>
          <tr> <td>fisher, ranksum, 3dmhg, 4dmhg</td> <td>11</td> <td>pos</td> <td>The number of sequences <b>labeled</b> as positive.</td> </tr>
          <tr> <td>fisher, ranksum, 3dmhg, 4dmhg</td> <td>12</td> <td>neg</td> <td>The number of sequences <b>labeled</b> as negative.</td> </tr>
	  <tr> <td colspan=4></td>
          <tr> <td>fisher</td> <td>13</td> <td>PWM_min</td> <td>The optimal threshold on PWM score for <b>classifying</b> sequences as positive;
                sequences with PWM score greater than or equal to the threshold are classified as positive.</td> </tr>
          <tr> <td>fisher</td> <td>14</td> <td>TP</td> <td>The number of true positive sequences: sequences both <b>labeled</b> and <b>classified</b> as positive</td> </tr>
          <tr> <td>fisher</td> <td>15</td> <td>%TP</td> <td>The percentage of true positive sequences: percentage of sequences <b>labeled</b> positive and <b>classified</b> as positive.</td> </tr>
          <tr> <td>fisher</td> <td>16</td> <td>FP</td> <td>The number of false positive sequences: sequences <b>labeled</b> negative but <b>classified</b> as positive.</td> </tr>
          <tr> <td>fisher</td> <td>17</td> <td>%TP</td> <td>The percentage of false positive sequences: sequences <b>labeled</b> negative but <b>classified</b> as positive.</td> </tr>
	  <tr> <td colspan=4></td>
          <tr> <td>ranksum</td> <td>13</td> <td>U</td> <td>The value of the Mann-Whitney <i>U</i> statistic.</td> </tr>
          <tr> <td>ranksum</td> <td>14</td> <td>pleft</td> <td>The left-tailed <i>p</i>-value of the rank-sum test.</td> </tr>
          <tr> <td>ranksum</td> <td>15</td> <td>pright</td> <td>The right-tailed <i>p</i>-value of the rank-sum test.</td> </tr>
          <tr> <td>ranksum</td> <td>16</td> <td>pboth</td> <td>The two-tailed <i>p</i>-value of the rank-sum test.</td> </tr>
          <tr> <td>ranksum</td> <td>17</td> <td>adj_pleft</td> <td>The left-tailed <i>p</i>-value of the rank-sum test, adjusted for multiple tests.</td> </tr>
          <tr> <td>ranksum</td> <td>18</td> <td>adj_pright</td> <td>The right-tailed <i>p</i>-value of the rank-sum test, adjusted for multiple tests.</td> </tr>
          <tr> <td>ranksum</td> <td>19</td> <td>adj_both</td> <td>The two-tailed <i>p</i>-value of the rank-sum test, adjusted for multiple tests.</td> </tr>
	  <tr> <td colspan=4></td>
	  <tr> <td>pearson</td> <td>10</td> <td>Pearson_CC</td> <td>The correlation coefficient of the PWM and FASTA scores of positive sequences.</td> </tr>
	  <tr> <td>pearson</td> <td>11</td> <td>mean_squared_error</td> <td>The mean-squared error of the regression line between PWM and FASTA scores.</td> </tr>
	  <tr> <td>pearson</td> <td>12</td> <td>slope</td> <td>The slope of the regression line.</td> </tr>
	  <tr> <td>pearson</td> <td>13</td> <td>intercept</td> <td>The y-intercept of the regression line.</td> </tr>
	  <tr> <td colspan=4></td>
	  <tr> <td>spearman</td> <td>10</td> <td>Spearman_CC</td> <td>The correlation coefficient of the PWM and FASTA ranks of positive sequences./td> </tr>
        </table>
	  </ul>
	</p>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'ame-sequences-tsv':
      html = `
        <p>AME outputs a tab-separated values (TSV) file ('sequences.tsv') containing one line for
	each sequence classified as 'positive' by AME for each significant motif.
        The first line contains the (tab-separated) names of the fields.
	The lines are grouped by motif, and groups are separated by a line
	starting with the character "#".
        </p>
        <p>
          The names and meanings of each of the fields in <a href="` + site_url + `/doc/ame-output-format.html#tsv_sequences">AME Significant Sequences TSV Format</a>,
          are described below.
        </p>
        <table class="dark" style="width:100%" border=1>
          <tr> <th>field</th> <th>name</th> <th>contents</th> </tr>
	  <tr> <td>1</td> <td>motif_DB</td> <td>` + get_doc_text('motif-db', 'the motif.') + `</td> </tr>
	  <tr> <td>2</td> <td>motif_ID</td> <td>the ID of the motif</td> </tr>
	  <tr> <td>3</td> <td>seq_ID</td> <td>the ID of the sequence</td> </tr>
	  <tr> <td>4</td> <td><i>label_score</i> (either FASTA_score or PWM_score)</td> <td>the value of the score used to label it as positive</td> </tr>
	  <tr> <td>5</td> <td><i>class_score</i> (either PWM_score or FASTA_score)</td> <td>the value of the score used to classify it as positive</td> </tr>
	  <tr> <td>6</td> <td>class</td> <td>whether the sequence is a true positive, 'tp', or a false positive, 'fp'</td> </tr>
        </table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'centrimo-results-tsv':
      html = `
	<p>
          CentriMo outputs a tab-separated values (TSV) file ('centrimo.tsv') that contains one line for each 
	  region found to be significantly enriched for a motif.
	  The lines are sorted in order of decreasing statistical significance.
          The first line contains the (tab-separated) names of the fields.
          Your command line is given at the end of the file in a comment line starting with the
          character '#'.
	</p>
	<p>
          The names and meanings of each of the fields in <a href="` + site_url + `/doc/centrimo-output-format.html#tsv_results">CentriMo Results TSV Format</a>,
          which depend on whether or not you provide control sequences to CentriMo, are described below.
	</p>
        <table class="dark" style="width:100%" border=1>
          <tr> <th>field</th> <th>name</th> <th>contents</th> </tr>
	  <tr> <td>1</td> <td>db_index</td> <td>The index of the motif file that contains the motif.  Motif
	      files are numbered in the order the appeared in the command line.</td> </tr>
	  <tr> <td>2</td> <td>motif_id</td> <td> ` + get_doc_text('motif-id') + `
	      If more than one motif has the same ID, CentriMo uses only the first such motif.
	      The name is single-quoted and preceded with '+' or '-' if you scanned separately with 
	      the reverse complement motif (using the <code>--sep</code> option).</td> </tr>
	  <tr> <td>3</td> <td>motif_alt_id</td> <td> ` + get_doc_text('motif-alt-id') + `</td> </tr>
	  <tr> <td>4</td> <td>consensus</td> <td> ` + get_doc_text('motif-cons') + `</td> </tr>
	  <tr> <td>5</td> <td>E-value</td> <td> ` + get_doc_text('centrimo-evalue') + `</td> </tr>
	  <tr> <td>6</td> <td>adj_p-value</td> <td> ` + get_doc_text('centrimo-adj-pvalue') + `
	      By default, a <i>p</i>-value is calculated by using the one-tailed binomial
	      test on the number of sequences with a match to the motif
	      that have their best match in the reported region;
              if you provided control sequences, the <i>p</i>-value of Fisher\'s exact test on the enrichment of 
              best matches in the positive sequences relative to the negative sequences is computed instead;
	      if you used the <code>--cd</code> option, the <i>p</i>-value is the probability that the average 
	      distance between the best site and the sequence center would be as low or lower than observed, 
	      computed using the cumulative Bates distribution, optimized over different score thresholds.
	      In all cases, the reported <i>p</i>-value has been adjusted for the number of regions
	      and/or score thresholds tested.</td> </tr>
	  <tr> <td>7</td> <td>log_adj_p-value</td> <td>Log of adjusted <i>p</i>-value.</td> </tr>
	  <tr> <td>8</td> <td>bin_location</td> <td>Location of the center of the most enriched region, or
		0 if you used the <code>--cd</code> option.
	  <tr> <td>9</td> <td>bin_width</td> <td> ` + get_doc_text('centrimo-bin-width') + `</td> </tr>
	  <tr> <td>10</td> <td>total_width</td> <td>The maximum number of regions possible for this motif
              <br>&nbsp;&nbsp;
	      round(sequence_length - motif_length + 1)/2,<br>
	      or the number of places the motif will fit if you used the <code>--cd</code> option.</td> </tr>
	  <tr> <td>11</td> <td>sites_in_bin</td> <td> ` + get_doc_text('centrimo-sites-in-bin') + `</td> </tr>
	  <tr> <td>12</td> <td>total_sites</td> <td>The number of sequences containing a match to the motif
	      above the score threshold.
	  <tr> <td>13</td> <td>p_success</td> <td>The probability of a random match falling into the enriched region:
	      <br>&nbsp;&nbsp;
	      bin_width / total_width</td> </tr>
	  <tr> <td>14</td> <td>p-value</td> <td>The uncorrected <i>p</i>-value before it gets adjusted for the
	      number of multiple tests to give the adjusted <i>p</i>-value.</td> </tr>
	  <tr> <td>15</td> <td>mult_tests</td> <td> ` + get_doc_text('centrimo-mult-tests') + `</td> </tr>
	  <tr> <th colspan=3>The following additional columns are present when you provide control sequences to CentriMo
	  (using the <code>--neg</code> option).</th> </tr>
	  <tr> <td>16</td> <td>neg_sites_in_bin</td> <td>The number of negative sequences where the best
	      match to the motif falls in the reported region.
	      This value is rounded but the underlying value may contain fractional counts.
	      Note: This number may be less than the number of negative have a best match in the region.
	      The reason for this is that a sequence may have many matches that score equally best.
	      If n matches have the best score in a sequence, 1/n is added to the
	      appropriate bin for each match.</td> </tr>
	  <tr> <td>17</td> <td>neg_sites</td> <td>The number of negative sequences containing a match to the
	      motif above the minimum score threshold.
	      When score optimization is enabled the score threshold may be raised
	      higher than the minimum.</td> </tr>
	  <tr> <td>18</td> <td>neg_adj_pvalue</td> <td>The probability that any tested region in the negative
	      sequences would be as enriched for best matches to this motif
	      according to the Binomial test.</td> </tr>
	  <tr> <td>19</td> <td>log_neg_adj_pvalue</td> <td>Log of negative adjusted <i>p</i>-value.</td> </tr>
	  <tr> <td>20</td> <td>fisher_adj_pvalue</td> <td>Fisher adjusted <i>p</i>-value before it gets adjusted for the
	      number of motifs in the input files(s).</td> </tr>
	  <tr> <td>21</td> <td>log_fisher_adj_pvalue</td> <td>Log of Fisher adjusted <i>p</i>-value.</td> </tr>
        </table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'centrimo-sites-txt':
      html = `
	<p>
          CentriMo outputs a text file ('site_counts.txt') that contains,
	  for each motif, pairs of values (bin_position, site_count),
	  or triples of values (bin_position, site_count, neg_site_count) if
	  you provided control sequences to CentriMo.
	  This data can be used to plot the density of motif best matches (sites)
	  along the input sequences.  Fractional counts are possible if multiple (n)
	  bins contain the best match for a given sequence, with each bin  
	  receiving an incremental count of 1/n. 
	</p>
	<p>
	  The data for each motif begins with a header line with the format:
	  <br>&nbsp&nbsp
		DB &lt;db_number&gt; MOTIF &lt;id&gt; &lt;alt&gt;
	  </br>
	  where &lt;id&gt; and &lt;alt&gt; are as described above.
	  The following lines (up to the next header line) 
	  each contain a single value-pair or value-triple for the motif
	  named in the header line.
	</p>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'meme-chip-results-tsv':
      html = `
	<p>
          MEME-ChIP outputs a tab-separated values (TSV) file ('summary.tsv') that 
	  contains one line for each motif found by MEME-ChIP.
	  The lines are sorted in order of decreasing statistical significance.
	  The first line contains the (tab-separated) names of the fields.
	  Your command line is given at the end of the file in a comment line starting with the
	  character '#'.
	</p>
	<p>
          The names and meanings of the fields in the 
	  <a href="` + site_url + `/doc/meme-chip-output-format.html">MEME-ChIP Results Summary TSV Format</a> 
	  are described below. 
	</p>
        <table class="dark" style="width:100%" border=1>
          <tr> <th>field</th> <th>name</th> <th>contents</th> </tr>
          <tr> <td>1</td> <td>MOTIF_INDEX</td> <td>The index of the motif in the "Motifs in MEME text format" file ('combined.meme') 
		output by MEME-ChIP.</td> </tr>
          <tr> <td>2</td> <td>MOTIF_SOURCE</td> <td>The name of the program that found the <i>de novo</i> motif, or the
                name of the motif file containing the known motif.</td> </tr>
          <tr> <td>3</td> <td>MOTIF_ID</td> <td> ` + get_doc_text('motif-id') + `</td> </tr>
          <tr> <td>4</td> <td>ALT_ID</td> <td> ` + get_doc_text('motif-alt-id') + `</td> </tr>
          <tr> <td>5</td> <td>CONSENSUS</td> <td>The ID of the <i>de novo</i> motif, or a consensus sequence
                computed from the letter frequencies in the known motif
                (as described <a href="#consensus_doc">below</a>).</td> </tr>
          <tr> <td>6</td> <td>WIDTH</td> <td>The width of the motif.</td> </tr>
          <tr> <td>7</td> <td>SITES</td> <td>The number of sites reported by the <i>de novo</i> program, or the number
                of "Total Matches" reported by CentriMo.</td> </tr>
          <tr> <td>8</td> <td>E-VALUE</td> <td>The statistical significance of the motif.</td> </tr>
          <tr> <td>9</td> <td>E-VALUE_SOURCE</td> <td>The program that reported the <i>E</i>-value.</td> </tr>
          <tr> <td>10</td> <td>MOST_SIMILAR_MOTIF</td> <td>The known motif most similar to this motif according to Tomtom.</td> </tr>
          <tr> <td>11</td> <td>URL</td> <td>A link to a description of the most similar motif, or to the known motif.</td> </tr>
        </table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'meme-chip-combined-motifs':
      html = `
        <p>
          MEME-ChIP outputs a text file ('combined.meme') containing all the significant motifs found by MEME-ChIP.
          The motifs are in <a href="` + site_url + `/doc/meme-format.html">Minimal MEME Motif format</a>, 
	  and their IDs correspond to the motif indices given in the "Summary in TSV Format" file ('summary.tsv').
        </p>
        </p>
          <b>Note:</b> The 'nsites=' and 'E=' fields in the motif headers are only
          relevant for the MEME and DREME motifs.  For known motifs, those values do
          not refer to the number of sites in the input sequences.
        </p>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'spamo-results-tsv':
      html = `
	<p>
          SpaMo outputs a tab-separated values (TSV) file ('spamo.tsv') that contains one line for each motif
	  found to be significantly enriched.
	  The lines are grouped by secondary motif and sorted in order of decreasing statistical significance.
          The first line contains the (tab-separated) names of the fields.
	  Your command line is given at the end of the file in a comment line starting with the character '#'.
	</p>
	<p>
	  The names and meanings of each of the fields in <a href="` + site_url + `/doc/spamo-output-format.html#results-tsv">SpaMo Results TSV Format</a> 
	  are described below. 
	</p>
	<table class="dark" style="width:100%" border=1>
	  <tr> <th>field</th> <th>name</th> <th>contents</th> </tr>
          <tr> <td>1</td> <td>prim_db</td> <td> ` + get_doc_text('motif-db', 'the primary motif.') + `</td> </tr>
          <tr> <td>2</td> <td>prim_id</td> <td> ` + get_doc_text('motif-id', 'primary') + `</td> </tr>
          <tr> <td>3</td> <td>prim_alt</td> <td> ` + get_doc_text('motif-alt-id', 'primary') + `</td> </tr>
          <tr> <td>4</td> <td>prim_cons</td> <td> ` + get_doc_text('motif-cons', 'primary') + `</td> </tr>
          <tr> <td>5</td> <td>sec_db</td> <td> ` + get_doc_text('motif-db', 'the secondary motif.') + `</td> </tr>
          <tr> <td>6</td> <td>sec_id</td> <td> ` + get_doc_text('motif-id', 'secondary') + `</td> </tr>
          <tr> <td>7</td> <td>sec_alt</td> <td> ` + get_doc_text('motif-alt-id', 'secondary') + `</td> </tr>
          <tr> <td>8</td> <td>sec_cons</td> <td> ` + get_doc_text('motif-cons', 'secondary') + `</td> </tr>
          <tr> <td>9</td> <td>trim_left</td> <td>Number of positions trimmed from left of secondary motif.</td> </tr>
          <tr> <td>10</td> <td>trim_right</td> <td>Number of positions trimmed from right of secondary motif.</td> </tr> 
          <tr> <th colspan=3>If the next three fields are not blank, the motif is redundant with a more significant ('parent') motif.</th> </tr>
          <tr> <td>11</td> <td>red_db</td> <td> ` + get_doc_text('motif-db', 'the parent motif.') + `</td> </tr>
          <tr> <td>12</td> <td>red_id</td> <td> ` + get_doc_text('motif-id', 'parent') + `</td> </tr>
          <tr> <td>13</td> <td>red_alt</td> <td> ` + get_doc_text('motif-alt-id', 'parent') + `</td> </tr>
          <tr> <td>14</td> <td>E-value</td> <td>The expected number motifs that would have least one spacing as enriched as the best spacing for this secondary. 
	    The <i>E</i>-value is the best spacing <i>p</i>-value multiplied by the number of motifs in the input database(s).</td> </tr>
          <tr> <td>15</td> <td>gap</td> <td>The distance between the edge of the primary and the (trimmed) secondary motif.</td> </tr>
          <tr> <td>16</td> <td>orient</td> <td>The (combination) of quadrants for which occurrences of this spacing are combined.</td> </tr>
          <tr> <td>17</td> <td>count</td> <td>The number of occurrences of the secondary motif with the given spacing and orientation to the primary motif.</td> </tr>
          <tr> <td>18</td> <td>total</td> <td>The total number of occurrences of the secondary motif within the margins around the best primary motif occurrence.</td> </tr>
          <tr> <td>19</td> <td>adj_p-value</td> <td>The <i>p</i>-value of the gap and orientation, adjusted for nine combinations of quadrants times the number of gaps tested (as controlled by the <code>-range</code> option).</td> </tr>
          <tr> <td>20</td> <td>p-value</td> <td>The <i>p</i>-value of the gap and orientation adjusted only for the number of gaps tested.</td> </tr>
        </table>
        <br>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'spamo-dumpseqs-tsv':
      html = `
	<p>
	  By specifying the options <code>--dumpseqs</code> or <code>--dumpsigs</code> 
	  you can have SpaMo create TSV (tab-separated values) files
	  describing the motif matches SpaMo used to make the histograms in its HTML output.
	  The files are named 
          '<code>seqs_&lt;primary_motif&gt;_&lt;secondary_db&gt;_&lt;secondary_motif&gt;.txt</code>'.
	  The rows in each file are sorted by sequence name.  
	  The first line contains the (tab-separated) names of the fields.
        </p>
	<p>
	  The names and meanings of each of the fields in <a href="` + site_url + `/doc/spamo-output-format.html#tsv_dumpseqs">SpaMo Dumpseqs TSV Format</a> 
	  are described below. 
        </p>
	<table class="dark" style="width:100%">
	  <tr><th>field</th><th>name</th><th>contents</th></tr>
	  <tr><td>1</td><td>matches</td><td>Trimmed lowercase sequence centered on primary match with matches in uppercase.</td></tr>
	  <tr><td>2</td><td>sec_pos</td><td>Position of the secondary match within the whole sequence.</td></tr>
	  <tr><td>3</td><td>pri_match</td><td>Sequence fragment that the primary matched.</td></tr>
	  <tr><td>4</td><td>pri_strand</td><td>Strand of the primary match (+/-).</td></tr>
	  <tr><td>5</td><td>sec_match</td><td>Sequence fragment that the secondary matched.</td></tr>
	  <tr><td>6</td><td>sec_strand</td><td>Strand of the secondary match (+/-).</td></tr>
	  <tr><td>7</td><td>same_opp</td><td>The primary match on the same (s) or opposite (o) strand as the secondary.</td></tr>
	  <tr><td>8</td><td>down_up</td><td>The secondary match is downstream (d) or upstream (u) of the primary.</td></tr>
	  <tr><td>9</td><td>gap</td><td>The gap between the primary and secondary matches.</td></tr>
	  <tr><td>10</td><td>seq_name</td><td>The name of the sequence.</td></tr>
	  <tr><td>11</td><td>adj_p-value</td><td>The <i>p</i>-value of the bin containing the match, adjusted for the number of bins.</td></tr>
	  <tr><th colspan="3">If the sequence names are in UCSC Genome Browser position
	  format (e.g., "chr5:36715616-36715623"), the following additional fields will be present:</th></tr>
	  <tr><td>12</td><td>pri_bed_chr</td><td>Position of primary match in BED coordinates.</td></tr>
	  <tr><td>13</td><td>pri_bed_start</td><td>"</td></tr>
	  <tr><td>14</td><td>pri_bed_end</td><td>"</td></tr>
	  <tr><td>15</td><td>pri_browser</td><td>Position of primary match in UCSC Genome Browser coordinates.</td></tr>
	  <tr><td>16</td><td>sec_bed_chr</td><td>Position of secondary match in BED coordinates.</td></tr>
	  <tr><td>16</td><td>sec_bed_start</td><td>"</td></tr>
	  <tr><td>16</td><td>sec_bed_end</td><td>"</td></tr>
	  <tr><td>19</td><td>sec_browser</td><td>Position of secondary match in UCSC Genome Browser coordinates.</td></tr>
	</table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'fimo-results-tsv':
      html = ` 
        <p>
          FIMO outputs a tab-separated values (TSV) file ('fimo.tsv') that contains one line for each
          significant match to a motif.
          The lines are sorted in order of decreasing statistical significance (increasing <i>p</i>-value).
          The first line contains the (tab-separated) names of the fields.
          Your command line is given at the end of the file in a comment line starting with the
          character '#'.
        </p>
	<p>
	  The names and meanings of each of the fields in <a href="` + site_url + `/doc/fimo-output-format.html#results-tsv">FIMO Results TSV Format</a> 
	  are described below. 
	</p>
	<table class="dark" style="width:100%" border=1>
	  <tr> <th>field</th> <th>name</th> <th>contents</th> </tr>
          <tr> <td>1</td> <td>motif_id</td> <td> ` + get_doc_text('motif-id') + `</td> </tr>
          <tr> <td>2</td> <td>motif_alt_id</td> <td> ` + get_doc_text('motif-alt-id') + `</td> </tr>
          <tr> <td>3</td> <td>sequence_name</td> <td> ` + get_doc_text('sequence-id') + ` --OR-- the `
	    + get_doc_text('sequence-name') + `</td> </tr>
          <tr> <td>4</td> <td>start</td> <td> ` + get_doc_text('match-start-seq', 'motif occurrence') + ` --OR-- `
            + get_doc_text('match-start-genomic', 'motif occurrence') 
            + get_doc_text('parse-genomic-coord', 'The latter case occurs when FIMO') + `</td> </tr>
          <tr> <td>5</td> <td>stop</td> <td> ` + get_doc_text('match-stop-seq', 'motif occurrence') + ` --OR-- `
            + get_doc_text('match-stop-genomic', 'motif occurrence') 
            + get_doc_text('parse-genomic-coord', 'The latter case occurs when FIMO') + `</td> </tr>
          <tr> <td>6</td> <td>strand</td> <td>The strand '<code>+</code>' indicates the motif matched the forward
	    strand, '<code>-</code>' the reverse strand, and '<code>.</code>'
	    indicates strand is not applicable (as for amino acid sequences).</td> </tr>
          <tr> <td>7</td> <td>score</td> <td>The score for the motif occurrence. `
            + get_doc_text('motif-match-score') + `</td> </tr>
          <tr> <td>8</td> <td>p-value</td> <td> The <i>p</i>-value of the motif occurrence. `
            + get_doc_text('motif-match-p-value') + `</td> </tr>
          <tr> <td>9</td> <td>q-value</td> <td>The q-value of the motif occurrence. `
            + get_doc_text('bh-q-value', 'FIMO') + ` <b>Note:</b> This column is empty if you used the <span class=popt>--text</span> switch.</td> </tr>
          <tr> <td>10</td> <td>sequence</td> <td>The region of the sequence matched to the motif.</td> </tr>
        </table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'fimo-results-gff3':
      html = `
        <p>
          FIMO outputs a GFF3</a> file ('fimo.gff') that contains one line for each
          significant match to a motif.
        </p>
        <p>
	  The GFF3 format is described <a href="http://gmod.org/wiki/GFF3">here</a>.
	  The 'score' reported in the FIMO GFF3 output</a> (in column 6) is<br/>
	  &nbsp;&nbsp;&nbsp;&nbsp;<code>min(1000, -10*(log10(pvalue)))</code>, <br/>
	  and the 'display name' ('Name=' tag in column 9) is composed of the contents of three 
	  fields as follows <br/>
	  &nbsp;&nbsp;&nbsp;&nbsp;<code>&lt;motif_id&gt;_&lt;sequence_name&gt;&lt;strand&gt;</code>.
        </p>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'gomo-results-tsv':
      html = `
        <p>
          GOMo outputs a tab-separated values (TSV) file ('gomo.tsv') that contains one line for each 
	  motif-GO-term pair found to be significantly enriched.
          The lines are grouped by motif and sorted in order of decreasing statistical significance.
          The first line contains the (tab-separated) names of the fields.
          Your command line is given at the end of the file in a comment line starting with the character '#'.
        </p>
        <p>
          The names and meanings of each of the fields in <a href="` + site_url + `/doc/gomo-output-format.html#results-tsv">GOMo Results TSV Format</a>
          are described below.
        </p>
        <table class="dark" style="width:100%" border=1>
          <tr> <th>field</th> <th>name</th> <th>contents</th> </tr>
          <tr> <td>1</td> <td>Motif_Identifier</td> <td> ` + get_doc_text('motif-id') + ` </td> </tr> 
          <tr> <td>2</td> <td>GO_Term_Identifier</td> <td> ` + get_doc_text('gomo-go-term') + ` </td> </tr>
          <tr> <td>3</td> <td>GOMo_Score</td> <td> ` + get_doc_text('gomo-score') + ` </td> </tr>
          <tr> <td>4</td> <td>p-value</td> <td> ` + get_doc_text('gomo-p-value') + ` </td> </tr>
          <tr> <td>5</td> <td>q-value</td> <td> ` + get_doc_text('bh-q-value', 'GOMo') + ` </td> </tr>
        </table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'gomo-results-xml':
      html = `
	<p>
	GOMo outputs an XML file ('gomo.xml') with the following format.
	</p>
	<table class="bordertable" border="1">
	  <tr>
	    <th>Tag</th><th>Child of</th><th>Description</th>
	  </tr>
	  <tr>
	    <td >&lt;gomo&gt;</td><td >Nothing</td>
	    <td>
	      Information about this run of GOMo.
	      <ul>
		<li>version - The version of GOMo that generated the XML file.</li>
		<li>release - The release date of the version that generated the XML.</li>
	      </ul>
	    </td>
	  </tr>
	  <tr>
	    <td >&lt;program&gt;</td>
	    <td >&lt;gomo&gt;</td>
	    <td>
	      Information about the state of the program when it ran.<br />
	      <ul>
		<li>name - name of the program.</li>
		<li>cmd - the command line passed to the program.</li>
		<li>gene_url - the url used to lookup further information on the gene ids. 
		The url has ampersands (&amp;) converted into <b>&amp;amp;</b> and the place where
		  the gene ID should be replaced by <b>!!GENEID!!</b> .</li>
		<li>outdir - the output directory that the program wrote to.</li>
		<li>clobber - true if GOMo was allowed to overwrite the output directory.</li>
		<li>text_only - true if GOMo wrote to stdout, in which case this file would
		  not exist so it must be false.</li>
		<li>use_e_values - true if GOMo used <i>E</i>-values (converted from <i>p</i>-values) as 
		  input scores, false if GOMo used gene scores.</li>
		<li>score_e_thresh - if GOMo used <i>E</i>-values then this is the threshold that 
		  GOMo assumed the worst <i>E</i>-value (<i>p</i>-value = 1.0) for the gene to smooth out noise.</li>
		<li>min_gene_count - the minimum number of genes that a GO term was annotated 
		  with before GOMo would calculate a score for it.</li>
		<li>motifs - if present then a space delimited list of the motifs that GOMo
		  calculated a score for, otherwise GOMo scored all motifs.</li>
		<li>shuffle_scores - the number of times GOMo generated a shuffled mapping of
		  gene id to gene id to be used to generate scores from the null model.</li>
		<li>q_threshold - GOMo filtered the results to only show those with a better
		(smaller) q-value.</li>
	      </ul>
	    </td>
	  </tr>
	  <tr>
	    <td>&lt;gomapfile&gt;</td>
	    <td>&lt;program&gt;</td>
	    <td>
	      Information about the GO mapping file.<br />
	      <ul>
		<li>path - the path to the mapping file.</li>
	      </ul>
	    </td>
	  </tr>
	  <tr>
	    <td>&lt;seqscorefile&gt;</td>
	    <td>&lt;program&gt;</td>
	    <td>
	      Information about the sequence scoring file.<br />
	      <ul>
		<li>path - the path to the sequence scoring file.</li>
	      </ul>
	    </td>
	  </tr>
	  <tr>
	    <td>&lt;motif&gt;</td>
	    <td>&lt;gomo&gt;</td>
	    <td>
	      Information about the motif.<br />
	      <ul>
		<li>id - the motif identifier.</li>
		<li>genecount - the number of scored sequences that were used to compute the result.</li>
	      </ul>
	    </td>
	  </tr>
	  <tr>
	    <td>&lt;goterm&gt;</td>
	    <td>&lt;motif&gt;</td>
	    <td>
	      Information about the GO term.<br />
	      <ul>
		<li>id - the GO identifier.</li>
		<li>score - the geometric mean across all species of the rank-sum test <i>p</i>-value.</li>
		<li>pvalue - the empirically calculated <i>p</i>-value.</li>
		<li>qvalue - the empirically calculated q-value.</li>
		<li>annotated - the number of genes annotated with the go term.</li>
		<li>group - the subgroup that the term belongs to. For the Gene Ontology 
		    b = biological process, c = cellular component and m = molecular function.</li>
		<li>nabove - the number of more general terms that link to this one.</li>
		<li>nbelow - the number of more specific terms that link from this one.</li>
		<li>implied - is the go term implied by other significant go terms? 
		  Allows values 'y', 'n' or 'u' (default) for yes, no or unknown.</li>
		<li>description - the GO term description.</li>
	      </ul>
	    </td>
	  </tr>
	  <tr>
	    <td>&lt;gene&gt;</td>
	    <td>&lt;goterm&gt;</td>
	    <td>
	      Information about the GO term's annotated genes for the primary species.<br />
	      <ul>
		<li>id - the gene identifier.</li>
		<li>rank - the rank of the scored gene.</li>
	      </ul>
	    </td>
	  </tr>
        </table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'mcast-results-tsv':
      html = ` 
        <p>
          MCAST outputs a tab-separated values (TSV) file ('mcast.tsv') that contains one line for each
          significant match of a cluster of motifs to a sequence region.
          The lines are sorted in order of decreasing statistical significance (increasing <i>p</i>-value).
          The first line contains the (tab-separated) names of the fields.
          Your command line is given at the end of the file in a comment line starting with the
          character '#'.
        </p>
	<p>
	  The names and meanings of each of the fields in 
          <a href="` + site_url + `/doc/mcast-output-format.html#results-tsv">MCAST Results TSV Format</a> 
	  are described below. 
	</p>
	<table class="dark" style="width:100%" border=1>
	  <tr> <th>field</th> <th>name</th> <th>contents</th> </tr>
          <tr> <td>1</td> <td>pattern_name</td> <td>A unique name that MCAST generates for the match. </td> </tr>
          <tr> <td>2</td> <td>sequence_name</td> <td> ` + get_doc_text('sequence-id') + ` --OR-- the `
	    + get_doc_text('sequence-name') + `</td> </tr>
          <tr> <td>3</td> <td>start</td> <td> ` + get_doc_text('match-start-seq', 'matched sequence region') + ` --OR-- `
            + get_doc_text('match-start-genomic', 'motif occurrence')
            + get_doc_text('parse-genomic-coord', 'The latter case occurs when MCAST') + `</td> </tr>
          <tr> <td>4</td> <td>stop</td> <td> ` + get_doc_text('match-stop-seq', 'matched sequence region') + ` --OR-- `
            + get_doc_text('match-stop-genomic', 'motif occurrence')
            + get_doc_text('parse-genomic-coord', 'The latter case occurs when MCAST') + `</td> </tr>
          <tr> <td>5</td> <td>score</td> <td> ` + get_doc_text('mcast-cluster-score') + ` </td> </tr>
          <tr> <td>6</td> <td>p-value</td> <td> ` + get_doc_text('mcast-cluster-p-value') + ` </td> </tr>
          <tr> <td>7</td> <td>E-value</td> <td> ` + get_doc_text('mcast-cluster-E-value') + ` </td> </tr>
          <tr> <td>8</td> <td>q-value</td> <td> ` + get_doc_text('bh-q-value', 'MCAST') + ` 
	    <b>Note:</b> This column is empty if you used the <span class=popt>--text</span> switch.</td> </tr>
          <tr> <td>9</td> <td>sequence</td> <td>The region of the sequence matched to the motif cluster.</td> </tr>
        </table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'mcast-results-gff3':
      html = `
        <p>
          MCAST outputs a GFF3</a> file ('mcast.gff') that contains one line for each
          significant match to a motif cluster.
        </p>
        <p>
	  The GFF3 format is described <a href="http://gmod.org/wiki/GFF3">here</a>.
	  The 'score' reported in the MCAST GFF3 output</a> (in column 6) is<br/>
	  &nbsp;&nbsp;&nbsp;&nbsp;<code>min(1000, -10*(log10(pvalue)))</code>, <br/>
	  and the 'unique identifier' ('ID=' tag in column 9) is the value of the
	  &lt;pattern_name&gt; field in the MCAST results TSV format.  Following the
	  unique identifier in column 9, the <i>p</i>-value, <i>E</i>-value and q-value
	  of the match are given.
        </p>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    case 'tomtom-results-tsv':
      html = `
	<p>
          Tomtom outputs a tab-separated values (TSV) file ('tomtom.tsv') that contains one line for each motif
	  found to be significantly enriched.
	  The lines are grouped by query motif and sorted in order of decreasing statistical significance.
          The first line contains the (tab-separated) names of the fields.
	  Your command line is given at the end of the file in a comment line starting with the character '#'.
	</p>
	<p>
	  The names and meanings of each of the fields in <a href="` + site_url + `/doc/tomtom-output-format.html#results-tsv">Tomtom Results TSV Format</a> 
	  are described below. 
	</p>
	<table class="dark" style="width:100%" border=1>
	  <tr> <th>field</th> <th>name</th> <th>contents</th> </tr>
          <tr> <td>1</td> <td>Query_ID</td> <td> ` + get_doc_text('motif-id', 'query') + `</td> </tr>
          <tr> <td>2</td> <td>Target_ID</td> <td> ` + get_doc_text('motif-id', 'target') + `</td> </tr>
          <tr> <td>3</td> <td>Optimal_offset</td> <td> ` + get_doc_text('tomtom-offset') + `</td> </tr>
          <tr> <td>4</td> <td>p-value</td> <td> ` + get_doc_text('tomtom-p-value') + `</td> </tr>
          <tr> <td>5</td> <td>E-value</td> <td> ` + get_doc_text('tomtom-E-value') + `</td> </tr>
          <tr> <td>6</td> <td>q-value</td> <td> ` + get_doc_text('bh-q-value', 'Tomtom') + `</td> </tr>
          <tr> <td>7</td> <td>Overlap</td> <td> ` + get_doc_text('tomtom-overlap') + `</td> </tr>
          <tr> <td>8</td> <td>Query_consensus</td> <td>A consensus sequence
                computed from the letter frequencies in the query motif
                (as described <a href="#consensus_doc">below</a>).</td> </tr>
          <tr> <td>9</td> <td>Target_consensus</td> <td>A consensus sequence
                computed from the letter frequencies in the target motif
                (as described <a href="#consensus_doc">below</a>).</td> </tr>
          <tr> <td>10</td> <td>Orientation</td> <td> ` + get_doc_text('tomtom-orientation', "<br>A value of '+' means that the target motif is as it appears in the database. A value of '-' means that the reverse complement of the target motif is shown.") + `</td> </tr>
        </table>
      `;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
      break;

    default:
      html = "Error--Unrecognized doc_type: " + doc_type;
      document.getElementById(id).insertAdjacentHTML('beforeend', html);
  }
} // print_doc

//
// Function to return the HTML text of a given type.
// This function can be used directly to document the output format (xx-output-format.html)
// and indirectly via print_doc_para for help pop-ups in the actual output HTML,
// to prevent duplication of documentation.
//
function get_doc_text(doc_type, extra) {
  var html;
  if (extra == undefined) {extra = ""};

  switch (doc_type) {

    // shared
    case 'motif-db':
      return(`
	The name of a file of motifs ("motif database file") that contains ` + extra + `
      `);
    case 'motif-id':
      return(`
	The name of the ` + extra + ` motif, which is unique in the motif database file.
      `);
    case 'motif-alt-id':
      return(`
	An alternate name for the ` + extra + ` motif, which may be provided in the motif database file.
      `);
    case 'motif-width':
      return(`
	The width of the motif. No gaps are allowed in motifs supplied to ` + extra + `
        as it only works for motifs of a fixed width.
      `);
    case 'motif-cons':
      return(`
	A consensus sequence computed from the ` + extra + ` motif (as described <a href="#consensus_doc">below</a>).
      `);
    case 'motif-match-score':
     return(`
	The score for the match of a position in a sequence to a motif is
	computed by summing the appropriate entry from each column of the
	position-dependent scoring matrix that represents the motif. ` + extra + `
     `);
    case 'motif-match-p-value':
      return(`
	The <i>p</i>-value of a motif match is the probability of a single random
	subsequence of the length of the motif
	<a href="javascript:help_refine('pop_motif_match_score')">scoring</a>
	at least as well as the observed match.
      `);
    case 'bh-q-value-method':
      return(`
        ` + extra + ` estimates q-values from all the match <i>p</i>-values 
	using the method proposed by Benjamini & Hochberg (<i>Journal of the Royal Statistical Society B</i>, 57:289-300, 1995).
	See also Storey JD, Tibshirani R. Statistical significance for
	genome-wide studies, <i>Proc. Natl. Acad. Sci. USA</i> (2003) <b>100</b>:9440&ndash;9445.
      `);
    case 'bh-q-value':
      return(`
	The minimum False Discovery Rate (FDR) required to consider this match significant.</br>` + get_doc_text('bh-q-value-method', extra) + `
      `);
    case 'sdb-name':
      return(`
	The name of the (FASTA) sequence database file.
      `);
    case 'sdb-psp':
      return(`
	The name of the position specific priors (PSP) file.
      `);
    case 'sdb-dist':
      return(`
	The name of the binned distribution of priors file.
      `);
    case 'sdb-count':
      return(`
	The number of sequences in the database.
      `);
    case 'sdb-letters':
      return(`
	The number of letters in the sequence database.
      `);
    case 'lastmod':
      return(`
	The date of the last modification to the ` + extra + ` database.
      `);
    case 'sequence-id':
      return(`
        The identifier of the sequence (from the FASTA sequence header line)` + extra + `
      `);
    case 'sequence-name':
      return(`
	` + extra + `name of the sequence extracted from the sequence identifier (in the FASTA sequence header line).<br>
	When you use the <code>--parse-genomic--coord</code> option, the sequence name ends at the
	first colon ':' (if any) present in the sequence\'s FASTA identifier.  Typically this is the
	chromosome or contig name.  With the <code>--parse-genomic--coord</code> option,
	the start and stop positions are in 0-based coordinates relative to the sequence start given 
	in the FASTA sequence identifier (just after the sequence name).</td> </tr>
      `);
    case 'sequence-desc':
      return(`
        The description appearing after the identifier of the sequence in the FASTA header line.
      `);
    case 'sequence-name':
    case 'alph-name':
      return(`
	The name of the alphabet symbol.
      `);
    case 'alph-bg':
      return(`
	The frequency of the alphabet symbol as defined by the background model.
      `);
    case 'match-start-seq':
      return(`
	The start position of the ` + extra + `; 1-based sequence coordinates.
      `);
    case 'match-stop-seq':
      return(`
	The end position of the ` + extra + `; 1-based sequence coordinates.
      `);
    case 'match-start-genomic':
      return(`
	The start position of the ` + extra + `; genomic coordinates.
      `);
    case 'match-stop-genomic':
      return(`
	The end position of the ` + extra + `; genomic coordinates.
      `);
    case 'parse-genomic-coord':
      return(`
	` + extra + ` was run with the <code>--parse-genomic-coord</code> option
	and has split the sequence identifier into sequence name, sequence start and sequence end 
	in genomic coordinates.
      `);

    // AME
    case 'ame-pvalue':
      return(`
	The optimal enrichment <i>p</i>-value of the motif according to the statistical test;
	not adjusted for multiple tests.
      `);
    case 'ame-adj-pvalue':
      return(`
	The optimal enrichment <i>p</i>-value of the motif according to the statistical test,
        adjusted for multiple tests using a Bonferroni correction. ` + extra + `
	If the best <i>p</i>-value is <i>p</i> before adjustment,
        and the number of multiple tests is <i>n</i>, then the adjusted
        <i>p</i>-value is 1 - (1-<i>p</i>)<i><sup>n</sup></i>.
      `);
    case 'ame-evalue':
      return(`
	The expected number of motifs that would be as enriched in the
        (primary) sequences as this one.  The <i>E</i>-value is the adjusted <i>p</i>-value
        multiplied by the number of motifs in the motif file(s).
      `);

    // CentriMo
    case 'centrimo-adj-pvalue':
      return(`
        The statistical significance of the enrichment of the motif, adjusted for multiple tests. ` + extra + `
      `);
    case 'centrimo-evalue':
      var evalue_html = `
        at least one region as enriched for best matches to the motif as the reported region
	(or would have optimal average distance to the sequence center as low as observed, 
	if you used the <code>--cd</code> option).
      `;
      return(`
	The expected number motifs that would have ` + (extra ? extra : evalue_html) + `
	The <i>E</i>-value is the adjusted <i>p</i>-value multiplied by the number of motifs in the
	input files(s).</td> </tr>
      `);
    case 'centrimo-bin-width':
      return(`
        The width (in sequence positions) of the most enriched region (default),
        <b>or</b> two times the average distance between the center of the best site
        and the sequence center if you used the option <code>--cd</code>.
        A best match to the motif is counted as being in the region if
        the center of the motif falls in the region.
      `);
    case 'centrimo-sites-in-bin':
      return(`
	The number of (positive) sequences whose best match to the motif `
	+ (extra ? extra : "falls in the reported region (default) or anywhere in the sequence (if you used the option <code>--cd</code>).") + `
	<br>Note: This number may be less than the number of
	(positive) sequences that have a best match in the region.
	The reason for this is that a sequence may have many matches that score
	equally best.  If <i>n</i> matches have the best score in a sequence, 1/<i>n</i> is added to the
	appropriate bin for each match.</td> </tr>
      `);
    case 'centrimo-mult-tests':
      return(`
	This is the number of multiple tests (<i>n</i>) done for this motif.
	It was used to adjust the <i>p</i>-value of a region for
	multiple tests using the formula:
	<br>&nbsp;&nbsp;
	  <i>p</i>' = 1 - (1-<i>p</i>)<sup><i>n</i></sup> where <i>p</i> is the unadjusted <i>p</i>-value.
	<br>
	The number of multiple tests is the number of regions
	considered times the number of score thresholds considered.
	It depends on the motif length, sequence length, and the type of
	optimizations being done (central enrichment, local enrichment, central distance or
	score optimization).</td> </tr>
      `);

    // GOMo
    case 'gomo-go-term':
      return(`
        The Gene Ontology Consortium term for a specific role or locality.
        Used for annotating genes with their functions.
      `);
    case 'gomo-score':
      return( `
	A score generated as the <a href="https://en.wikipedia.org/wiki/Geometric_mean">
	geometric mean</a> of <a href="https://en.wikipedia.org/wiki/Mann-Whitney_U_test">rank-sum test(s)</a> 
	for the particular Gene Ontology term. The two groups compared by the rank-sum test are scores of genes annotated 
	with the GO term and scores of genes not annotated with the GO term.</td> </tr>
      `);
    case 'gomo-p-value':
      return( `
	An empirically generated <i>p</i>-value for the enrichment of the GO term.<br>
	The null hypothesis is that by shuffling the labels on gene scores, 
	any possible association between the set of genes that a GO term annotates is destroyed. 
	A large number of scores are generated using the null hypothesis and the number of null 
	hypothesis scores that are better than each of the real scores is summed and then divided 
	by the total null hypothesis scores generated to calculate a <i>p</i>-value.</td> </tr>
      `);

    // MAST

    // MCAST
    case 'mcast-cluster-score':
      return( `
        The score that the hidden Markov model created by MCAST assigned to the motif cluster.<br>
        This is the sum of the scores of the individual motif matches in the cluster, plus a
        gap penalty, <i>g</i>, multiplied by the total size of the inter-motif gaps
        in the cluster.  Individual motif match scores are log2(<i>P(s)/p</i>), where <i>s</i> is the 
        <a href="javascript:help_refine('pop_motif_match_score')">log-odds score</a> 
        of the motif match, <i>P(s)</i> is the 
        <a href="javascript:help_refine('pop_motif_match_pvalue')"><i>p</i>-value</a> 
	of the motif match, and <i>p</i> is the user-specified <i>p</i>-value threshold (default: 0.0005).
        <div class="active" id="pop_motif_match_score_act"></div>
        <div class="active" id="pop_motif_match_pvalue_act"></div>
      `);
    case 'mcast-cluster-p-value':
      return( `
	The <i>p</i>-value of the motif cluster score.<br>
	MCAST estimates <i>p</i>-values by fitting an exponential distribution
	to the observed motif cluster scores.
      `);
    case 'mcast-cluster-E-value':
      return( `
        The <i>E</i>-value of the motif cluster score.<br>
	The <i>E</i>-value is an estimate of the <i>number</i> of false positive matches
	with <i>p</i>-values at least as good as this match\'s.  
        MCAST estimates this by multiplying the motif cluster score <i>p</i>-value
        times the (estimated) number of random matches found in the search.
      `);

    // Tomtom
    case 'tomtom-offset':
      return( `
        The offset of the query motif relative to the target motif in the optimal alignment.<br>
 	A positive value indicates the query is shifted right.
      `);
    case 'tomtom-p-value':
      return( `
        The probability that a random motif of the same width as the target would have an 
	optimal alignment with a match score as good or better than the target's.<br>
	Tomtom estimates the <i>p</i>-value using a null model consisting of sampling
	motif columns from all the columns in the set of target motifs.
      `);
    case 'tomtom-E-value':
      return( `
	The expected number of false positives in the matches up to this point.<br>
	Tomtom estimates the <i>E</i>-value by multiplying the <i>p</i>-value by
	the total number of target motifs in all the target databases.
      `);
    case 'tomtom-overlap':
      return( `
	The number of motif columns that overlap in the optimal alignment.
      `);
    case 'tomtom-orientation':
      return( `
	The orientation of the target motif that gave the optimal alignment. ` + extra + `
      `);
  }

} // get_doc_text

//
// Function to replace the innerHTML of element "id" with an HTML paragraph
// containing the text for 'doc_type', which is known to function get_doc_text.
// This function can be used in help pop-ups.
//
function print_doc_para(id, doc_type, extra) {
  html = `<p>` + get_doc_text(doc_type, extra) + `</p>`; 
  document.getElementById(id).insertAdjacentHTML('beforeend', html);
} // print_doc_para

    </script>
    <script type="text/javascript">
function make_alpha_bg_table(alph, freqs) {
  function colour_symbol(index) {
    var span = document.createElement("span");
    span.appendChild(document.createTextNode(alph.get_symbol(index)));
    span.style.color = alph.get_colour(index);
    span.className = "alpha_symbol";
    return span;
  }
  var table, thead, tbody, row, th, span, i;
  // create table
  table = document.createElement("table");
  table.className = "alpha_bg_table";
  // create header
  thead = document.createElement("thead");
  table.appendChild(thead);
  row = thead.insertRow(thead.rows.length);
  if (alph.has_complement()) {
    add_text_header_cell(row, "Name", "pop_alph_name");
    if (freqs != null) add_text_header_cell(row, "Freq.", "pop_alph_freq");
    if (alph.has_bg()) add_text_header_cell(row, "Bg.", "pop_alph_bg");
    add_text_header_cell(row, "");
    add_text_header_cell(row, "");
    add_text_header_cell(row, "");
    if (alph.has_bg()) add_text_header_cell(row, "Bg.", "pop_alph_bg");
    if (freqs != null) add_text_header_cell(row, "Freq.", "pop_alph_freq");
    add_text_header_cell(row, "Name", "pop_alph_name");
  } else {
    add_text_header_cell(row, "");
    add_text_header_cell(row, "Name", "pop_alph_name");
    if (freqs != null) add_text_header_cell(row, "Freq.", "pop_alph_freq");
    if (alph.has_bg()) add_text_header_cell(row, "Bg.", "pop_alph_bg");
  }
  // add alphabet entries
  tbody = document.createElement("tbody");
  table.appendChild(tbody);
  if (alph.has_complement()) {
    for (i = 0; i < alph.get_size_core(); i++) {
      var c = alph.get_complement(i);
      if (i > c) continue;
      row = tbody.insertRow(tbody.rows.length);
      add_text_cell(row, alph.get_name(i));
      if (freqs != null) add_text_cell(row, "" + freqs[i]);
      if (alph.has_bg()) add_text_cell(row, "" + alph.get_bg_freq(i));
      add_cell(row, colour_symbol(i)); 
      add_text_cell(row, "~");
      add_cell(row, colour_symbol(c)); 
      if (alph.has_bg()) add_text_cell(row, "" + alph.get_bg_freq(c));
      if (freqs != null) add_text_cell(row, "" + freqs[c]);
      add_text_cell(row, alph.get_name(c));
    }
  } else {
    for (i = 0; i < alph.get_size_core(); i++) {
      row = tbody.insertRow(tbody.rows.length);
      add_cell(row, colour_symbol(i)); 
      add_text_cell(row, alph.get_name(i));
      if (freqs != null) add_text_cell(row, "" + freqs[i]);
      if (alph.has_bg()) add_text_cell(row, "" + alph.get_bg_freq(i));
    }
  }
  return table;
}


    </script>
    <script type="text/javascript">
pre_load_setup();
var ame_alphabet = new Alphabet(data.alphabet, data.background['frequencies']);

/*
 * name_from_source
 *
 * Makes a file name more human friendly to read.
 */
function name_from_source(source) {
  if (source == "-") {
    return "-"
  }
  //assume source is a file name
  var file = source.replace(/^.*\/([^\/]+)$/,"$1");
  var noext = file.replace(/\.[^\.]+$/, "");
  return noext.replace(/_/g, " ");
}

/*
 * pre_load_setup
 *
 *  Sets up initial variables which may be
 *  required for the HTML document creation.
 */
function pre_load_setup() {
  var seq_db = data['sequence_db'];
  if (!seq_db['name']) seq_db['name'] = name_from_source(seq_db['source']);
  // get the names of the motif databases
  var dbs = data['motif_dbs'];
  var motif_count = 0;
  for (var i = 0; i < dbs.length; i++) {
    var db = dbs[i];
    if (!db['name']) db['name'] = name_from_source(db['source']);
    motif_count += db['count'];
  }
  dbs['count'] = motif_count;
}

/*
 * page_loaded
 *
 * Called when the page has loaded for the first time.
 */
function page_loaded() {
  post_load_setup();
}

/*
 * page_loaded
 *
 * Called when a cached page is reshown.
 */
function page_shown(e) {
  if (e.persisted) post_load_setup();
}

/*
 * post_load_setup
 *
 * Setup state that is dependant on everything having been loaded already.
 */
function post_load_setup() {
  update_scroll_pad();
}

/*
 * add_cell
 *
 * Add a cell to the table row.
 */
function add_cell(row, node, cls) {
  var cell = row.insertCell(row.cells.length);
  if (node) cell.appendChild(node);
  if (cls) cell.className = cls;
}

/*
 * add_text_cell
 *
 * Add a text cell to the table row.
 */
function add_text_cell(row, text, cls) {
  var node = null;
  if (text) node = document.createTextNode(text);
  add_cell(row, node, cls);
}

function toExp(num, prec) {
  if (typeof(num) === "number") return num.toExponential(prec);
  return "";
}

/*
 * make_results_table
 *
 * Add the results to the results table.
 */
function make_results_table() {
  // clear the table and add the items
  var tbl = $("results");
  // If the partition is fixed, the number of positives doesn't change.
  if (data['options']['fix_partition']) {
    tbl.className += " no_pos no_neg no_p_thresh";
  }
  if (data['options']['pvalue_method'] == "Fisher's exact test") {
    // Fisher only shows the optimal score if the partition is fixed.
    if (! data['options']['fix_partition']) tbl.className += " no_tp_thresh";
  } else {
    // Non-fisher methods only use the positive sequences.
    tbl.className += " no_neg"
    // Non-fisher methods don't classify the sequences.
    tbl.className += " no_tp_fp";
    // Non-fisher methods don't show a score threshold.
    tbl.className += " no_tp_thresh";
  }
  if (data['options']['pvalue_method'] == "Spearman's correlation coefficient") { 
    tbl.className += " spearman";
  } else if (data['options']['pvalue_method'] == "linear regression") { 
    tbl.className += " linreg";
  } else {
    tbl.className += " other";
  }
  var tbody = tbl.tBodies[0];
  while (tbody.rows.length > 0) {
    tbody.deleteRow(0);
  }
  var motif_dbs = data['motif_dbs'];
  var motifs = data['motifs'];
  // add the new rows to the table
  for (var i = 0; i < motifs.length; i++) {
    var motif = motifs[i];
    var row = tbody.insertRow(tbody.rows.length);
    row.id = "db_" + motif['db'] + "_motif_" + motif['id'];
    row['motif'] = motif;
    var canvas = document.createElement('canvas');
    var pspm = new Pspm(motif['pwm'], motif['id'], 0, 0, 
	motif['motif_nsites'], motif['motif_evalue']);
    var logo = logo_1(ame_alphabet, '', pspm);
    draw_logo_on_canvas(logo, canvas, false, 0.5);
    add_cell(row, canvas, 'col_logo');
    add_text_cell(row, motif_dbs[motif['db']]['name'], 'col_db');
    add_cell(row, make_link(motif['id'], motif['url']), 'col_id');
    add_text_cell(row, motif['alt'], 'col_name');
    add_text_cell(row, motif['corrected_pvalue'], 'col_corrected_pvalue');
    add_text_cell(row, motif['evalue'], 'col_evalue');
    add_text_cell(row, (motif['positive_thresh']).toFixed(2), 'col_p_thresh');
    add_text_cell(row, motif['pos'].toFixed(0), 'col_pos');
    add_text_cell(row, motif['neg'].toFixed(0), 'col_neg');
    add_text_cell(row, (motif['true_positive_thresh']).toFixed(2), 'col_tp_thresh');
    add_text_cell(row, motif['tp'].toFixed(0) + ' (' + (motif['pos']!=0 ? (100*motif['tp']/motif['pos']).toFixed(1) : '0') + "%)", 'col_tp');
    add_text_cell(row, motif['fp'].toFixed(0) + ' (' + (motif['neg']!=0 ? (100*motif['fp']/motif['neg']).toFixed(1) : '0') + "%)", 'col_fp');
    add_text_cell(row, toExp(motif['pearsons_rho'], 2), 'col_pearsons_rho');
    add_text_cell(row, toExp(motif['spearmans_rho'], 2), 'col_spearmans_rho');
    add_text_cell(row, toExp(motif['mean_square_error'], 2), 'col_mean_square_error');
    add_text_cell(row, toExp(motif['m'], 2), 'col_m');
    add_text_cell(row, toExp(motif['b'], 2), 'col_b');
  }
}

    </script>
    <style type="text/css">
table.inputs, table.alpha_bg_table {
  margin-top: 20px;
  border-collapse:collapse;
}
table.inputs * td, table.inputs * th, table.alpha_bg_table * td, table.alpha_bg_table * th {
  padding-left: 15px;
  padding-right: 15px;
  padding-top: 1px;
  padding-bottom: 1px;
}
table.results {
  border-collapse:collapse;
}
table.results * td, table.results * th {
  padding-left: 15px;
  padding-right: 15px;
  padding-top: 1px;
  padding-bottom: 1px;
}

table.results * td.logo {
  height: 20px;
}

table.no_tp_fp .col_fp,
table.no_tp_fp .col_tp,
table.no_p_thresh .col_p_thresh,
table.no_tp_thresh .col_tp_thresh,
table.no_pos .col_pos,
table.no_neg .col_neg,
table.linreg .col_spearmans_rho,
table.spearman .col_pearsons_rho,
table.spearman .col_mean_square_error, 
table.spearman .col_m, 
table.spearman .col_b,
table.other .col_pearsons_rho,
table.other .col_spearmans_rho,
table.other .col_mean_square_error, 
table.other .col_m, 
table.other .col_b {
  display: none;
}

div.pop_content {
  position:absolute;
  z-index:1;
  width:300px;
  padding: 5px;
  background: #E4ECEC;
  font-size: 12px;
  font-family: Arial;
  border-style: double;
  border-width: 3px;
  border-color: #AA2244;
  display:none;
}

    </style>
    <style type="text/css">
/* The following is the content of meme.css */
body { background-color:white; font-size: 12px; font-family: Verdana, Arial, Helvetica, sans-serif;}

div.help {
  display: inline-block;
  margin: 0px;
  padding: 0px;
  width: 12px;
  height: 13px;
  cursor: pointer;
  background-image: url(data:image/gif;base64,R0lGODlhDAANAIABANR0AP///yH5BAEAAAEALAAAAAAMAA0AAAIdhI8Xy22MIFgv1DttrrJ7mlGNNo4c+aFg6SQuUAAAOw==);
}

div.help:hover {
  background-image: url(data:image/gif;base64,R0lGODlhDAANAKEAANR0AP///9R0ANR0ACH+EUNyZWF0ZWQgd2l0aCBHSU1QACH5BAEAAAIALAAAAAAMAA0AAAIdDGynCe3PgoxONntvwqz2/z2K2ImjR0KhmSIZUgAAOw==);
}

p.spaced { line-height: 1.8em;}

span.citation { font-family: "Book Antiqua", "Palatino Linotype", serif; color: #004a4d;}

p.pad { padding-left: 30px; padding-top: 5px; padding-bottom: 10px;}

td.jump { font-size: 13px; color: #ffffff; background-color: #00666a;
  font-family: Georgia, "Times New Roman", Times, serif;}

a.jump { margin: 15px 0 0; font-style: normal; font-variant: small-caps;
  font-weight: bolder; font-family: Georgia, "Times New Roman", Times, serif;}

h2.mainh {font-size: 1.5em; font-style: normal; margin: 15px 0 0;
  font-variant: small-caps; font-family: Georgia, "Times New Roman", Times, serif;}

h2.line {border-bottom: 1px solid #CCCCCC; font-size: 1.5em; font-style: normal;
  margin: 15px 0 0; padding-bottom: 3px; font-variant: small-caps;
  font-family: Georgia, "Times New Roman", Times, serif;}

h4 {border-bottom: 1px solid #CCCCCC; font-size: 1.2em; font-style: normal;
  margin: 10px 0 0; padding-bottom: 3px; font-family: Georgia, "Times New Roman", Times, serif;}

h5 {margin: 0px}

a.help { font-size: 9px; font-style: normal; text-transform: uppercase;
  font-family: Georgia, "Times New Roman", Times, serif;}

div.pad { padding-left: 30px; padding-top: 5px; padding-bottom: 10px;}

div.pad1 { margin: 10px 5px;}

div.pad2 { margin: 25px 5px 5px;}
h2.pad2 { padding: 25px 5px 5px;}

div.pad3 { padding: 5px 0px 10px 30px;}

div.box { border: 2px solid #CCCCCC; padding:10px; overflow: hidden;}

div.bar { border-left: 7px solid #00666a; padding:5px; margin-top:25px; }

div.subsection {margin:25px 0px;}

img {border:0px none;}

th.majorth {text-align:left;}
th.minorth {font-weight:normal; text-align:left; width:8em; padding: 3px 0px;}
th.actionth {font-weight:normal; text-align:left;}

.explain h5 {font-size:1em; margin-left: 1em;}

div.doc {margin-left: 2em; margin-bottom: 3em;}

th.trainingset {
  border-bottom: thin dashed black; 
  font-weight:normal; 
  padding:0px 10px;
}
div.pop_content {
  position:absolute;
  z-index:50;
  width:300px;
  padding: 5px;
  background: #E4ECEC;
  font-size: 12px;
  font-family: Arial;
  border-style: double;
  border-width: 3px;
  border-color: #AA2244;
  display:none;
}

div.pop_content > *:first-child {
  margin-top: 0px;
}

div.pop_content h1, div.pop_content h2, div.pop_content h3, div.pop_content h4, 
div.pop_content h5, div.pop_content h6, div.pop_content p {
  margin: 0px;
}

div.pop_content p + h1, div.pop_content p + h2, div.pop_content p + h3, 
div.pop_content p + h4, div.pop_content p + h5, div.pop_content p + h6 {
  margin-top: 5px;
}

div.pop_content p + p {
  margin-top: 5px;
}

div.pop_content > *:last-child {
  margin-bottom: 0px;
}

div.pop_content div.pop_close {
  /* old definition */
  float:right;
  bottom: 0;
}

div.pop_content span.pop_close, div.pop_content span.pop_back {
  display: inline-block;
  border: 2px outset #661429;
  background-color: #CCC;
  padding-left: 1px;
  padding-right: 1px;
  padding-top: 0px;
  padding-bottom: 0px;
  cursor: pointer;
  color: #AA2244; /*#661429;*/
  font-weight: bold;
}

div.pop_content span.pop_close:active, div.pop_content span.pop_back:active {
  border-style: inset;
}

div.pop_content span.pop_close {
  float:right;
  /*border: 2px outset #AA002B;*/
  /*color: #AA2244;*/
}

div.pop_content:not(.nested) .nested_only {
  display: none;
}

div.pop_back_sec {
  margin-bottom: 5px;
}

div.pop_close_sec {
  margin-top: 5px;
}

table.hide_advanced tr.advanced {
  display: none;
}
span.show_more {
  display: none;
}
table.hide_advanced span.show_more {
  display: inline;
}
table.hide_advanced span.show_less {
  display: none;
}


/*****************************************************************************
 * Program logo styling
 ****************************************************************************/
div.prog_logo {
  border-bottom: 0.25em solid #0f5f60;
  height: 4.5em;
  width: 24em;
  display:inline-block;
}
div.prog_logo img {
  float:left;
  width: 4em;
  border-style: none;
  margin-right: 0.2em;
}
div.prog_logo h1, div.prog_logo h1:hover, div.prog_logo h1:active, div.prog_logo h1:visited {
  margin:0;
  padding:0;
  font-family: Arial, Helvetica,  sans-serif;
  font-size: 3.2em;
  line-height: 1em;
  vertical-align: top;
  display: block;
  color: #026666;
  letter-spacing: -0.06em;
  text-shadow: 0.04em 0.06em 0.05em #666;
}
div.prog_logo h2, div.prog_logo h2:hover, div.prog_logo h2:active, div.prog_logo h2:visited {
  display: block;
  margin:0;
  padding:0;
  font-family: Helvetica, sans-serif;
  font-size: 0.9em;
  line-height: 1em;
  letter-spacing: -0.06em;
  color: black;
}

div.big.prog_logo {
  font-size: 18px;
}


    </style>
  </head>
  <body onload="page_loaded()" onpageshow="page_shown(event)" onresize="update_scroll_pad()">
    <!-- Help popups -->
    <div class="pop_content" id="pop_alph_name">
      <script>print_doc_para('pop_alph_name', 'alph-name', 'The ');</script>
      <div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_alph_bg">
      <script>print_doc_para('pop_alph_bg', 'alph-bg', 'The ');</script>
      <div class="pop_close">[<a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_results">
      <script>print_doc("pop_results", "ame-results-header");</script>
      <script>print_doc("pop_results", "ame-results-description");</script>
      <script>print_doc("pop_results", "motif-consensus");</script>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_sequences">
      <script>print_doc("pop_sequences", "ame-sequences-tsv");</script>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_db">
      <script>print_doc_para("pop_db", "motif-db", "motif.");</script>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_id">
      <script>print_doc_para("pop_id", "motif-id");</script>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_name">
      <script>print_doc_para("pop_name", "motif-alt-id");</script>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_pos">
      <p>Labeling this number of sequences as 'positive' was found by AME
	 to be optimal (or your primary sequence file contains
	 this number of sequences and you also provided a control sequence file).  
	 The sequences were chosen as described above under "Labeling positives".
      </p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_neg">
      <p>Labeling this number of sequences as 'negative' was found by AME
	 to be optimal (or you provided a control sequence file containing
	 this number of sequences).  The sequences are those remaining after the
	 'positive' sequences were chosen as described above under "Labeling positives".
      </p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_tp">
      <p>The number (percentage) of sequences labeled as 'positive' (Pos) and classified as 'positive' (TP), 
	</br />&nbsp;&nbsp;&nbsp;&nbsp; %TP = 100 * TP/Pos.<br />
	Labeling of sequences as 'positive' is described in the help bubble for "Pos".
        Classification is described above under "Classifying positives".
      </p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_fp">
      <p>The number (percentage) of sequences labeled as 'negative' (Neg) but classified as 'positive' (FP), 
	<br />&nbsp;&nbsp;&nbsp;&nbsp; %FP = 100 * FP/Neg.<br />
	Labeling of sequences as 'negative' is described in the help bubble for "Neg".
        Classification is described above under "Classifying positives".
      </p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_p_thresh">
      <p>The (optimal) score threshold for <b>labeling</b> sequences as 'positive' determined by AME.
      (See the note about "Labeling positives" in the section above.)
      </p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_tp_thresh">
      <p>The (optimal) score threshold for <b>classifying</b> sequences as 'positive' determined by AME (or provided by
	you using <code>--hit-odds-fraction</code> or <code>--fasta-threshold</code>).
        <br>
        By default, the true positive and false positive sequences are those with sequence motif scores &ge; the value shown.
        <br>
        If you specified <code>--poslist pwm</code>, then the true positive and false positive sequences are those with
	FASTA scores &le; the value shown.
      </p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_adj_pvalue">
      <script>print_doc_para('pop_adj_pvalue', 'ame-adj-pvalue', 
         '(The number of multiple tests is given above.)') 
      </script>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_evalue">
      <script>print_doc_para('pop_evalue', 'ame-evalue');</script> 
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_pearsons_rho">
      <p>The value of Pearson's correlation coefficient (&rho;).  
      It varies between -1 and +1.
      High values of &rho; indicate strong correlation between the 
      sequence motif scores and the sequence FASTA scores.
      Note that positive &rho; means that sequence motif scores go <b>up</b>
      when sequence FASTA scores go <b>down</b>.  
      The <i>p</i>-value of this method is the probability that &rho; is at least as
      large as the observed value assuming there is no relation between the scores.
      </p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_spearmans_rho">
      <p>The value of Spearman's rank correlation coefficient (&rho;).  
      It varies between -1 and +1.
      High values of &rho; indicate strong correlation between the 
      sequence motif score ranks and the sequence FASTA score ranks.
      Note that positive &rho; means that sequence motif scores go <b>up</b>
      when sequence FASTA scores go <b>down</b>.
      The <i>p</i>-value of this method is the probability that &rho; is at least as
      large as the observed value assuming there is no relation between the scores.
      </p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_mse">
      <p>The value of the mean squared error of the linear regression of X vs Y.</p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_m">
      <p>The slope of the regression line: Y = mX + b.</p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <div class="pop_content" id="pop_b">
      <p>The Y-intercept of the regression line: Y = mX + b.</p>
      <div style="float:right; bottom:0px;">[ 
        <a href="javascript:help_popup()">close</a> ]</div>
    </div>
    <!-- Page start -->
      <div class="prog_logo big">
        <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIAAAACACAYAAADDPmHLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAADiwAAA4sBdcsX+wAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAACAASURBVHic7J13WBRZ1v8/3TRNzklARDICgoiCCuKICQOCAcwTnLyTmJ2ZneDk3clh2ck7wRwxC4Yxi2LOERQERUAURXLqpn9/dHdNJ6BRJ7zv7/0+j49aVV11695T9557zvecI1IoFPwf/v+F+M9uwP/hz8X/CcD/55D82Q34vSDKyDAHAoBgIAhwAWwAa9Xf6j8AtRp/6lR/3wTygTzgkiI9vemPbP8fBdH/Bh1AlJFhCcQCQ4EIlIPek/s3w7UBxSiF4RSwC8hVpKc33Kf7/2n4HykAoowMU2AQygFPAGIAaVfvY29mRk1LC2131wctwCFgJ0qB2K9IT2+9mxv9mfgfJQCijIwBwCxgCuDU0bVeNjaM9/XFxcKCwupq1hYUUNf62/gkeHmxY9IkZm3ZwuK8PL3fD3J3J9bDA7FIRG5ZGfvKyjpr3i1gBbBIkZ5+sIuv9qfhLy8AoowMX2Cm6k+A7nk3S0vG+Pgw1seHx7Zt405zM4+EhvJWTAw/nTlDWX09Ec7O7CgpYWNRkfC7BaNG0SKXE+jgwJCVK7Xu+cOwYQxyd+e/Z84gAsb7+XGrsZFpmzfzeFgY7lZWbLt6lcPXryM33H+XgMXAYkV6+uX71xv3H39ZARBlZPQB5gAT0VjLRUBfV1fG+vgwzteXbpaWLM3PZ2NREfvLyvC0tub4jBmEL1pEWX29wXtbm5pS8Mgj+M+bx6mZMxmzbh35VVUAJPv5kTFkCOGLF1Pb0iL8JrFnT7YUF+Nja8toHx9G9ujBA927s+vaNbZdvcrWK1couHNH91FtwBrgA0V6+sn72D33DX85ARBlZAxEOfBjNY/729vzTEQE269e5dmICDYWFWEtlZLo7c0Dq1YJ170QGUlMt25M37y53Wc8FBLCMC8vHvz1V96OicFGKuWVvXsBWDhqFAV37vD+oUMdtjPK1ZWs5GQmZWczokcP3CwteWbXro5+shGlIBzopAv+UPxltoGijIx+wCcolToBw3v04IXISPzs7Pj65El2akzlLhYWvDdgAPZmZtxpbgbAzMSEZrm8w2c9HBLC4gsX8LOz4+D16yxOTGTO/v20yOX42tmxtrCw0/aO9PZm65UrHCgv50B5uda5qUFB+NvZ8a/DhzUPjwXGijIydgKvKtLTj3b6kD8Af7oAiDIy7IAPgKdRTfUSsZjZoaE816cPdS0t9LS1xeuXX5C1tWn99mZjI8du3CCxZ0+W5+cDcLm6mrTAwHaf521rS6SLCwqFgmnBwcjb2miSyUjx8yPz4kUaZTKsTU07bfdIb29+PHNG7/hEf3/eHTAALxsbXQFQIwE4JMrI+B6Yo0hPr+70Yb8j/lRLoCgjYwZKY8szmm2Rt7XhYWXFxKwsBq5Ywe3mZsb5+Bi8R9bly1rnthQX429nR5yHh9Z1g9zd8bCy4sFevVhw4QIJq1czfPVqRq1dy+u5uTweFgbAmoICZoeGYiIStdtuK1NTBnTrxrarV7WOj+7Zky/j40nesKGzWUiseud8VR/8afhTZgBRRkZP4Bc0pvtgR0ceCQnB3syMZ3ft4t2Dv+2k/nPiBC/27cs6A1NzdlERL0dFYSISIVcoqGttZcaWLawcO5afzp7lSk0NA9zdiXJ1Zdz69Xjb2vL9qVNa91hbUMAEf39spFJ+OXeOOA8P9qSmsvLSJVrb2kj09mb2tm1UNjYCMKR7d87euiX8H2Colxc/jxjBqDVruN7QQLNMZvDdJWIx8rY2VJqXG7BYlJExG3hUkZ5efDf9eS/4w5VAUUbGBGAuYK8+9kxEBI+GhvLVyZNcunOHM5WV1Gho4BYSCdcee4wRa9Zw/MYNvXsWz57NrF9/ZW9pqXCsm5UVk/z9sZRIOH/7NtuvXu1UN9BEpKsrI3r0QATsLS1lv8Y6nzFkCHWtrby5fz+gnF22TpzIkrw8XsrJwVoq5cCUKfjMnSv85qGQED4YNAiJWMytpiZ+OnOGr06e1DRC3QFmK9LT1xrdyPuAP0wARBkZUuBz4DnN4wPc3fkhIYHYzEzqW9s3pH0UG4untTUP/vqr3rmnw8M5XVlJbufGGkCpKNpIpdio1vra1lZqW1qMFpCXo6LYfe0aRysqiHJ1ZfOECbywezchTk5MDQzk+I0bhDs702vhQkA5+J8PHszE7Gz2lpbibGHBPwcOxMvGhuQNG3RtCV8DLyvS01sMPft+4w8RAJUxJxOI0j33nwceoKS2ls+PHevwHt2trcl/+GH8582jvJ39vRoSsZi+rq4EOzgQ5OBAoOpvT2trbKRSTMWGVZ/WtjZqW1oorasjv6qKi1VV5FdVkVdVxfEbN/SUUIBHQkOpbWlh1aVLAEhNTHg2IoIHe/Wiz5Il2EillD3+OBOzsvR0hm+HDuXkzZv8dPas7m2PAWl/hBHpd9cBRBkZsUAW4GCwASIRNtLOzfjX6ur4+exZejk6GhQAK1NTRnl7k+znxzgfHxzNzbvcVlOxGEdzcxzNzent7Kx17nZTE9lFRawvLOTXK1eE2WreuXNa17XI5RwsLxd2IqkBAZy6eVMYfBGo13/ePnAAMxMTQ02JAo6KMjKSFOnpuV1+kS7gd50BRBkZ41B++RYANlKplnUNYJyPD98mJOA/bx6tOl/YJ3FxvHPwIE3tKFSulpYk+fiQ7OfHiB49MJf8MTptk0zGtqtXWV9YSFZRETcatJ2C4c7OvBEdzdRNm3g5KgovGxte2L0bgJeioghxdOTVffu0lMh20IhyJsj+Pd4DfkcBEGVkPAT8jGqWGe/ry4oxY/jhzBneP3iQKpXhRiwScXDqVI5WVPC3nTuF388MDubpiAgGZ2ZqeetspFIeDwtjgr8/g9zdEXewXfsj0KZQsL+8nLUFBfx09qyegA/p3p1P4uIYsHw5oHzfD2NjmRYUJAi9vZkZc6KjGentza3GRn46e5ZlKrsGIAMeU6SnL/g92v+7CIAoI+Nl4FOUMx6gtN8vHDWKtYWFzA4N5bOjR/n21ClaVXv+7ZMmcae5md3XrhHl6orUxISpmzZRofq6TMVingoP562YGFwsLO57m+8HbjY28s9Dh/jh9Gmt2WxxYiJNcjnP79pFo0zGntRU1hYW8u/jx+lubc22iRM5XVnJx0eVxsE50dHUtbTwyNat6uVCAfxDkZ7++f1u830XANXgf2bo3MWHH2bMunW0trXxaVwcfVxceHXfPtYVFmIiEpHk64uHtTXHKio4dP268n5AWmAgH8TG4mdnd1/b+nuhsLqaObm5ZF68iALlO7zSrx8zg4O52diIk7k5UUuXIlco2DJhAiW1tTyxfTuaI7F8zBhyy8r4+qSWD+mV+y0E91UAVNP+PDS+fE38a9AgWuRy3j90iFm9evFqv37Ut7bSKJPx/O7dnK6s1Lo+wcuLT+Li6Ofmdt/a+EfiaEUFr+7bx86SEkA5i/W0taWmpYWKhgb6uLiwMSWFgHnzaNDRc9ytrFgzbhwDV6zQPKwAHrmfy8F9EwCVwrcW1Zof5uREaV2dsNarj60eN46sy5eJcHEhbeNG7jQ381BICLuvXaO4pgZQKlGfxMWR2LPnfWnbn40txcW8um+fnoBPDghgalAQk7OVOt4YHx/spFJh/Xe1tNRTMFHqBBPul2J4X3wBqq1eJiCRmpiwYNQofho+nMGenlrXnb11C5lCgbOFBaPXraOquRkFMP/8eWHwX+/fnxMzZvyvGXxQcglOzJjB6/37ax0/VlFBbycnQZE9VF7O0+HhvBEdDWBo8EH5gWWq+vyecc8CoDLyZKHa6v04bBhldXUMWrGCDZeVdowB7u7C9cvz87nT3KxnVDGXSFiSmMiHsbG/q2Zf29JCflUVu0pKWJKXx5K8PHaVlJBfVaWnwd9PqLX/JYmJwna1qKaGbVevkjFkCCKguqWFJrkc/851HQsgS9X394R7WgJU5t39qCx8Md268eWQIcStWCEoNC9ERvL54MF8cvQob+7fT6CDA79OmKBlJ/ewsmLd+PH0v49r/Y2GBrKKithVUsK1ujrK6uoor6/X4gUagrWpKe5WVnhYW9Pd2pqhXl4k+fjgaml539p2pKKClA0bKKuvx0Qk4rPBg+nfrRu3m5qwMjVl9Nq1WrsIDysrpgYFMSM4mK9PnmT++fPqU8eAQfdiNr5Xy8nnaJh3B7i7s7OkRBj8UJVtvNfChawcOxZXS0ue2L6dUWt/83f0d3Nj3fjxeFhZ3WNTlNr3uoIC1hUWsr+8/K7YvnWtrVy6c4dLKnrXkrw8xCIRg9zdSfHzI8Xf/553I/3d3DgyfTopGzZwpKKCv+fkEOzoyHAvLxbl5Qm2gUn+/swIDsbP3p4VFy/S2taGg7aFMwrlGDx/t2256xlA5dVbo3lsenAwj4aGMmz1auGY1MSEFrmcCBcXPo6NZfS6dVrX/zJ8+D1Z8Err6vjvmTOsLSjg7K1bRv1mkLs743x9USgUKFCq1up/b796VcuraAhhTk5M8Pfnyd698bS2vuu2N8lkPLp9O0s1WMnjfHyYHRrKSG9v6ltbmbZ5M7uvXSPSxYXlY8YQumgRLfpOq4l360W8Kx1A5c+fq3s8+/JlAh0ceK5PH+FYi1yOCHgrJobvT59W/h701sOuorq5mddzcwmYP59/HjrU7uC/N3Ag3wwdqnWsqrmZi1VVXLpzhwgXF2Ld3Smrr9dbIiwlEib6+zM1KIgoV1fh+Nlbt/jnoUMEzJ/P67m5VGvsdLoCTb1HU+vZXFxM959/5uD16/Rzc6NNoeDLIUN4IzfX0OADzFWNSZdxVzOAKCNjByoyh7lEQrNMJkz7ES4ubJ84kbWFhXxz8iRikYi3YmIoq6/nORVp8sPYWD2N2Fg0y+V8e+oUHxw+zO2mzqO1bj31lGChW2KA//9RbCwWEgnpe/ZoHY/18GBdUhI5paUUVlfTz9UVS1NT0jZu5Gptrda1jubmzImO5pmIiPacO53ioyNHeCNX2+9jZWrK7smTOXnzJmFOTro2AV3sVKSnD+vqc7s8A6goTAmgJE2cmD6dWb16CedP3bxJ6KJFAKwcO5Yfhg1jU1GRMPjTg4PvavDbFAoWXrhA0IIFvJSTY9Tg25uZIRGLSdu4kS/i4w3qGSKRCN1PwFYqZWNyMn/PyWFSdjb/2LuXhNWr+fToUa3ZbaK/P6ZiMbebmngpJ4egBQtYeOHCXeker/fvz/TgYK1j9a2tpG3cyOzQUF7KyRGOe1hZMczLi3Btj2XC3dDLujQDqAic+YCbu5UVuyZPZuaWLRytqABgSmAgRysqKKw2zHPs7+ZGTmpql6f98vp6JmZnc1CDlSM1MeGp3r05dP26YDbWRT83N34aPpzIJUt4KyaGAd26MXb9eq1rPomLw1Qs5u8aHZweGcmM4GD6L1umd08169jMxITvExKI9fDgrQMHWKky+4JSGV4zbhzuXVRsm2Qy4leu5IiqP0EpGFFubkzOzsba1JRvhg4lLTCQM5WVWJmacr2hgQlZWeotbAUQ1BWiaVdngA9Q8th4ITKSL48fFwY/wN6ez+PjOTp9Om9ER+uRLtRbva4O/tGKCvovW6Y1+ADO5uZ8EhfH2qQkJvn7G/ytv729IIwfHTmCp7U1s0NDta4xNANEd+vGpuJig/dUs4aa5XJmb9tG6saNfBwXx6Fp0xjSvTsAB8vL6b9smdA3xsJcItHbEY3z9eW1ffuwlEjISU3F0dyc4AULiFm+nLBFi9hQWMiKMWPUl7uhHCOjYbQAqHj7TwsN8/FhbUGBcP71/v15bNs2YpYvZ7yvL4+pWLbtvZgxWJafT/zKlZTW1emdK6+vpw0Yu349n8fH83KUHtkIfzs7ullaMm/kSM7OmoWVqSmfDh5MDxsb4RoRyh2AJuykUqqMWGJAuW10tbDg+9On+XzwYLKTk+lmZUVpXR3xK1dqunWNgu6HMmz1agru3OHZPn0oq68nJStLSwf56uRJrE1N6fubkvq0aqyMQldmgE80r2+Wy7U68okdO/j1yhUuVlVx8uZNrmg08pfhw7tk5FEAc/bvZ/rmzTSqnCQJXl6M9/XVuqaktpaG1lYGrVjBtKAgvk9I0KJz+9vbU9nYyPrCQoasXEnA/Pn8cPo0c0eOFLRuTYaOGjmlpXpm7PbwSVwcC86fZ965c0QvW8bCCxeEIJVGmYzpmzczZ/9+vWd0hP5ubvwyfDiAQIYZ5O7Of06cMKhfbL1yRdNhJkY5VkbBKAFQhWtpRewsvHCBH4YNw1nlm1ebdnva2jLY05OtV64AhpWbjlDf2sqErCw+1AiqmBIYyE/Dh1Ok8heoUVxTg7+9PeX19bx38CCP9+5NdnKyQDHzs7fnp7NnWVdYKPAK/nnoEDnXrgmUsW1Xr7JD5a1T4/vTp/GxteWL+Hi8bGwQodz72+pQ12I9PBjRo4dAYVcAmRcv6jGYPjx8mAlZWR2SXnWhqyxXNTe3S5379coVVl68qHkoQTVmncIoJVCUkZGNTqyeqVjMT8OHk+DlxUdHjrD96lV62Njw7yFDeHHPHnaUlBDu7MyJGTOMtu0rgAlZWazX4P+nR0byZO/ejFizhms6S8H3CQmYqFysNqamvH3gAFODgujn5sbYdetYMXYsD/36q6GgzU5hb2bGsxERTA0KwsHcnMI7dxi2erVgohUBh6ZNY9WlS3x69Kjeb0f37ImjuTkHy8s5pqKyJ/v5sTYpybCv3ADaFAoilyzhdGUlA93d+TI+ngdWrTKWvbxRkZ4+rrOLOhUAVZTuifbOpwYE8HxkJF7W1hRWV/PB4cOC/3tzSkqXvHpz9u8XvnwR8OngwYz18cHT2pphq1frKVX/6NePv4WH8+yuXWRrhH5PCwpiS3GxlivazMQEHzs73K2s6GZpibuVlaCll6uMQNcbGiivr6eourrTTp4RHMy/Bg0ieMECrWvH+fiwYNQoLty+zambNxns6cmRigqe3rmTFrmcN6Kj+WDQIKP7ZEtxsWA9fW/gQOI9PXl13z4Ot7Pz0UFkZ1HJxqjkczo6ufLSJVaqKNGaSPDy6tLgL8vP15r2548ahZ1UStTSpTwaGsrmlBQGr1xJ3u3bwjXFNTVcrq7WGnz1vQAczMwY6+NDir8/id7eWBkR8wfKZWjLlSusKyhgY1GRliCBcvb7MDaWV/ft0xr8eE9PMseO5fHt2wWjk0Qs5vuEBD6NiyN9zx4+PHyYMCcnpgUFGdWWxJ49SfDyYmdJCe8cOMAwLy9udU4mVWMOkNrRBR3OACp34yVAHN2tG3Oio6lvbaVBJlP+3dpKfWsr9TIZDa2tLM7Lo761FRFweNo0o5k8RysqiF+5UlD4AAZ7erK/rEwImng7JobHwsKIzcykRKVg9ndzY+XYsfScq22VTvL1JT0yknhPTyTtxAAYC1lbGzmlpWScOEHW5d9o+pGurpzQiVK69PDDfH3yJF9p07iQmphwZuZMhqxaxfX6eixUW7qu9E/0smUoUO4SHgkNxc7MDFupFDupFFupFFszM+Hfk7Kz1ctOGxDQUXxBZ70zU31NqJMTg9zdWV9YyKaiIo7fuIGvnR2JPXviaWWFv729oOSkBQYa/XLlqq1No47itLe0VCti5v1Dh1hTUMDWCRMExTOvqkrLuRTdrRt7UlPZMH48CV5e9zz4oPyCE7y82DB+PHtSU4nu1g1Ab/B9bG1xsbDgO5W/QxMtcjn/PHRI2Fo2ymSkZGV1GuCiRj83NyHO4EZjI7ZSKbebmujj4sKV2lo+P36cZ3ft4i1VnIHpb+ZoMcoxbBedzQAX0UjL8kpUFA+GhDBizRqu19ezJzWVv+/ZIyg5oJweLzz0kFEu0zaFgtjMTMHIE+fhwQ/DhvGPffvYpDOtg1IvmD9qFCGOjiSsXi0QOLxtbfk0Lq7DsPD7icyLF/nHvn1c0diVdLe25uDUqXT/+Wej7zPA3Z3ctDSjlOTC6mp6LVigxRMY5uXFC5GRjN+wgX5ubqwZN46ZW7aQo+3NvKRIT2+3Y9r9RFQJmbRy8nx27Bg/nz1LTmoq4319EYtEWoMP8FR4uNH+8sV5ecLgR7i4sDAxkRlbthgcfFDuEh7dto2D168L27ihXl4cnz79Dxt8UM5wx6dPZ6iXl3DsWl0d1xsatGwVneFgebnBBFWG4Gdnx1Ph4VrHdpSU4GFtzcMhIaxLSmKGxuBriFSAaiwNoqM5cpahg/85cYIvjx9nbVIS83XComykUt6KiensXQClIentA79lS3l3wABeysnh1M2bHf5O1tbGc7t2caWmhqfDw9k6YcJdhYHdKxzNzdk6YQJPawxKanY2n8fH82xEhJ7NoD28feCA0UGpb6nS2WhiRX4+Xw8dytRNmwQeg9TEhLyHHtKcWQyOJbQjAKo8fFPa+9EPp0/z+PbtvDdwoFYM3eNhYUYHbXx76pTWFDqmZ0+t/b8mJgcEaFkSJWIx3w4dyncJCfdlnb9bSMRivktI4NuhQ5GIxRTV1BCXmUm4iwubJ0zAwcxM63pTsZgQR0dSAwJ4d8AAhnl5caWmhm918hW0BxcLCyGRBcCIHj34e9++JK5dq5XGrkUup6imhkG/cTGnqMZU/x3aedYgOsnDN/fcOVrkctIjI3l02zYAJrTjlNFFdXMzH+ikT5ErFPSwsRHYwZoId3bGydxc8JL9Z8gQ/hYRYdSz/gio2/LMrl3caGjgie3bcTQ3Z3iPHoQ6ORHi5ESokxO+trYU19Rw5tYtghwc6OPiwo6SEj44fJhHVZp9Z5jg78+Xx48DSsPSpOxsrdwFamy/epVkPz+1YDihHNM9ute19/kMbec4TubmhDs7M7pnT8wlEuaokiS4WlpqSlyH+PjoUT1//vzz5/nvsGFY6HgLpSYmTAsKEkzLT4eH/6UGX42/RURoLQdyhYLZoaE4WVggAs5WVuLwww/0WriQ706dwsHMTMgqdrupSQgL6wyD3N0Fguqzu3ZpDX4fFxfeionhyLRpvDtggK4+YnBM25sBEkC55/wiPh5Pa2s8ra1xtrCgobUVS1NTckpLudnQIDBUk3x8jNJmS+vq+M8JfcPinNxcspKT2ZuWxpzcXHaWlOBuZcU3Q4ey8tIlimpqGOrlxVcPPNDpM/4sfPXAA+SpKOfVzc3CFtXa1JQzs2bhZ2dHo0zG0tGjmZydreXl/M+JEzwbEdEpx1AsEpHk48MvKv0rwcuLif7+JPn6Ym5iQnZREf88dIhtV6+yJzWVXo6OXFAazxKAd3Xvp7cNVCVergKk8Z6ebJ4wgZdzcsi8eJHbTU2M8/Ulxc9PmPbV2DB+PElGaMBvHzjAP9vJwWcukfD3vn15oU8fXC0tudnYSMaJE3x4+DDetrYcnz79T1H4uoLbTU30XbpUS78B5c7huT59cDQ3J+P4cUNJIXgrJob3B3buw8m6fJnxGzYA8PXQoTTKZKwvLOSADhN6TnQ0coWCj48cAWVuYwfdBNeGZoBYVImXc0pLGb56NQtHjcLT2pr3Dh5kRI8eepkurExNGdGjR6cNB7Q4BCN69KCyqUkwqjTJZHx4+DAfHT6MtVRKs1wukCA/jYv7yw8+KHcHn8bFMWXTJq3jqy5d4ov4eDYUFhocfFD2jTECMKJHD6xMTalvbRWodoaw8MIFTVaSFOXYan25hgRAa604UF5O1NKlfJuQwP4pU3CysNDLfzfK29sopk9hdbXA3n2ge3e+iI8XJFkTCtCK0onu1u2+7PPlbW0cunSJ7WfOcLWyktLbyvSwno4O9HB2Znjv3sQEBGByjzuLtMBAvjh+XMth80V8PJerq/XIp5o4e+sWhdXVndpRzCUSRnl7s0bjY9KEpUSCSCSipLZWMJurMBQjBCAClOuWrVRKWX09NS0tzNqyhenBwXz9wANMDwriPydOCCSHZD+/DhusxjqNBr8/cCBP79xpUOt/Z8AAFmjEC342eLBR928PCoWCH3fs4LP1G6jUYfRq4svsjTjb2PBK8nieGDYM0T2EqH02eLCQhNpcIlHGIqxfr5cFRRfrCgp4yQC7SRfJfn5aAmAqFvPugAFMDw6mp60tALllZTy3e7em2VpPezYk6sEAI7y9eVWHvbs0L4+opUuZHBDA+yqXpkQsbjeJoy7Uef5EKNk67WX1sjE1JVaV6DHJ15d4I9k5hnC54gZjP/6EVxcv6XDw1aisreXVxUsY+/EnXK7QT0lnLOI9PQWdqEkmI23TJmI9POjv5oa7lVW7CrOhXIiGMM7HR8sGsmrcOPp368Zj27Zh+913WH7zDf89c4b1SUl0+20Z0GPmaM0AqjIrPQF87ewMfp3FNTUMWbmS7io6WF9XV6PW5hsNDcKWxUwiwczEpN28vnZmZkKARnpkZKf3NgSFQsF/t23nvVWraLyLoM/9+fnEvvUW70yezJMjht/VbJAeGSl4EJvlcp6JiMDV0pJgBwdEIhHl9fWU1tVxra6Oa7W1LLhwgf3l5dxoaOg0FtHR3Jy+rq4cvn6dKFdXultb03/ZMi0lcNGFC3jZ2DAjOJgvlFnYeooyMsw1y9/ozgAB6mM+trb0sLFhgLu7pgQByj2uWssNdjCY/EsPWUVFQuOaZDKOVFQYNBvbSKWM9PZmf1kZDmZmd/31/3fbdl5buvSuBl+NxpYWXlu6lP9u235Xv4/39BSsgdfr6xmxZg0RixeTVVRESlYWg1as4IXdu8m+fJlZvXqRGhBAm0JBVju+EF2o+97f3p7NxcUG+YI7S0o0o5rE6Ph3dHUAYYrwsbXFytSU4T160NPWFhORiOKaGoqqqymuqeGN/fupbm4myEgB2KXDu3t8+3b2T5lCgpcXv5w9y+nKShzNzXl3wADmnTvHzcZGZgYH35Wp93LFDd7TSCF/r3hv1SpGRkTg6+ba+cUakIjFjPXx0XP47CwpId7Tk01FRQQ7OPD+wIHMP39eSFG/q6SER3Xo64ag7vu61tZ2wGnM+AAAIABJREFUYxDspFJdV3swIGS51hUAgabiY2fHoBUrBIars4UFPra2+NjZ0dPWViA+BhopALp8vpLaWkIXLuSp8HBmh4URaG9PcU0N886f5zuVbTzFSNOyJhQKBc/OnXtPX74uGltaeHbuXDa+9mqXl4IUf3+DArB89GisTE1J8vHh4a1b2X3tmnBet6/ag7rvd5aU8M3QoXhaW+tR6IMcHFh04YLWIc3/6AqACyiVNFupVBh8gMrGRiobG7WiVtQPMAZlBl6qpqWFT48e1SNVgpLDl+jtbdS9NfHjjh3s7yIX3xjsz8/nxx07eFJF1zYWid7eerqOo5kZYc7OnLt1iz5Llmj1MxjuK0NQ9706x9JYHx+9FPa67CRUY6yG7vxqA0pz4xM7dhjVCGPDo41lv6jhY2dnNIdPDXlbG5+t17cr3C98tn4D8k62cbqwMjXFR2Nf/2q/fmxITmbmli3M2LJFb/DB+L7S7Pusy5cN1i8AODh1Km6/KZU2mud0ZwBrUCp5mqQMW6lU0PTlCoWWccGYNK+1LS2CVm8hkfB6//58fOQIDTIZr/fvL4RAN8hkPL59Oy1yeZfj6gAOXbpk1FZPgEKBrPQaCpkM5G0o2tpA9UfRJkckkWDq64dIJYiVtbUcunSJQUYSOtVwt7ISyKyF1dX0Wby4w2m+TpW8urO+NabvQcm80oha1vpidQVAkI7148fz6dGj5JaV8XR4OG8PGEBDayvmEgkPrFzJsRs3lPwzI5Q0zeJNM4OD6W5tLaRFG+ThwflbtyiprWV0z55MCwpiwfnzdLuLlCzb2/kCDEHR0kJDzh5kBpJBiCQSpL16IQ0IFAZf8xldFQDNd1llgEFtCGX19QR1MsCmYrGwvEhNTDA3MUFqYoJULEaqGhupiQkS1d8qdDgD2IByW+FqYaFlqHkpJ4cfTp9mpLc3j/fuzbEdO4yWQM01Lb57d614ORORiBUXL3L8xg3yqqpIDQhgwfnzdzUDXNVJw9Ye2mprqd++jTZdO4dYjDQwEPPwCETtEFuMfYYm7uZdyurqjNKvbKRSmhsbmT9yJNOCgmiRy2nW+dPNygrpbx9q5wLgY2vLCQ1q1oXbt4UvNu/2bfztlbUebIxcozVnAEczM630Z0/v3CmEbTXJZMJSczedprbtd4bG3H16gy+ysMBq5ChM7O3b+VXXnqGJuxIAI/UAG1NTKhsbeWzbtnYrpW1MTm53BjA4f99sbNSidm24fJntKg9gNysrg4qLsSisrtaijF+pqRG2lDHu7mrf9e8G2bUSZAbCthWNjdRv2ULz6VMofsd0cX1cXO7JtN0e3h84kPXjxxsk5TSqlghD0BWAWlB+8f3c3AhxdNQ6KQJejIxkj2rPWmtksKNmWPiPZ87wVkyMnvt4rI8Pr0RFCXvWru4aQOnV6wwyA/QpNRTNTTSdOEHtqpU0HTuKwkAEjjHP0IXmu1iZmrJizBg2JifrZvjQgrGh9OoxeOfgQTYXF5M5diz70tJI8vUVmMG/FhdrRhNpacm6S0AtKO3Wz+3aJWS1LqquxsHcnFHe3lyvrxcMNcYmVvTQ2K6cvXWLZ3ftYuGoUZhLJFyvr8fV0pIGmYzHtm0T0rPdjQD06KBD1WjT0b7FdnYo6uuVOwEVFK2tNJ89S/OFC0j9AzALC0OsegdjnqELzXfJLSvDd9485o0cybHp01l+8SJv79+vF/nsYeT2Wj0G9a2tSh2tRw9yy8t5rk8fPo6L47OjR5l//rymF7JzAQBlNa7oZct4KCSEECcnbjU28sHhw1qpUJrlclrb2jrdCehK89qCAjYXFxPi6Eh3a2tK6uo4f+uWlrHkuuE0qR1ieO/efJm9scNrTJydaVUTWsRirIYNQ2QqpfnsWVry87QEAbmclvw8Wi5exNTXB/Ow3gzv3bvL7dJ9l0aZjJqWFh7fvh13KysOTJ3KiosX+dehQ9xUfanGzACtbW16zjRHc3NWX7rEF8eO0dvZmbdiYvjXoEH4zZunvrZDAdD6PIpqarTKt6kf0CSTCUphbUtLp95AG6kUa1NTrRRsTTIZx2/cMFgFDO5uBogJCMDZxqZDW4CJRro3s6BgxDZK37l5v36YhYbSfPYMzfn5oNmxijZaCwtpLSzkUxS88cYbRKvy+RoDQ+/ib2fH96dPM//8eb4/fZpX+/Uj/+GHSVq/nlM3bxptX9GFuuRNkq8vY318iHJ1ZW9pqWZ6Oa0xNqgDgDKyNsxJnxn+RO/e/DxihPB/Q+lbDKEjTdjR3JxZvXqxatw4HlE5QYqqq7uUUAHARCzmleTxHV/j5AwiMSKpFDMddrHIwgLz/tHYTJyENLgXGFCc1q9fT0xMDMOHD2enRoWT9lDf2kqRgaRZfvb2+Nja8mZ0NEsSE3ksLIzSujpqW1qM3jVo9v3wHj34fPBg/O3tWZuURLCDAxknTtD955+ZmJ2tmaFE6+vQFQBh7/d6dDQxqkBITXx85AhhTk50V61R6qrbnaG9NS3A3p7rTzzBY2FhHCov54U+fYju1o1muZwtKip4V/DEsGEdGmpEEgkmjg6YhUcgaoeHL7a0xCImBpsJE5EGBoGBJW7Hjh0MGzaMmJgY1q9fr5dnSI0tV64I07SjuTk/Dh/OqZkzcbaw4Pk+fTCXSPjm1CkC5s+n96JFnK6sNHr91+z7dwYM4E5zMwNXrMD7l194ZtcuthQXG+JbaIVe6S4BgoUmyMHBYGJFUHqfwl1cuFZXx0UjBaC7xkuN7tlTGdWSk8OlO3e4VlfHpOxsKhsbOX7jBm/0709KVhbrCgrazQDWHkQiEd/Mnk3sW2+16xE09fXDzIi0NWIrKywGDsQuKoqhIliTmYlMJ4r58OHDpKSkEBoaymuvvcbUqVORaPAjNWlw9mZm5N2+zc9nz3Lixo126WHdjRQAzb4fnJlp1G/QGGPQnwGEEb/V2IhrO9YwD2trwSli7AygGUjZx8VFK9PVogsX6KXacu6+dk2YeTYWFRms1dcZfN1ceWfy5HbPm4WEGPyq28P7Dz/MiiVLyM/P56GHHsLEwNJw7tw5Zs2aRWBgIN9//z1NTU3I2tqESuegLGz9pYos2hE3ULOvOoJm3wc6OPBRbCzLx4zhu4QERnl7t0c70/qqdXvhEsqkAmy4fJmXoqL0NPwJ/v4MdHdnj8qGnmekAGgGjlQ0NGitc+8cOCAENopEIswlEkxEIqqam3VDnY3GkyOG8/H06VgYaa42BAuplI+nT+fJEUoXsK+vL/Pnz+fChQvMmDEDsQEhKioq4m9/+xs+Pj488+abVGnGP/r4sHzMGLZPmsQncXFEuuoTTNSBH8ZA3feD3N3ZPnEiYpGInNJS5ZZw2DC2TZyoWwm9DeUYCzAUGFII+IqAhYmJ9HN1ZXNxMSZiMbEeHnhZWzMhK0vg90nEYiqeeMIoXuDgzEz2lZXhY2vLlgkTCFm4ULdsKs9ERJAWGCgwapN8fdkwvmPFriNcrrjBs3PndpkjMCgoiG9mz+6QBXThwgXee+89MjMz29UBsLCAwYN57vnneXrgQDJOnKBJJmOEtzeT/f355tQpXtm7V7g8zsODvWlpnbbvdlMTbj/+iKytjQNTpvBiTo5WMk2JWMzcESO40dDAy7/d/7IiPV2Lwm1oHswDJTf/wS1beGbXLprkchpaW/ns6FH85s3TikeTtbXp5ehpDykq+nhRTQ17y8pYMno0gQ4O2Km4f4sSE3lv4ECt+oFZly/f9SwAyuVg42uv8snMGTjb2HR6vbONDZ/MnMHG117tlALWq1cvli9fzunTp5k0aZJhtlBjI2zdys+pqUSmpvLj3r0svHCBWVu2ELl0KRP9/XlGYzeSYiTFPlu1PLpZWmJmYqKXSVXW1sZLOTlMCtCiAOopdYbiAk4BY0ApBDtLSoSsX2qEOjkxvEcPIcZvfWEhD2okjG4PKf7+gjQ+tWMHH8XGsi8tDRcLC240NLCmoIAoVVhVDxsbvGxsyC0r45W9ezk0dWqn928PIpGIJ4cP57GEhN8lMCQsLIxVq1Zx6tQp3nnnHdbr5CMGaGxshJ07Yc8e6N8fEhLIAx7eupUv4uOFEHFjaXDqUHobHeaWJqqam3XTy+vFoRsSgF3A6x09/GJVFXOio3myd2/+e+YMv165QpNM1ml0kJ+dHWFOTsriUW1tvLJ3L6/s3SuEOalhIhKxZPRotl25Qm5ZGYevXyfz4sV7jg4yEYsZFBTUZX++sYiIiGDdunV8vHYtr7/5JvxW2uU3yOVw8CAcOgQRERwfPRq3xERAmYzSmOwqTTIZv6q2yJerq3G1tMRSItErPTemZ0/d0HG9ODJDop6LMpAQUBYw1OWot6qydLzYty+gNHboxgu2B0M5BHQNPm/GxCiLLCmDGgH4x759RqWI/7Nxu6mJH6qq4PHHIT0d2ttuKhRw8iT1H31E8rhxcPmy0fkVtl29KvRZm0JB2KJFekm2fGxt+TguTrMqewvKsdWCngCookeF8N2B7u4Gy5fdamqirrVVyNjVXnYPXTzZu7deDgBNxHp48HJUFM0ymRb/7kpNDZM3bryrbeEfBVlbG5M3bvwtMtjbG558Ep5/HjqYvU7m5MDXX/PrSy+xuR2fviZ0+9ra1BQnnS17WX09Y9et49xvlVQO6UYGQ/sJIgQtrFEm091KAMpp2sPKStjPagZ+dARPa2teaCfax87MjMWJiUzdtImalhae1SjOAEq+/POqKtx/RTy/e7cQ/2ApkTAtKIjPBg/mmZQUHF98EZ59FjpQ8g4fOMCYMWOIjIwkMzOTNgPCrhk40tPWlt2TJ1P51FOUPvYY52fMYLa/Pyb19TRXVlJUWAhlZXDlChw4cFkkEo0WiUSJIpFIGHeDaeJEGRlDgN2gTN5kIhbzlioTCCgl7gdVNo9J2b8VsNyblkacKqavI1Q3N+M7b57elL58zBhK6+p4KScHV0tLDk+dSnJWFk7m5nhaWwtcgW+HDv3LZQn57tQpIeOHp7U1G8aPJ7+qivyqKoZ4ehLu4sKT27ezuqAALl2CzZuhk91TQEAA0dHRNDY20tTURGNjIzeqqzlXUQGtrXSTSqmuq1MqmK2tSkJrxzgBTFcoFMJuoL25eD9wC3D68vhxtk6cSKSLC8dv3MDLxoYkX18OlJczTScGfm1BgVECYKcql65ZBuXhkBD87Ox4fvduxvj4MMjdHblCwfHp0zl/+zb7y8pYU1BAfWsrL6hCrP8qQvDdqVNCmwCWJCby7+PHtQJCpgUFsTAxkcaNG9kEEBAA+flKQWjH53Hp0iUudUAiNSpbMIBIBCLRF7S1vaFQKLTs4+0mihRlZHwL/A1+y9MT5epKaV0d269e1csPCMotSeEjjxiVKaxZLidowQJhvdyXlkaQgwMmYjGHr1/nQHk5+8vKOHT9OjUtLZiZmPDewIEcq6hgdUEBbQoFT4eH89UDD/xpmcJkbW08v3u3UA0NlMkfPxw0iASN0nlqPBYWxvN9+hC+eLFwzNvWlgwXF/713nsc+01hu3+wt4cxY9YrlixJMXS6IwEYABwweFIHfnZ2VDU3c7upief69DE6j8/CCxd46NdfAeV6ZmVqyvlbtwwWV3AwM2OwpyczVPHvo9et43ZTE0O9vFg1duwfnj3kdlMTkzdu1It5jPPwIM7TU52WRQtSExOqn34aj59+EhJQLxg1SrChrF+/nnfffZeT+tE8d4c+fSA1FSwtByrS0w8ausSoVLHmEgkbxo9n9Nq1WqbbPi4uvNa/P6O8vZm2eTNbiovvKVWssXiuTx/G+/oyYo2ybuVfIVWsJgxVIQFlUMzNJ5/E9rvvaFMoDKaKVSgUrF69mpkzZ9J8l+RbK2trpKmpVCmzlt1dqlgVFoPS8HC9vp4pKgPKkO7d2ZySwtaJEzlbWYnv3LlsURVZam1rY06u3nbT8MNFItaMG9fl6ptfnzxJgL29QE+/UlPDlE2biFm+/J7Mxp0hp7SUmOXLmbJpE1dqajCXSHiyd292TJrE3rQ0Xo6KwsrUtN3yMO8MGMAPp0/TplDgaW3NmnHj9Dx2IpGIyZMn8+KLL95VGwP69OGJBQvUgw+qMWwPRqeL7+XoyNqkJKqamnC3suLzY8f45dw5PQMEcF/SxYMyM+bNdnLjn3/wQZ7euZM9164R4eKilWL2j0gX383KiuzkZE7cuMGC8+exkEh4f+BA2oCRa9ZoGbciXFx4OjycYAcHkjZsUN6zk3TxdXV1BAYGUt6F2VFkYkLU7Nkc7dkTlMa7TtPFG1MxZCUwGWD1uHFcra3llb17DRpkbKVSalQkjAQvL3ZMmmR045fl52sFNlhIJByYMoU+S5boXRvl6sq+tDTcf/qJO83NlDz2GCPXrNGLKfi9CkaIgH1TpvDv48e1Qr0sJRJ2TJrEmVu3eGL7b0klHgoJwdzEhHnnz9Mil7N09GijCkYsXLiQhx56yKg2a8HUVLn+9+mzU/Hjjx1WE+1SyZi+rq589cADxBlgn3hYWbEhOZlfi4uF7KH3UjIGYNno0RyuqODfqtSo6ufsnDyZZfn5vKcirDY++6yWYmUI97NkTIC9PcemTyd00SJKamu11vxwZ2cOT5uG37x5BvmSXSkZo1Ao6NW3L/n3phSeAj5WKBTLDZ3sdH5U1ZzZCHD8xg1qW1oMRraU1dcTs2wZ/vb2gqXv1X37ulRG9V+DBmllHHtyxw7SAgL4ecQIxvr48ExEBAenTuVIRYUgKJYSCeYSCS9FRWmGQPNIaCjPa1gSm+Vy8m7fZldJCcvy8/ny+HHBGfXl8eMsy89nV0kJebdv6w3+osREfDWU2kt37vDLuXPMHTGCh0JC2DpxonDudGUlBXfuEGAgxCzZz49/daFekAKQp6Qo9/F3jwig3VAkYxdIoRrl49u3c6SiwqA9X65Q8NSOHTzXpw9SExNOV1bypoYFsTOIUBpR1EJQ09JCXGYmJ2/c4OGQEOI9PfnHvn3M2rJFMEE7W1hQ2dhIo0zGoalT+XH4cP7Rrx8fxcYKCqGzhQW3nnpKr80SsZh+bm7M6tWLJF/fdrn4FhIJU3R2GK/n5tLd2prUgABmqbay6newkEj0rJzJfn4sSUw0umIYwJv791NgZwf9Oq4DmTxxIo46UVwaqAZ+bO+kUQKgSE8/gMo/cK2ujha5vN31vaq5mUaZTPgaPzpyhKVGFkUAZejU2qQk3lDx7uUKBd+cOkXqxo1M2bSJ5TrMHmcLC8rq6/ng8GGCFy7Ew8qKFyIjGbpqFSdVimE/NzduNzVpKZlRrq4cmTaN1/r3J8TRkSRfX45On27QT7EiP5+pOmu2us7vuPXrua7B+3+uTx+u1NRwWiOK+I3oaNYmJXUp4cXSvDw+UtsSkpIw6cC41m3cONLXroWxY0FfiL9XKBTtBkp0pZDvqyi9hGK5QoGDmRn2ZmZ6ZISetrZ4WltrlV59dPt2AhwcjK4eKgI+GDSIMCcnHt22zeBOQw0nc3MhuvjR0FB6OToyODOTyxpc/ChXV72SczHu7lysqtJK6eppbc3pmTM5WlGhFRqfXVTE3JEjCXZ01Kpapt6hBDo44GVtzWNhYXS3sSFJRQixkEj4ZcQIoyuEqXGkooJHNZRIbGyQJibSuHatwetnxcQwds8eGD4cBg9u46uvfqCsbCpgBfyno2cZvUdSpKcfBb5X/39RXp5emjcvGxuykpP5+exZ6lpbebFvX3xUCaVSNmwwOuRZjWlBQeSkpnZoJ3BWsYn+3rcvz0ZEEL9ypdbgg3IG0DVd1zQ36+XnL62rU9K1dNhNjTIZGwoLmdqOoWlaUBBPhoeTVVTEkJUrudPcjKe1NTmpqV0e/LL6elI2bNCrPrroww+xbyeqePG1a1SrP0Qzs+8VpaXPoMz3OFGhUHToMujqJnkOyhLlfHr0KIEODmSOHctT4eF8PXQo52bN4tfiYl7bt49uVlbMiY5GplIC23uxztDPzY0j06YxoJ1aBM4WFgz18mJWr17tFpo2NAPUqJTZK48+ysdxccLxUzdvEmEgAHTFxYt6y4Aa7x08SNrGjSzNyxMsfEe6YAdRQ/NDsZVKSY+MpJejI+mRkbSIRNwZPVr/R5aW/PAb86gCVZ1HhUJRq1AoNun/QBtdEgBVXfqXQGkgmZCVxdqCAvq6ulLf2kr/Zct4ee9e2hQK3o6JYfGFC1r5hPSmNiPhbmVFbloaC0aNwluVB1eNrMuXWXj+PENXrTJoNHK1tKS7jY1eDOLWq1fx/uUXAufP57V9+4TjErGYegNCuuXKFdwsLZno78+c6GhOzJihJ5TetrYsGDWK3LS0u0oK8ahKwQblchTk4MDGlBT+PWQIbQoFSUlJmOhyL7Vnx5dUY2Q0jKodrPejjIwd6BST1oSfnR370tKoaWlhzLp1FOpMyR/GxmoVRu4KmuVyvj11ig8OH26XIhbi6MhoHx+OVVTgZmnJ+wMHErRggVH3Xzp6NCW1tbyqIRTBjo5M9vcnvW9fSmprybx4kZWXLgk1iR3NzZkTHc0zERGayZi6hI+OHOGNdkzo/vb2jPL2ZqS3N841NcT3749cLaT+/vDMMwA7FenpHRp9DOFuBaAnSuOQwXwqS0eP5vD165y8eZNvExJ4QLUuxnl6squkRKnk3YMQgJJU8vHRo/znxAk9JdHXzo6ZwcH0dXUlys0NG1NTtly5wrGKCrKLigSL4Uexscw9d46imhqczM15NCyMx8PCiFq6VBCuMT4+fDhoECsvXSLz4kUhfwEolbwXIiN5rV8/o+r9tIePjhxhTm4upiYmvBMTw3g/P8QiERuLinj3wAEtsqepWIzr1q2UblSFwUdGwoMP3kFZJ7i4q8++KwEAEGVkTADW6B7v4+LC6nHj6LVwIS1yOXOio3mwVy/szMy4XF3NjM2bhWQI04OD+WX4cKNqDbSH0ro6/nvmDGsLCoRaBLpwtbQkytWVvq6uHKmo0Ko/9GCvXnS3saG6uZk9167x9oED3OqEfBrm5MQEf3+e7N27y44sTTTJZDy6fTtL8/KQmpiwdcIEckpL+fHMGdytrPggNhZPa2viMzO129TUBB98AHV1EB8PEyZMVKSnG94idIK7FgAAUUbGV8Bzmsc2paSwvrCQJrmcR0ND8ba1ZfGFCyy4cEEIZrSUSJApFLTI5fR3c2Pd+PFGp0TpCIXV1awrKGBdYSH7dcqn3AvEIhGD3N1J8fMjxd/f6MKYHUGtFKvX/DnR0dhKpVpLj1gkYtGoUTiYmzNGo0QuAAcOQGYmxMQcUBw8aLx5UQf3KgBSlPSxKFA6X64+9phAE59/7hy7rl3TGggTkYj148ezv7xcMOd6WFmxbvx4o+0ExuBGQwNZRUXsKinhWl0dZXV1lNfXayWpMARrU1PcrazwsLamu7U1Q728SPLx6TR9e1dwpKJCb1u8KSWF2du2aRmVQLnMlD3+OOGLF2tX/1Ao4JNPGrh9O13R0vLT3bblngQABJfxUcABoJejI9dUiQ4M4ecRIwi0t2fEmjU0y+WCI8VcIuGX4cOZbkTY9r2gtqWFsvp6yurqhAHwUA24h5WV0bkP7xZL8/J4dPt2ve2wo7l5u0rtntRU3sjN1S2wUcXPPz/FuXO1CoWicy55O7hnAQAQZWSoixF1SAZ8d8AA0gIDiV2xgqrmZqQmJuSkpjJoxQphlni9f3/+NWiQUSXo/iehTaHgzf37BfNukIMDE/z9sZBIWHnxYrv6i52ZGWdnzaL3okWaVtdGYIQiPT1XJBJJdYmeXcF9YVMq0tNzgTRAEGs3S0tWjBkjfFGPh4XxeO/eJK5dK7hto93ckIrFWkvER0eOELlkicAw+t+ALcXFRC5ZIgz+Z4MHs3T0aGRtbZiZmJCTlsazBhjOTubmLBs9mo+PHNEcfBmQpupz7mXw4T7NAMLNMjIeAuahKl79clQUU4OC+O7UKT6PjydBw0EDSifJgG7dSMnKMqiwJXh58UlcXJctan8VHK2o4NV9+7SCa98bOBBbqZS/79kjcAh6Ozuza/JkZm7ZIgj+WzExPBwSwidHj2pmAVcAjyjS040zahiB+yoAAKKMjJeBz9T/nxIYyIJRo/h7To6QX1CNLRMmYCmR0KZQ8NDWrVypqWFI9+70dnLiG9W1IpRl2D6Ijb0v2vcfgcLqaubk5pKpkVIPlEry5dmz8Zk7V8+JNlvFX1AzoDR3Shp4RZGe/vn9bOt9J9SrGvgKKpLMiosXGblmDS9HRWkRSUxEIvq7uZG4di1fHj/OtokTmTdyJMtHj6ZUQxNWqO7Ra8ECnt+9u12O4F8BNxsbeX73bnotWMAK1eBbatg4qpqb2VRUxDdDh9LX1ZVfNLKtLbxwgRAnJ8Em0iCTaQ6+gt9h8OF3EAAQhOARVDpBTmkpY9et499Dhgg6QV9XV/KqqmiQyTh/+zbVzc3Ym5kRsWQJawsK+Cg2ljOzZpGdnMyUwEBa29r4+uRJ/ObN46WcHPaVld23ff69oE2hYF9ZGS/l5OA3bx5fnzwpkFX6u7lxbPp0rTQ7T+3cSXS3bixJTBRmOVBS1qqbmw0RRmQop/37PvjwOywBWjfPyBgHZKLaHYhFImHQXoqKwtHMTJmMcsAA3sjNZaEq9q+HjQ1nZs0iaMEC7KRS3o6JwdbMjEnZ2VpToqulJUk+PiT7+TGiR497sih2BU0yGduuXmV9YSFZRUVa0dOO5uY82bu3oPDtnzKF5fn5WqVbAuztud7QoLVV/iI+HkuJhKe1cw82olT4svmd8LsKAAhbxCxUdgI11o8fT7izMwV37jB72zYtI8d/hw1DamLCI1u3Csd+HjGCS1VVfNJOmXUrU1NGeXuT7OfHOB+f+x4pdLupieyiItYXFvLrlSvtJrHs7ezM2qQk/OfNA5Th9VnJyfjPm6e37nvZ2BDk4MALkZG0yuVM2bRJM3tYFZCk1vZIcSEGAAAGcklEQVR/L/zuAgCCsSgTlcUQlB0T7uzMj2fO6AVSpPj58WZMDIfKy4WI21He3rwcFSVEA6kx2NOTA+XlWjR1iVhMX1dXgh0cCHJwIFD1t6e1NTZSabu5jVvb2qhtaaG0ro78qiouqqJ786qqOH7jhh4VXl3qRtPVHO7szPrx4/GZO1c4tnLsWK7U1GgmawKUDKYHvLxYmpfHZu1t7zGUX367fP77hT9EAEAwG3+Oju+gIwTY2wvet4dDQpgRHKwlAG6WlhTNno3bjz9S29LC7NBQNly+TGUniqKZiQk2UqlQ8KJWVaOnIyp4rIcHkS4uSE1MyLp8mUt37pDYsydzR4xg6KpVeKooYero6B6//CL81s/OjpMzZxKxeDGXq6vpbm2Nk4WFVjCLBr4GXlakp/9+RQs08IeF1SrS01sU6enPAxOBO4auEYEWO0fT9fpwSIjuV0KKnx87S0qEtXSivz8PdO+udY2VqSkPhYRgKhYLClazXE5lYyNFNTUU1dRQ2dhIs1yOpURCbx02UC9HR45Mm8ZjYWGIRCKsTU3Zm5bGY2FhbCku5u85OZyeOZM3oqNZrKqtrJtkqrC6mp/PnuXL+Hj+OWgQu1NT8dEhtqj6ZKIiPf35P2rw4Q8UADVUbstINLKQqCESiQi0t+eL+HitkK5nIiLoYWvLz2fPal0/0d+ftRqpWE/evMnQ7t1x0lj/X+3XjyRfX6W+8cgjeu15ondvIddRoIMD2cnJWuc/jI1l/vnzPLJ1K1+fPMn7hw4xYs0avnrgAbxsbFien0/6nj04m5uTW1ZGo0yGiY4Z20RVJ3h4jx7I2tqIWLxYt0j0TpT+/Lty6d4L/hi1WQcq4sIwUUbGDOALwA2UW6qpmzfz7oABnJs1i9yyMnra2uJlY8Ok7Gwh7AyUeXfjPT2ZsWWLcOzEjRv8MGwYQ7p3x9rUlAu3bxPl5kbU0qWCkqnOUgbKXckb/fuTqiJXVDQ0aAWXABwsLxeypkvEYmRtbZyprCRLVe/3w8OH+f70aZwtLFiflMT4DRu0BKCfmxvzR44kt6wM//nzdb19FShpXPrxb38Q/hQBUEORnr5ElJGRjTLw5GlA3CKX80ZuLh8ePkycpyeNMhmHr1/XY/2M8/Hh4PXrWuv9iZs3udHYSNiiRUhNTNiUksLnx44Jg5958SL70tKo0thvl9fXCz75m42NSE1MtOjuuWVlvNKvHw+oqp2p6/serqjQIo/+89AhnMzNWTBqlJYj61pdHWkbN3JeO26xDSXDek5XOXz3G3+qAIBANH1WlJExH/gEFdewrrVVzyH0cEgIDTIZW4qL9aZ/UNYYUGfOHOjujruVlVZc4YqLFxnj40OEKkPHxuRkrTQusrY2KhsbcbO0FATgSEUFk7OzOXz9uhY1S6FQ6HksX9yzhwWjRmlVXLteX6/71e8EXlXR7P90/OkCoIaqQ4aJMjIGoqQ2j9W95nJ1NdOCgvg0Lg5Pa2uOVFRohZArgLOVlfR2dubroUN5dtcurazcJ2/exEIiIdDBgSaZjHAXF1ZlZWk9Q70MqDNxN8vlWoWdQbmLmBoUxCc6WUAUwIMaYWI62Ah8oIqy+svgLyMAaqg6aJwqKnkOyl2DGJQm5ZzSUkQoI3tS/PzYk5pK1uXLApXqxM2bfDP0/7V3/i4JhGEc/xgmdVeb0g9EJ8nByKmwQHIocAyJlpqaA/+ECNqCbm8skGbnpKVWKSGECMIkEEsH9YQGa3jP0O4Ms7LT/MDDu9x2X7j3Ht7384S4zud1+hYQn4F1j4dRq5WjVEqnbTfaB/gdDpadTu60X7htn4+zTEYYvz6nhjg3ua9dsjUdXesDdIrWRNrUymP0TKNqdtXtZi8QYC0eN5zVM2u3EwuHGbfZmI/FyKkq1qEhJiWJCUniIBjksVzmplBgWpY5TCZ5qlbZ8fuZkmUeSqWmI+EtuEWYOU660cz5DqYPQCOauGoL2AD0A43aJBGJkC4W363kKy4Xu4EAuUqFnKo21Xk2266i9hk4BY5bCZnMSE8FoI5FUYaBRSCE2DQuAL97mE/PC+KybAIhYb58jUa/NuXKBPRkAD5iURQJWEIEYg7wIi5H/lSjqwbcI3z7V4gXfmHk3u01+iIARlgUZQSxZ/ACM4ADMTh5TFvrBWKUWr3K2ppHDFhKI1Rr5leVd0DfBmBAe/yNY3WAaRgE4J/zBubyScGhopWBAAAAAElFTkSuQmCC" alt="AME Logo">
        <h1>AME</h1>
        <h2>Analysis of Motif Enrichment</h2>
      </div>
      <p>
        For further information on how to interpret these results please access
        <a href="http://meme-suite.org//doc/ame-output-format.html">http://meme-suite.org//doc/ame-output-format.html</a>.<br>
	To get a copy of the MEME software please access 
        <a href="http://meme-suite.org">http://meme-suite.org</a>. 
      </p>
      <p>If you use AME in your research, please cite the following paper:<br />
        <span class="citation">
          Robert McLeay and Timothy L. Bailey, 
          &quot;Motif Enrichment Analysis: A unified framework and method evaluation&quot;, 
          <i>BMC Bioinformatics</i>, <b>11</b>:165, 2010, doi:10.1186/1471-2105-11-165.
          <a href="http://www.biomedcentral.com/1471-2105/11/165">[full text]</a>
        </span>
      </p>
    <!-- navigation -->
    <div class="pad2">
      <a class="jump" href="#data_sec">Enriched Motifs</a>
      &nbsp;&nbsp;|&nbsp;&nbsp;
      <a class="jump" href="#input_sec">Input Files</a>
      &nbsp;&nbsp;|&nbsp;&nbsp;
      <a class="jump" href="#info_sec">Program information</a>
      &nbsp;&nbsp;|&nbsp;&nbsp;
      <a class="jump" href="ame.tsv">Results in TSV Format</a>&nbsp;<span id="results_help"></span>&nbsp;<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABwAAAAQCAMAAAAyEe/dAAAACVBMVEX//wAAAAD////s2cV/AAAAdUlEQVQYlXVRBxLAIAhL+P+jC2HZhXcBZEWEldDsZcLIcAhHWWnK8SDcWQhMFUHdAQ1CqQ5+CWPmlHojl+nCJNRtzu4qRc3IUzmTVpXYK0nox0z0PI1stgchdK7lEv7ekhvalw8WW547Gyzedt/2/gLx8WXjXF/1AYFriNAWAAAAAElFTkSuQmCC" alt="NEW" id="new_1">
      <script>
        make_help_button($("results_help"), "pop_results");
      </script>
      <script>
        if (data['options']['noseq'] == false) {
          document.write("&nbsp;&nbsp;|&nbsp;&nbsp;\n");
          document.write('<a class="jump" href="sequences.tsv">Positive sequences for each motif</a>&nbsp;<span id="sequences_help"></span>&nbsp;<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABwAAAAQCAMAAAAyEe/dAAAACVBMVEX//wAAAAD////s2cV/AAAAdUlEQVQYlXVRBxLAIAhL+P+jC2HZhXcBZEWEldDsZcLIcAhHWWnK8SDcWQhMFUHdAQ1CqQ5+CWPmlHojl+nCJNRtzu4qRc3IUzmTVpXYK0nox0z0PI1stgchdK7lEv7ekhvalw8WW547Gyzedt/2/gLx8WXjXF/1AYFriNAWAAAAAElFTkSuQmCC" alt="NEW" id="new_2">\n');
          make_help_button($("sequences_help"), "pop_sequences");
        }
      </script>
    </div>
    <!-- alert the user when their browser is not up to the task -->
    <noscript><h1 style="color:red">Javascript is required to view these results!</h1></noscript>
    <script>
      if (!window.HTMLCanvasElement)
        document.write("<h1 style=\"color:red\">Your browser does not support canvas!</h1>\n");
    </script>
    <!-- write out the job description -->
    <script>
      if (data['job_description']) {
        document.write('<h2 id="description">Description</h2>\n');
        document.write('<div class="box">\n');
        var paragraphs = data['job_description'].split("\n\n");
        for (var i = 0; i < paragraphs.length; i++) {
          document.write('<p>');
          document.write(paragraphs[i].split("\n").join('<br>'));
          document.write('</p>\n');
        }
        document.write('</div>\n');
      }
    </script>
    <h2 class="mainh pad2" id="data_sec">Enriched Motifs</h2>
    <div class="box">
      <!-- messages -->
      <p>
      <script>
        // Motif scoring stuff.
        document.write('Sequence motif score: ' + data['options']['scoring'] + '<br>');
        if (data['background']['source'] == '--sequences--') {
	  document.write('Background model source: letter frequencies in (primary) sequences file (' + data['sequence_db']['source'] + ')<br>');
        } else if (data['background']['source'] == '--motif--') {
	  document.write('Background model source: frequencies given in the (first) motif input file (' + data['motif_dbs']['0']['source'] + ')<br>');
        } else if (data['background']['source'] == '--uniform--') {
	  document.write('Background model source: uniform frequencies<br>');
        } else {
	  document.write("Background model source: background file '" + data['background']['source'] + "'<br>");
        }
	document.write('Background model frequencies: ' + data['background']['frequencies'] + '<br>');
	document.write('Total pseudocount added to a motif column: ' + data['options']['pseudocount'] + '<br>');
        if (!data['control_db'] && !data['options']['fix_partition']) {
          document.write('FASTA scores provided: ' +  (data['sequence_db']['fasta_scores'] ? 'yes' : 'no') + '<br>');
        }
       
        // Statistical test stuff.
	var fisher_method = data['options']['pvalue_method'] == "Fisher's exact test";
	var fisher_partition = fisher_method && ! data['options']['fix_partition'];
	var fisher_motif = fisher_method && data['options']['fix_partition'];
        if (data['options']['pvalue_method'] == "Pearson's correlation coefficient" 
          || data['options']['pvalue_method'] == "Spearman's correlation coefficient") {
          document.write('<br>Statistical test: the <i>p</i>-value of ' + data['options']['pvalue_method'] + '<br>');
        } else {
          var extra = fisher_partition ? " (optimized over partitions)" : fisher_motif ? " (optimized over motif scores)" : "";
          document.write('<br>Statistical test: ' + data['options']['pvalue_method'] + extra + '<br>');
        }

	// Labeling positives
        var fasta_scores = data['sequence_db']['fasta_scores'];
        if (data['control_db']) {
          document.write("Labeling positives: all " + data['sequence_db']['count'] + " primary sequences are labeled as 'positive';" 
            + " all " + data['control_db']['count'] + " control sequences are labeled as 'negative'<br>");
        } else if (data['options']['fix_partition']) {
          if (data['options']['positive_list'] == 'FASTA') {
            var extra1 = fasta_scores ? " sequences with the lowest FASTA scores" : " sequences";
            if (fasta_scores) {
              document.write("Labeling positives: the " + data['options']['partition'] + " sequences with the lowest FASTA scores are labeled as 'positive' (fixed partition)<br>");
            } else {
              document.write("Labeling positives: the first " + data['options']['partition'] + " sequences are labeled as 'positive' (fixed partition)<br>");
            }
          } else {
            document.write("Labeling positives: the " + data['options']['partition'] + " sequences with the highest motif scores are labeled as 'positive (fixed partition)'<br>");
          }
        } else {
          if (data['options']['positive_list'] == 'FASTA') {
            if (fasta_scores) {
              document.write("Labeling positives: sequences with FASTA scores &le; 'Pos Thresh' are labeled as 'positive'</br>");
            } else {
              document.write("Labeling positives: sequences with input ranks &le; 'Pos Thresh' are labeled as 'positive'</br>");
            }
          } else {
            document.write("Labeling positives: sequences with the sequence motif scores &ge; 'Pos Thresh' are labeled as 'positive'<br>");
          }
        }

        // Classifying positives
        if (fisher_method) {
          if (data['options']['fix_partition']) {
            document.write("Classifying positives: sequences with the motif scores &ge; 'TP Thresh' are classified as positive<br>");
          } else {
            if (data['options']['positive_list'] == "FASTA") {
              document.write("Classifying positives: sequences with motif score &ge; " + 100*data['options']['hit_odds_fraction'] + "% of the maximum possible score are classified as 'positive'<br>");
            } else {
              document.write("Classifying positives: sequences with FASTA score &le; " + data['options']['fisher_fasta_threshold'] + " are classified as 'positive'<br>");
            }
          }
        } else {
          document.write("Classifying positives: not applicable; the test is based on only the 'positive' sequences and the remaining sequences are ignored <br>");
        }

        // Special for no-threshold methods
        if (data['options']['pvalue_method'] == "Pearson's correlation coefficient" 
          || data['options']['pvalue_method'] == "Spearman's correlation coefficient") {
          var f = data['options']['log_fscores'] ? 'log(FASTA ID value)' : 'FASTA ID value';
          var m = data['options']['log_pwmscores'] ? 'log(sequence motif score)' : 'sequence motif score';
          if (data['options']['linreg_switchxy']) {
            document.write('X: ' + m + '<br>');
            document.write('Y: ' + f + '<br>');
          } else {
            document.write('X: ' + f + '<br>');
            document.write('Y: ' + m + '<br>');
          }
        }

	// General stuff.
        if (fisher_partition) {
          document.write('Observed sequence motif scores (maximum median of all motifs): ' + data['max_median_sequence_score'] + '</br>');
        }
        if (data['motifs'].length != 0) {
          // document.write('Number of multiple tests for Bonferroni correction of <i>p</i>-value: ' + data['motifs'][0]['number_of_tests']);
	  document.write('<br><i>E</i>-value threshold for reporting results: ' + 
	    data['options']['evalue_report_threshold'] + '<br>');
        }
      </script>

      </p>
      <table class="results" id="results">
        <thead>
          <tr>
            <th class="col_logo">Logo</th>
            <th class="col_db">Database&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_db')"></div></th>
            <th>ID&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_id')"></div></th>
            <th class="col_name">Alt ID&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_name')"></div></th>
            <!-- Used by: all methods -->
            <th class="col_corrected_pvalue"><i>p</i>-value&nbsp;<div 
                class="help" onclick="help_popup(this, 'pop_adj_pvalue')"></div></th>
            <th class="col_evalue"><i>E</i>-value&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_evalue')"></div></th>
            <th class="col_p_thresh">Pos Thresh &nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_p_thresh')"></div></th>
            <th class="col_pos">Pos&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_pos')"></div></th>
            <th class="col_neg">Neg&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_neg')"></div></th>
            <th class="col_tp_thresh">TP Thresh &nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_tp_thresh')"></div></th>
            <th class="col_tp">TP (%)&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_tp')"></div></th>
            <th class="col_fp">FP (%)&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_fp')"></div></th>
            <!-- Used by: LINREG_METHOD -->
            <th class="col_pearsons_rho">Pearsons's CC (&rho;) &nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_pearsons_rho')"></div></th>
            <!-- Used by: SPEARMAN_METHOD -->
            <th class="col_spearmans_rho">Spearman's Rank CC (&rho;) &nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_spearmans_rho')"></div></th>
            <!-- Used by: LINREG_METHOD -->
            <th class="col_mean_square_error">Mean squared error&nbsp;<div 
                class="help" onclick="help_popup(this, 'pop_mse')"></div></th>
            <th class="col_m">m&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_m')"></div></th>
            <th class="col_b">b&nbsp;<div class="help" 
                onclick="help_popup(this, 'pop_b')"></div></th>
          </tr>
        </thead>
        <tbody>
        </tbody>
      </table>
      <script>
        if (data['motifs'].length > 0) {
          make_results_table();
        } else {
          document.write("<br><b>No motifs passed the <i>E</i>-value threshold (" + data['options']['evalue_report_threshold'] + ").</b>");
        }
      </script>
    </div>
    <h2 id="input_sec" class="mainh pad2">Input Files</h2>
    <div class="box">
      <h4>Alphabet</h4>
      <span id="bg_source"></span>
      <div class="pad" id="alpha_bg"></div>
      <script>
      {
        $("bg_source").appendChild(make_background_source("Background source", data.background['source']));
        $("alpha_bg").appendChild(make_alpha_bg_table(ame_alphabet));
      }
      </script>
      <h4>Sequences</h4>
        <!-- This script prints the whole table so things will be "balanced". -->
	<script>
          {
            var db = data['sequence_db'];
            document.write("<table class=\"inputs\">\n");
            if (data['control_db']) {
              document.write("<tr><th>Primary Sequences</th><th>Number</th><th>Control Sequences</th><th>Number</th></tr>");
            } else {
              document.write("<tr><th>Primary Sequences</th><th>Number</th></tr>");
            }
            document.write("<tr><td>" + db['source'] + "</td>");
            document.write("<td>" + db['count'] + "</td>");
            if (data['control_db']) {
              var db = data['control_db'];
              if (db['source'] == '--shuffle--') {
		document.write("<td>primary sequences shuffled conserving " + data['options']['kmer'] + "-mer frequencies</td>");
              } else {
		document.write("<td>" + db['source'] + "</td>");
              }
              document.write("<td>" + db['count'] + "</td>");
            }
            document.write("</tr></table>");
          }
	</script>
      <h4>Motifs</h4>
      <table class="inputs">
        <tr><th>Database</th><th>Source</th><th>Motif Count</th></tr>
        <script>
        {
          var motif_dbs = data['motif_dbs'];
          for (var i = 0; i < motif_dbs.length; i++) {
            var db = motif_dbs[i];
            document.write("<tr>");
            document.write("<td>"  + db['name'] + "</td>");
            document.write("<td>" + db['source'] + "</td>");
            document.write("<td>" + db['count'] + "</td>");
            document.write("</tr>");
          }
        }
        </script>
      </table>
    </div>
    <div id="info_sec" class="bar">
      <div class="subsection">
        <h5 id="version">AME version</h5>
        <script>document.write(data["version"]);</script> 
        (Release date: <script>document.write(data["release"]);</script>)<br>
        Copyright &copy; Robert McLeay &amp; Timothy Bailey 
        <a href="mailto:meme-suite@uw.edu">meme-suite@uw.edu</a>, 2009.
      </div>
      <div class="subsection">
	<h5 id="command">Command line summary</h5>
	<textarea id="cmd" rows="5" style="width:100%;" readonly="readonly">
	</textarea>
	<script>$("cmd").value = data["cmd"].join(" ");</script>
      </div>
    </div>
    <div id="scrollpad"></div>
  </body>
</html>
